{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819c0140-24d4-4a83-9cdd-9f1f21a18d53",
   "metadata": {},
   "source": [
    "# BTE metakg visualization\n",
    "\n",
    "The goal of this notebook is to visualize the metaKG that BTE uses.  Similar to the [subway diagram](https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/diagrams/smartapi_metagraph.png) we've used before, but updated to the size and scale of the current metakg.\n",
    "\n",
    "This notebook takes as input an ndson file with the SmartAPI metakg (originally provided by Chunlei on 2023-03-01).\n",
    "\n",
    "Optimizations\n",
    "* remove less-commonly-used node types from subject/object\n",
    "* only count in one direction (`A-treats-B` gets merged with `B-treated_by-A`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "472db681-b76f-48e7-b9f6-86eb863d3dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json5\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3b92f98-9a9e-4890-bf19-a58ccb51bc48",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in data\n",
    "\n",
    "First, read in the Smart API ndjson file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5af40473-2748-4a6c-b16f-c5db1a830b31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Occurrent</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175543</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>causes</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175544</th>\n",
       "      <td>Transcript</td>\n",
       "      <td>CellLine</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175545</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175546</th>\n",
       "      <td>Device</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175547</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>has_participant</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175548 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 subject             object  \\\n",
       "0       AnatomicalEntity  MolecularActivity   \n",
       "1       AnatomicalEntity    MolecularEntity   \n",
       "2       AnatomicalEntity         NamedThing   \n",
       "3       AnatomicalEntity  NucleicAcidEntity   \n",
       "4       AnatomicalEntity          Occurrent   \n",
       "...                  ...                ...   \n",
       "175543              Cell         Phenomenon   \n",
       "175544        Transcript           CellLine   \n",
       "175545           Disease  PhenotypicFeature   \n",
       "175546            Device            Vitamin   \n",
       "175547         Procedure          Procedure   \n",
       "\n",
       "                                 predicate  \\\n",
       "0                      affects_activity_of   \n",
       "1                      affects_activity_of   \n",
       "2                      affects_activity_of   \n",
       "3                      affects_activity_of   \n",
       "4                      affects_activity_of   \n",
       "...                                    ...   \n",
       "175543                              causes   \n",
       "175544           physically_interacts_with   \n",
       "175545  entity_positively_regulates_entity   \n",
       "175546  entity_negatively_regulates_entity   \n",
       "175547                     has_participant   \n",
       "\n",
       "                                                      api provided_by  \n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "...                                                   ...         ...  \n",
       "175543  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175544  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175545  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175546  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175547  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "\n",
       "[175548 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_json('data/smartapi_metakg_03012023.ndjson.gz', lines=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e981e0d-f064-4f5b-b7a3-50929a1e58b7",
   "metadata": {},
   "source": [
    "parse out a couple lines for the API name and ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "479dce85-28b0-45e2-b494-7756a1b42bed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>c409da43f5177ae52392e2b346dbc585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>c409da43f5177ae52392e2b346dbc585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>c409da43f5177ae52392e2b346dbc585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>c409da43f5177ae52392e2b346dbc585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Occurrent</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>c409da43f5177ae52392e2b346dbc585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175543</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>causes</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175544</th>\n",
       "      <td>Transcript</td>\n",
       "      <td>CellLine</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175545</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175546</th>\n",
       "      <td>Device</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175547</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>has_participant</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175548 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 subject             object  \\\n",
       "0       AnatomicalEntity  MolecularActivity   \n",
       "1       AnatomicalEntity    MolecularEntity   \n",
       "2       AnatomicalEntity         NamedThing   \n",
       "3       AnatomicalEntity  NucleicAcidEntity   \n",
       "4       AnatomicalEntity          Occurrent   \n",
       "...                  ...                ...   \n",
       "175543              Cell         Phenomenon   \n",
       "175544        Transcript           CellLine   \n",
       "175545           Disease  PhenotypicFeature   \n",
       "175546            Device            Vitamin   \n",
       "175547         Procedure          Procedure   \n",
       "\n",
       "                                 predicate  \\\n",
       "0                      affects_activity_of   \n",
       "1                      affects_activity_of   \n",
       "2                      affects_activity_of   \n",
       "3                      affects_activity_of   \n",
       "4                      affects_activity_of   \n",
       "...                                    ...   \n",
       "175543                              causes   \n",
       "175544           physically_interacts_with   \n",
       "175545  entity_positively_regulates_entity   \n",
       "175546  entity_negatively_regulates_entity   \n",
       "175547                     has_participant   \n",
       "\n",
       "                                                      api provided_by  \\\n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN   \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN   \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN   \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN   \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN   \n",
       "...                                                   ...         ...   \n",
       "175543  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN   \n",
       "175544  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN   \n",
       "175545  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN   \n",
       "175546  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN   \n",
       "175547  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN   \n",
       "\n",
       "                                      api_name  \\\n",
       "0                                   CAM-KP API   \n",
       "1                                   CAM-KP API   \n",
       "2                                   CAM-KP API   \n",
       "3                                   CAM-KP API   \n",
       "4                                   CAM-KP API   \n",
       "...                                        ...   \n",
       "175543  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "175544  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "175545  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "175546  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "175547  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "\n",
       "                                  api_id  \n",
       "0       c409da43f5177ae52392e2b346dbc585  \n",
       "1       c409da43f5177ae52392e2b346dbc585  \n",
       "2       c409da43f5177ae52392e2b346dbc585  \n",
       "3       c409da43f5177ae52392e2b346dbc585  \n",
       "4       c409da43f5177ae52392e2b346dbc585  \n",
       "...                                  ...  \n",
       "175543  e248aefca0f469229e82cca80fbabc89  \n",
       "175544  e248aefca0f469229e82cca80fbabc89  \n",
       "175545  e248aefca0f469229e82cca80fbabc89  \n",
       "175546  e248aefca0f469229e82cca80fbabc89  \n",
       "175547  e248aefca0f469229e82cca80fbabc89  \n",
       "\n",
       "[175548 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(api_name = lambda x: pd.json_normalize(x['api'])['name'])\n",
    "df = df.assign(api_id = lambda x: pd.json_normalize(x['api'])['smartapi.id'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aafe106-7f9a-4713-84fe-59ba83195f1b",
   "metadata": {},
   "source": [
    "Read in the BTE config file that specifies currently-allowed APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "837e2003-2ac6-429b-a123-2715f1b93da0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'include': [{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       "  {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       "  {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       "  {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       "  {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "   'name': 'Ontology Lookup Service API'},\n",
       "  {'id': '38e9e5169a72aee3659c9ddba956790d',\n",
       "   'name': 'BioThings BindingDB API'},\n",
       "  {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "   'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       "  {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "   'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       "  {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       "  {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       "  {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       "  {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "   'name': 'BioThings EBIgene2phenotype API'},\n",
       "  {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "   'name': 'BioThings GO Biological Process API'},\n",
       "  {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "   'name': 'BioThings GO Cellular Component API'},\n",
       "  {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "   'name': 'BioThings GO Molecular Function API'},\n",
       "  {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       "  {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       "  {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       "  {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "   'name': 'BioThings MGIgene2phenotype API'},\n",
       "  {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       "  {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       "  {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       "  {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       "  {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       "  {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       "  {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       "  {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       "  {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "   'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       "  {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "   'name': 'Multiomics ClinicalTrials KP'},\n",
       "  {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "   'name': 'Multiomics EHR Risk KP API'},\n",
       "  {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "   'name': 'Multiomics Wellness KP API'},\n",
       "  {'id': '978fe380a147a8641caf72320862697b',\n",
       "   'name': 'Text Mining Targeted Association API'},\n",
       "  {'id': '25085b05fd1afcebb497724d147cfb44',\n",
       "   'name': 'Automat-biolink(Trapi v1.3.0)'},\n",
       "  {'id': '89aa98098ba4329aa49b43ff8d21ffbb',\n",
       "   'name': 'Automat-ctd(Trapi v1.3.0)'},\n",
       "  {'id': '463525b319d7e1ef8f705bce953d01bd',\n",
       "   'name': 'Automat-cord19(Trapi v1.3.0)'},\n",
       "  {'id': '539873b1f2f2eb913efaee411e09eaa7',\n",
       "   'name': 'Automat-drug-central(Trapi v1.3.0)'},\n",
       "  {'id': '272c2357763e4faf737f1cba94beaa1d',\n",
       "   'name': 'Automat-gtex(Trapi v1.3.0)'},\n",
       "  {'id': 'fb1f3780ad67b030cb0617363afa8f61',\n",
       "   'name': 'Automat-gwas-catalog(Trapi v1.3.0)'},\n",
       "  {'id': 'aacdd9e863bb659574d302f084084976',\n",
       "   'name': 'Automat-gtopdb(Trapi v1.3.0)'},\n",
       "  {'id': '7382f0fabffce3cc7f7b8b6358c69259',\n",
       "   'name': 'Automat-hgnc(Trapi v1.3.0)'},\n",
       "  {'id': 'ee2d2eae42ca30fe82946d3f42febaa0',\n",
       "   'name': 'Automat-hmdb(Trapi v1.3.0)'},\n",
       "  {'id': '830600da121a5accc955cbf62e60f802',\n",
       "   'name': 'Automat-hetio(Trapi v1.3.0)'},\n",
       "  {'id': 'f30339a6894a146a19a34974712ca2e3',\n",
       "   'name': 'Automat-human-goa(Trapi v1.3.0)'},\n",
       "  {'id': '44e7a1147ca8657f50af6bb25982762d',\n",
       "   'name': 'Automat-icees-kg(Trapi v1.3.0)'},\n",
       "  {'id': 'e0687431a9ff88344d20e83e0c99ee7d',\n",
       "   'name': 'Automat-intact(Trapi v1.3.0)'},\n",
       "  {'id': '4c6f9117581531161849e60ea906f0be',\n",
       "   'name': 'Automat-ontology-hierarchy(Trapi v1.3.0)'},\n",
       "  {'id': '2a879882329b000c7e7f08c2d71ccffd',\n",
       "   'name': 'Automat-panther(Trapi v1.3.0)'},\n",
       "  {'id': '0295640b4d060133e4296dda4c31da47',\n",
       "   'name': 'Automat-pharos(Trapi v1.3.0)'},\n",
       "  {'id': 'ef9027a7d2246c6540cc7b3ce202d89f',\n",
       "   'name': 'Automat-uberongraph(Trapi v1.3.0)'},\n",
       "  {'id': '79310edd01cae96f3d8495250e625886',\n",
       "   'name': 'Automat-viral-proteome(Trapi v1.3.0)'},\n",
       "  {'id': '51c178099fa2dc99b5d8fff8bf9f1a0d', 'name': 'COHD TRAPI 1.3'},\n",
       "  {'id': '855adaa128ce5aa58a091d99e520d396',\n",
       "   'name': 'Connections Hypothesis Provider API'}]}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_url = \"https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/src/config/apis.js\"\n",
    "r = requests.get(bte_config_url)\n",
    "str_bte_config = r.text\n",
    "#print(str_bte_config)\n",
    "str_bte_config = re.sub(\"exports.API_LIST = \",  \"\", str_bte_config)                       # remove variable assignment step\n",
    "str_bte_config = re.sub(\"\\s*//.*\",              \"\", str_bte_config)                       # remove commented lines\n",
    "str_bte_config = re.sub(r'^$\\n',                '', str_bte_config, flags=re.MULTILINE)   # remove blank lines\n",
    "str_bte_config = re.sub(r',\\s*exclude:[^\\]]*]', '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "str_bte_config = re.sub(r';$',                  '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "#print(str_bte_config)\n",
    "\n",
    "bte_config = json5.loads(str_bte_config)\n",
    "bte_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "689ae4af-34a1-4a5c-97e3-9f97fce9ef7e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['d22b657426375a5295e7da8a303b9893',\n",
       " '43af91b3d7cae43591083bff9d75c6dd',\n",
       " 'dca415f2d792976af9d642b7e73f7a41',\n",
       " '1f277e1563fcfd124bfae2cc3c4bcdec',\n",
       " '1c056ffc7ed0dd1229e71c4752239465',\n",
       " '38e9e5169a72aee3659c9ddba956790d',\n",
       " '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       " 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       " '00fb85fc776279163199e6c50f6ddfc6',\n",
       " 'e3edd325c76f2992a111b43a907a4870',\n",
       " 'a7f784626a426d054885a5f33f17d3f8',\n",
       " '1f47552dabd67351d4c625adb0a10d00',\n",
       " 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       " 'f339b28426e7bf72028f60feefcd7465',\n",
       " '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       " '316eab811fd9ef1097df98bcaa9f7361',\n",
       " 'a5b0ec6bfde5008984d4b6cde402d61f',\n",
       " '32f36164fabed5d3abe6c2fd899c9418',\n",
       " '77ed27f111262d0289ed4f4071faa619',\n",
       " 'edeb26858bd27d0322af93e7a9e08761',\n",
       " '03283cc2b21c077be6794e1704b1d230',\n",
       " '1d288b3a3caf75d541ffaae3aab386c8',\n",
       " 'ec6d76016ef40f284359d17fbf78df20',\n",
       " '8f08d1446e0bb9c2b323713ce83e2bd3',\n",
       " '671b45c0301c8624abbd26ae78449ca2',\n",
       " '59dce17363dce279d389100834e43648',\n",
       " '09c8782d9f4027712e65b95424adba79',\n",
       " 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       " '08a5ddcde71b4bf838327ef469076acd',\n",
       " 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       " '02af7d098ab304e80d6f4806c3527027',\n",
       " '978fe380a147a8641caf72320862697b',\n",
       " '25085b05fd1afcebb497724d147cfb44',\n",
       " '89aa98098ba4329aa49b43ff8d21ffbb',\n",
       " '463525b319d7e1ef8f705bce953d01bd',\n",
       " '539873b1f2f2eb913efaee411e09eaa7',\n",
       " '272c2357763e4faf737f1cba94beaa1d',\n",
       " 'fb1f3780ad67b030cb0617363afa8f61',\n",
       " 'aacdd9e863bb659574d302f084084976',\n",
       " '7382f0fabffce3cc7f7b8b6358c69259',\n",
       " 'ee2d2eae42ca30fe82946d3f42febaa0',\n",
       " '830600da121a5accc955cbf62e60f802',\n",
       " 'f30339a6894a146a19a34974712ca2e3',\n",
       " '44e7a1147ca8657f50af6bb25982762d',\n",
       " 'e0687431a9ff88344d20e83e0c99ee7d',\n",
       " '4c6f9117581531161849e60ea906f0be',\n",
       " '2a879882329b000c7e7f08c2d71ccffd',\n",
       " '0295640b4d060133e4296dda4c31da47',\n",
       " 'ef9027a7d2246c6540cc7b3ce202d89f',\n",
       " '79310edd01cae96f3d8495250e625886',\n",
       " '51c178099fa2dc99b5d8fff8bf9f1a0d',\n",
       " '855adaa128ce5aa58a091d99e520d396']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_ids = [ x['id'] for x in bte_config['include'] ]\n",
    "bte_config_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96ca64b-7295-4039-a8c7-9fe1c9912473",
   "metadata": {},
   "source": [
    "## Join SmartAPI data with BTE config IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7bf98b05-b49a-4308-88ec-eed57eba83a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>SequenceVariant</td>\n",
       "      <td>Gene</td>\n",
       "      <td>is_sequence_variant_of</td>\n",
       "      <td>{'name': 'LitVar API', 'smartapi': {'metadata'...</td>\n",
       "      <td>infores:dbsnp</td>\n",
       "      <td>LitVar API</td>\n",
       "      <td>dca415f2d792976af9d642b7e73f7a41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5115</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>orthologous_to</td>\n",
       "      <td>{'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-biolink(Trapi v1.3.0)</td>\n",
       "      <td>25085b05fd1afcebb497724d147cfb44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5116</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>superclass_of</td>\n",
       "      <td>{'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-biolink(Trapi v1.3.0)</td>\n",
       "      <td>25085b05fd1afcebb497724d147cfb44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5117</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>biomarker_for</td>\n",
       "      <td>{'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-biolink(Trapi v1.3.0)</td>\n",
       "      <td>25085b05fd1afcebb497724d147cfb44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>phenotype_of</td>\n",
       "      <td>{'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-biolink(Trapi v1.3.0)</td>\n",
       "      <td>25085b05fd1afcebb497724d147cfb44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130260</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>{'name': 'Text Mining Targeted Association API...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130261</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>{'name': 'Text Mining Targeted Association API...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130262</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>{'name': 'Text Mining Targeted Association API...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130263</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>{'name': 'Text Mining Targeted Association API...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130264</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>{'name': 'Text Mining Targeted Association API...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4781 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject         object               predicate  \\\n",
       "3000      SequenceVariant           Gene  is_sequence_variant_of   \n",
       "5115                 Gene           Gene          orthologous_to   \n",
       "5116                 Gene           Gene           superclass_of   \n",
       "5117                 Gene           Gene           biomarker_for   \n",
       "5118                 Gene           Gene            phenotype_of   \n",
       "...                   ...            ...                     ...   \n",
       "130260  PhenotypicFeature  SmallMolecule       contribution_from   \n",
       "130261      SmallMolecule        Disease          contributes_to   \n",
       "130262            Disease  SmallMolecule       contribution_from   \n",
       "130263      SmallMolecule        Disease          contributes_to   \n",
       "130264            Disease  SmallMolecule       contribution_from   \n",
       "\n",
       "                                                      api    provided_by  \\\n",
       "3000    {'name': 'LitVar API', 'smartapi': {'metadata'...  infores:dbsnp   \n",
       "5115    {'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...            NaN   \n",
       "5116    {'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...            NaN   \n",
       "5117    {'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...            NaN   \n",
       "5118    {'name': 'Automat-biolink(Trapi v1.3.0)', 'sma...            NaN   \n",
       "...                                                   ...            ...   \n",
       "130260  {'name': 'Text Mining Targeted Association API...            NaN   \n",
       "130261  {'name': 'Text Mining Targeted Association API...            NaN   \n",
       "130262  {'name': 'Text Mining Targeted Association API...            NaN   \n",
       "130263  {'name': 'Text Mining Targeted Association API...            NaN   \n",
       "130264  {'name': 'Text Mining Targeted Association API...            NaN   \n",
       "\n",
       "                                    api_name                            api_id  \n",
       "3000                              LitVar API  dca415f2d792976af9d642b7e73f7a41  \n",
       "5115           Automat-biolink(Trapi v1.3.0)  25085b05fd1afcebb497724d147cfb44  \n",
       "5116           Automat-biolink(Trapi v1.3.0)  25085b05fd1afcebb497724d147cfb44  \n",
       "5117           Automat-biolink(Trapi v1.3.0)  25085b05fd1afcebb497724d147cfb44  \n",
       "5118           Automat-biolink(Trapi v1.3.0)  25085b05fd1afcebb497724d147cfb44  \n",
       "...                                      ...                               ...  \n",
       "130260  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "130261  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "130262  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "130263  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "130264  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "\n",
       "[4781 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte = df.query('api_id in @bte_config_ids')\n",
    "df_bte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6764f282-a1e6-4b3c-936f-11b4d3971792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_bte.to_csv(\"results/bte_operations.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac384894-b9af-47c0-b0b5-e4920b30cf17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BioThings SEMMEDDB API                     2497\n",
       "Automat-icees-kg(Trapi v1.3.0)              454\n",
       "Automat-ctd(Trapi v1.3.0)                   441\n",
       "Multiomics EHR Risk KP API                  204\n",
       "Multiomics Wellness KP API                  196\n",
       "Automat-hetio(Trapi v1.3.0)                 190\n",
       "Automat-biolink(Trapi v1.3.0)               134\n",
       "Automat-pharos(Trapi v1.3.0)                 99\n",
       "Automat-hmdb(Trapi v1.3.0)                   66\n",
       "Automat-human-goa(Trapi v1.3.0)              58\n",
       "COHD TRAPI 1.3                               50\n",
       "Automat-gtopdb(Trapi v1.3.0)                 42\n",
       "Automat-viral-proteome(Trapi v1.3.0)         35\n",
       "Automat-panther(Trapi v1.3.0)                34\n",
       "Text Mining Targeted Association API         32\n",
       "MyDisease.info API                           26\n",
       "BioThings DGIdb API                          24\n",
       "MyChem.info API                              20\n",
       "BioLink API                                  19\n",
       "Multiomics BigGIM-DrugResponse KP API        18\n",
       "MyGene.info API                              18\n",
       "UBERON Ontology API                          16\n",
       "MyVariant.info API                           16\n",
       "Connections Hypothesis Provider API          12\n",
       "Automat-intact(Trapi v1.3.0)                 11\n",
       "Gene Ontology Biological Process API         10\n",
       "BioThings iDISK API                           8\n",
       "BioThings pfocr API                           6\n",
       "BioThings Rhea API                            5\n",
       "MGIgene2phenotype API                         4\n",
       "Gene Ontology Cellular Component API          4\n",
       "EBIgene2phenotype API                         4\n",
       "Gene Ontology Molecular Activity API          4\n",
       "Automat-hgnc(Trapi v1.3.0)                    4\n",
       "BioThings BioPlanet Pathway-Disease API       2\n",
       "Multiomics ClinicalTrials KP                  2\n",
       "BioThings GTRx API                            2\n",
       "BioThings BioPlanet Pathway-Gene API          2\n",
       "BioThings BindingDB API                       2\n",
       "BioThings DDInter API                         2\n",
       "Human Phenotype Ontology API                  2\n",
       "DISEASES API                                  2\n",
       "QuickGO API                                   1\n",
       "Ontology Lookup Service API                   1\n",
       "EBI Proteins API                              1\n",
       "LitVar API                                    1\n",
       "Name: api_name, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte['api_name'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398517c7-4afd-4e3c-b9e4-993dce33e07d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Summarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515b78f8-2606-4c88-9403-60e0fc2e34c6",
   "metadata": {},
   "source": [
    "### by subject, object; count # of APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a53da5f5-34dd-4f43-bdad-855bc4139fb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           subject                           object  count  \\\n",
      "0         Activity                         Activity      1   \n",
      "1         Activity                       Annotation      1   \n",
      "2         Activity                BiologicalProcess      1   \n",
      "3         Activity                   ChemicalEntity      1   \n",
      "4         Activity                 ChemicalExposure      1   \n",
      "5         Activity                ClinicalAttribute      1   \n",
      "6         Activity             ClinicalIntervention      1   \n",
      "7         Activity          ComplexMolecularMixture      1   \n",
      "8         Activity                           Device      1   \n",
      "9         Activity                          Disease      1   \n",
      "..             ...                              ...    ...   \n",
      "672  SmallMolecule                       Phenomenon      1   \n",
      "673  SmallMolecule                PhenotypicFeature      9   \n",
      "674  SmallMolecule             PhysiologicalProcess      3   \n",
      "675  SmallMolecule                      Polypeptide      7   \n",
      "676  SmallMolecule  PopulationOfIndividualOrganisms      1   \n",
      "677  SmallMolecule                        Procedure      5   \n",
      "678  SmallMolecule                          Protein      7   \n",
      "679  SmallMolecule                      Publication      1   \n",
      "680  SmallMolecule                    SmallMolecule     12   \n",
      "681      Treatment                          Disease      1   \n",
      "\n",
      "                                                  list  \n",
      "0                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "1                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "2                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "3                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "4                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "5                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "6                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "7                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "8                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "9                     [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "..                                                 ...  \n",
      "672                      [Automat-hetio(Trapi v1.3.0)]  \n",
      "673  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
      "674  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
      "675  [Automat-biolink(Trapi v1.3.0), Automat-hetio(...  \n",
      "676                   [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "677  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
      "678  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
      "679                   [Automat-icees-kg(Trapi v1.3.0)]  \n",
      "680  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
      "681                     [Multiomics ClinicalTrials KP]  \n",
      "\n",
      "[682 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"api_name\"]]\n",
    "api_stats = df1.groupby(['subject','object'], group_keys=False)['api_name'].nunique().rename(\"count\").to_frame()\n",
    "api_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['api_name'].unique().apply(list)\n",
    "api_stats = api_stats.reset_index()\n",
    "\n",
    "with pd.option_context('display.min_rows', 20,\n",
    "                       'display.max_columns', None,\n",
    "                       'display.precision', 3,\n",
    "                       ):\n",
    "    print(api_stats)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d73842dd-d6a2-4309-b820-1594a85c44d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "api_stats.to_csv(\"results/api_stats.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abbb70cd-070a-4380-b9e2-63cb798d09f8",
   "metadata": {},
   "source": [
    "### by subject, object; count # of predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cc40d6f8-87ed-48b1-b8a1-d3c04a9d8893",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           subject                           object  count  \\\n",
      "0         Activity                         Activity      1   \n",
      "1         Activity                       Annotation      1   \n",
      "2         Activity                BiologicalProcess      1   \n",
      "3         Activity                   ChemicalEntity      1   \n",
      "4         Activity                 ChemicalExposure      1   \n",
      "5         Activity                ClinicalAttribute      1   \n",
      "6         Activity             ClinicalIntervention      1   \n",
      "7         Activity          ComplexMolecularMixture      1   \n",
      "8         Activity                           Device      1   \n",
      "9         Activity                          Disease      1   \n",
      "..             ...                              ...    ...   \n",
      "672  SmallMolecule                       Phenomenon      1   \n",
      "673  SmallMolecule                PhenotypicFeature     16   \n",
      "674  SmallMolecule             PhysiologicalProcess      9   \n",
      "675  SmallMolecule                      Polypeptide     13   \n",
      "676  SmallMolecule  PopulationOfIndividualOrganisms      1   \n",
      "677  SmallMolecule                        Procedure      8   \n",
      "678  SmallMolecule                          Protein      7   \n",
      "679  SmallMolecule                      Publication      1   \n",
      "680  SmallMolecule                    SmallMolecule     21   \n",
      "681      Treatment                          Disease      1   \n",
      "\n",
      "                                                  list  \n",
      "0                                    [correlated_with]  \n",
      "1                                    [correlated_with]  \n",
      "2                                    [correlated_with]  \n",
      "3                                    [correlated_with]  \n",
      "4                                    [correlated_with]  \n",
      "5                                    [correlated_with]  \n",
      "6                                    [correlated_with]  \n",
      "7                                    [correlated_with]  \n",
      "8                                    [correlated_with]  \n",
      "9                                    [correlated_with]  \n",
      "..                                                 ...  \n",
      "672                                           [causes]  \n",
      "673  [treats, correlated_with, causes, ameliorates,...  \n",
      "674  [correlated_with, causes, disrupts, affects, e...  \n",
      "675  [subclass_of, superclass_of, part_of, interact...  \n",
      "676                                  [correlated_with]  \n",
      "677  [correlated_with, causes, positively_correlate...  \n",
      "678  [subclass_of, correlated_with, increases_trans...  \n",
      "679                                  [correlated_with]  \n",
      "680  [subclass_of, superclass_of, correlated_with, ...  \n",
      "681                                       [related_to]  \n",
      "\n",
      "[682 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"predicate\"]]\n",
    "predicate_stats = df1.groupby(['subject','object'], group_keys=False)['predicate'].nunique().rename(\"count\").to_frame()\n",
    "predicate_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['predicate'].unique().apply(list)\n",
    "predicate_stats = predicate_stats.reset_index()\n",
    "\n",
    "with pd.option_context('display.min_rows', 20,\n",
    "                       'display.max_columns', None,\n",
    "                       'display.precision', 3,\n",
    "                       ):\n",
    "    print(predicate_stats)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a2da419-97e7-438d-a294-69fa0946081f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "predicate_stats.to_csv(\"results/predicate_stats.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e955923-8436-4453-b380-2f8b8bb9ce3f",
   "metadata": {},
   "source": [
    "## Filter by most common types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9104cf-516b-4457-9632-0ff9a42ff036",
   "metadata": {},
   "source": [
    "Filter to only include the most common types of entities.  Also, since we _mostly_ have the same info in both directions, only keep one direction to simplify visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad7563bf-3c03-43ac-811c-186145df2acb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BiologicalProcess',\n",
       " 'ChemicalEntity',\n",
       " 'Disease',\n",
       " 'Gene',\n",
       " 'MolecularMixture',\n",
       " 'PhenotypicFeature',\n",
       " 'PhysiologicalProcess',\n",
       " 'Procedure',\n",
       " 'Protein',\n",
       " 'SmallMolecule'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NUM_TYPES_TO_KEEP = 10\n",
    "\n",
    "keep = set(pd.concat([predicate_stats['subject'], predicate_stats['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4b29e182-3e47-4e8f-9925-3eb0803566a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>10</td>\n",
       "      <td>[subclass_of, superclass_of, process_negativel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>3</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Gene</td>\n",
       "      <td>3</td>\n",
       "      <td>[correlated_with, actively_involves, has_parti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>2</td>\n",
       "      <td>[superclass_of, correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>6</td>\n",
       "      <td>[actively_involved_in, correlated_with, preced...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>12</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>15</td>\n",
       "      <td>[treats, biomarker_for, correlated_with, ameli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>44</td>\n",
       "      <td>[correlated_with, regulates, directly_physical...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>4</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>10</td>\n",
       "      <td>[treats, correlated_with, causes, ameliorates,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>6</td>\n",
       "      <td>[correlated_with, causes, produced_by, entity_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>3</td>\n",
       "      <td>[correlated_with, causes, preceded_by]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[superclass_of, correlated_with, increases_tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>11</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>36</td>\n",
       "      <td>[contributes_to, has_phenotype, phenotype_of, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>21</td>\n",
       "      <td>[has_biomarker, phenotype_of, contribution_fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Disease</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>9</td>\n",
       "      <td>[has_biomarker, treated_by, correlated_with, i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>17</td>\n",
       "      <td>[subclass_of, has_phenotype, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>10</td>\n",
       "      <td>[correlated_with, contains_process, associated...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>15</td>\n",
       "      <td>[correlated_with, positively_correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Protein</td>\n",
       "      <td>5</td>\n",
       "      <td>[correlated_with, contribution_from, is_amelio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>23</td>\n",
       "      <td>[has_biomarker, treated_by, correlated_with, i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>32</td>\n",
       "      <td>[orthologous_to, superclass_of, biomarker_for,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>Gene</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>30</td>\n",
       "      <td>[correlated_with, regulated_by, directly_physi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>12</td>\n",
       "      <td>[has_phenotype, correlated_with, genetic_assoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>7</td>\n",
       "      <td>[correlated_with, genetic_association, causes,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein</td>\n",
       "      <td>10</td>\n",
       "      <td>[correlated_with, increases_secretion_of, resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>61</td>\n",
       "      <td>[correlated_with, regulated_by, directly_physi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>3</td>\n",
       "      <td>[subclass_of, superclass_of, similar_to]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>8</td>\n",
       "      <td>[treats, correlated_with, causes, ameliorates,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>2</td>\n",
       "      <td>[correlated_with, causes]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>2</td>\n",
       "      <td>[correlated_with, causes]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[correlated_with, increases_transport_of]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>13</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>4</td>\n",
       "      <td>[correlated_with, associated_with, causes, man...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>12</td>\n",
       "      <td>[correlated_with, positively_correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[correlated_with, contribution_from, is_amelio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>16</td>\n",
       "      <td>[treated_by, correlated_with, caused_by, affec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>8</td>\n",
       "      <td>[correlated_with, coexists_with, affects, affe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>4</td>\n",
       "      <td>[correlated_with, affected_by, entity_positive...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>5</td>\n",
       "      <td>[correlated_with, produces, disrupted_by, affe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[correlated_with, caused_by, disrupted_by, aff...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>8</td>\n",
       "      <td>[correlated_with, positively_correlated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Protein</td>\n",
       "      <td>1</td>\n",
       "      <td>[correlated_with]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>8</td>\n",
       "      <td>[correlated_with, caused_by, positively_correl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>Protein</td>\n",
       "      <td>Protein</td>\n",
       "      <td>5</td>\n",
       "      <td>[similar_to, subclass_of, superclass_of, corre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>Protein</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>7</td>\n",
       "      <td>[superclass_of, correlated_with, affects_abund...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>21</td>\n",
       "      <td>[subclass_of, superclass_of, correlated_with, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "56      BiologicalProcess     BiologicalProcess     10   \n",
       "57      BiologicalProcess        ChemicalEntity      1   \n",
       "62      BiologicalProcess               Disease      3   \n",
       "65      BiologicalProcess                  Gene      3   \n",
       "69      BiologicalProcess      MolecularMixture      1   \n",
       "71      BiologicalProcess     PhenotypicFeature      2   \n",
       "72      BiologicalProcess  PhysiologicalProcess      1   \n",
       "74      BiologicalProcess             Procedure      1   \n",
       "75      BiologicalProcess               Protein      6   \n",
       "77      BiologicalProcess         SmallMolecule      1   \n",
       "114        ChemicalEntity        ChemicalEntity     12   \n",
       "121        ChemicalEntity               Disease     15   \n",
       "124        ChemicalEntity                  Gene     44   \n",
       "128        ChemicalEntity      MolecularMixture      4   \n",
       "132        ChemicalEntity     PhenotypicFeature     10   \n",
       "133        ChemicalEntity  PhysiologicalProcess      6   \n",
       "136        ChemicalEntity             Procedure      3   \n",
       "137        ChemicalEntity               Protein      3   \n",
       "139        ChemicalEntity         SmallMolecule     11   \n",
       "258               Disease               Disease     36   \n",
       "262               Disease                  Gene     21   \n",
       "266               Disease      MolecularMixture      9   \n",
       "269               Disease     PhenotypicFeature     17   \n",
       "270               Disease  PhysiologicalProcess     10   \n",
       "273               Disease             Procedure     15   \n",
       "274               Disease               Protein      5   \n",
       "277               Disease         SmallMolecule     23   \n",
       "344                  Gene                  Gene     32   \n",
       "349                  Gene      MolecularMixture     30   \n",
       "353                  Gene     PhenotypicFeature     12   \n",
       "354                  Gene  PhysiologicalProcess      7   \n",
       "356                  Gene             Procedure      1   \n",
       "357                  Gene               Protein     10   \n",
       "360                  Gene         SmallMolecule     61   \n",
       "434      MolecularMixture      MolecularMixture      3   \n",
       "436      MolecularMixture     PhenotypicFeature      8   \n",
       "437      MolecularMixture  PhysiologicalProcess      2   \n",
       "440      MolecularMixture             Procedure      2   \n",
       "441      MolecularMixture               Protein      2   \n",
       "443      MolecularMixture         SmallMolecule      4   \n",
       "492     PhenotypicFeature     PhenotypicFeature     13   \n",
       "493     PhenotypicFeature  PhysiologicalProcess      4   \n",
       "496     PhenotypicFeature             Procedure     12   \n",
       "497     PhenotypicFeature               Protein      3   \n",
       "500     PhenotypicFeature         SmallMolecule     16   \n",
       "522  PhysiologicalProcess  PhysiologicalProcess      8   \n",
       "525  PhysiologicalProcess             Procedure      4   \n",
       "526  PhysiologicalProcess               Protein      5   \n",
       "528  PhysiologicalProcess         SmallMolecule      9   \n",
       "590             Procedure             Procedure      8   \n",
       "591             Procedure               Protein      1   \n",
       "593             Procedure         SmallMolecule      8   \n",
       "620               Protein               Protein      5   \n",
       "622               Protein         SmallMolecule      7   \n",
       "680         SmallMolecule         SmallMolecule     21   \n",
       "\n",
       "                                                  list  \n",
       "56   [subclass_of, superclass_of, process_negativel...  \n",
       "57                                   [correlated_with]  \n",
       "62       [subclass_of, superclass_of, correlated_with]  \n",
       "65   [correlated_with, actively_involves, has_parti...  \n",
       "69                                   [correlated_with]  \n",
       "71                    [superclass_of, correlated_with]  \n",
       "72                                   [correlated_with]  \n",
       "74                                   [correlated_with]  \n",
       "75   [actively_involved_in, correlated_with, preced...  \n",
       "77                                   [correlated_with]  \n",
       "114  [subclass_of, superclass_of, correlated_with, ...  \n",
       "121  [treats, biomarker_for, correlated_with, ameli...  \n",
       "124  [correlated_with, regulates, directly_physical...  \n",
       "128  [subclass_of, superclass_of, correlated_with, ...  \n",
       "132  [treats, correlated_with, causes, ameliorates,...  \n",
       "133  [correlated_with, causes, produced_by, entity_...  \n",
       "136             [correlated_with, causes, preceded_by]  \n",
       "137  [superclass_of, correlated_with, increases_tra...  \n",
       "139  [subclass_of, superclass_of, correlated_with, ...  \n",
       "258  [contributes_to, has_phenotype, phenotype_of, ...  \n",
       "262  [has_biomarker, phenotype_of, contribution_fro...  \n",
       "266  [has_biomarker, treated_by, correlated_with, i...  \n",
       "269  [subclass_of, has_phenotype, correlated_with, ...  \n",
       "270  [correlated_with, contains_process, associated...  \n",
       "273  [correlated_with, positively_correlated_with, ...  \n",
       "274  [correlated_with, contribution_from, is_amelio...  \n",
       "277  [has_biomarker, treated_by, correlated_with, i...  \n",
       "344  [orthologous_to, superclass_of, biomarker_for,...  \n",
       "349  [correlated_with, regulated_by, directly_physi...  \n",
       "353  [has_phenotype, correlated_with, genetic_assoc...  \n",
       "354  [correlated_with, genetic_association, causes,...  \n",
       "356                                  [correlated_with]  \n",
       "357  [correlated_with, increases_secretion_of, resp...  \n",
       "360  [correlated_with, regulated_by, directly_physi...  \n",
       "434           [subclass_of, superclass_of, similar_to]  \n",
       "436  [treats, correlated_with, causes, ameliorates,...  \n",
       "437                          [correlated_with, causes]  \n",
       "440                          [correlated_with, causes]  \n",
       "441          [correlated_with, increases_transport_of]  \n",
       "443  [subclass_of, superclass_of, correlated_with, ...  \n",
       "492  [subclass_of, superclass_of, correlated_with, ...  \n",
       "493  [correlated_with, associated_with, causes, man...  \n",
       "496  [correlated_with, positively_correlated_with, ...  \n",
       "497  [correlated_with, contribution_from, is_amelio...  \n",
       "500  [treated_by, correlated_with, caused_by, affec...  \n",
       "522  [correlated_with, coexists_with, affects, affe...  \n",
       "525  [correlated_with, affected_by, entity_positive...  \n",
       "526  [correlated_with, produces, disrupted_by, affe...  \n",
       "528  [correlated_with, caused_by, disrupted_by, aff...  \n",
       "590  [correlated_with, positively_correlated_with, ...  \n",
       "591                                  [correlated_with]  \n",
       "593  [correlated_with, caused_by, positively_correl...  \n",
       "620  [similar_to, subclass_of, superclass_of, corre...  \n",
       "622  [superclass_of, correlated_with, affects_abund...  \n",
       "680  [subclass_of, superclass_of, correlated_with, ...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicate_stats_filt = predicate_stats.query(\"subject in @keep & object in @keep & subject <= object\")\n",
    "predicate_stats_filt.to_csv(\"results/predicate_stats_filt.tsv\", sep=\"\\t\")\n",
    "predicate_stats_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dabb9bd0-808b-4e46-a51c-ff15d25ac4dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-viral-proteome(Trapi v1.3.0), Automat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-viral-proteome(Trapi v1.3.0), Automat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Gene</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-viral-proteome(Trapi v1.3.0), Automat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-viral-proteome(Trapi v1.3.0), Automat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>6</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Protein</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-viral-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>11</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Ontology Looku...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>16</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Disease</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>11</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), BioThings SEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>14</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>15</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-hetio(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>Gene</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-pharo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>14</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>6</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-hetio(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hmdb(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>9</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), BioThings SEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), BioThings SEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), BioThings SEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), BioThings SEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Protein</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-icees-kg(Trapi v1.3.0), Automat-hetio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>Protein</td>\n",
       "      <td>Protein</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-viral-proteome(Trapi v1.3.0), Automat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>Protein</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>12</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-icees-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "56      BiologicalProcess     BiologicalProcess      5   \n",
       "57      BiologicalProcess        ChemicalEntity      1   \n",
       "62      BiologicalProcess               Disease      3   \n",
       "65      BiologicalProcess                  Gene      3   \n",
       "69      BiologicalProcess      MolecularMixture      1   \n",
       "71      BiologicalProcess     PhenotypicFeature      3   \n",
       "72      BiologicalProcess  PhysiologicalProcess      1   \n",
       "74      BiologicalProcess             Procedure      1   \n",
       "75      BiologicalProcess               Protein      3   \n",
       "77      BiologicalProcess         SmallMolecule      1   \n",
       "114        ChemicalEntity        ChemicalEntity      8   \n",
       "121        ChemicalEntity               Disease      6   \n",
       "124        ChemicalEntity                  Gene      7   \n",
       "128        ChemicalEntity      MolecularMixture      7   \n",
       "132        ChemicalEntity     PhenotypicFeature      5   \n",
       "133        ChemicalEntity  PhysiologicalProcess      3   \n",
       "136        ChemicalEntity             Procedure      3   \n",
       "137        ChemicalEntity               Protein      7   \n",
       "139        ChemicalEntity         SmallMolecule      8   \n",
       "258               Disease               Disease     11   \n",
       "262               Disease                  Gene     16   \n",
       "266               Disease      MolecularMixture      5   \n",
       "269               Disease     PhenotypicFeature     11   \n",
       "270               Disease  PhysiologicalProcess      2   \n",
       "273               Disease             Procedure      4   \n",
       "274               Disease               Protein      3   \n",
       "277               Disease         SmallMolecule     14   \n",
       "344                  Gene                  Gene     15   \n",
       "349                  Gene      MolecularMixture      5   \n",
       "353                  Gene     PhenotypicFeature      8   \n",
       "354                  Gene  PhysiologicalProcess      3   \n",
       "356                  Gene             Procedure      1   \n",
       "357                  Gene               Protein      5   \n",
       "360                  Gene         SmallMolecule     14   \n",
       "434      MolecularMixture      MolecularMixture      6   \n",
       "436      MolecularMixture     PhenotypicFeature      4   \n",
       "437      MolecularMixture  PhysiologicalProcess      2   \n",
       "440      MolecularMixture             Procedure      2   \n",
       "441      MolecularMixture               Protein      2   \n",
       "443      MolecularMixture         SmallMolecule      7   \n",
       "492     PhenotypicFeature     PhenotypicFeature      9   \n",
       "493     PhenotypicFeature  PhysiologicalProcess      2   \n",
       "496     PhenotypicFeature             Procedure      4   \n",
       "497     PhenotypicFeature               Protein      2   \n",
       "500     PhenotypicFeature         SmallMolecule      9   \n",
       "522  PhysiologicalProcess  PhysiologicalProcess      2   \n",
       "525  PhysiologicalProcess             Procedure      2   \n",
       "526  PhysiologicalProcess               Protein      2   \n",
       "528  PhysiologicalProcess         SmallMolecule      3   \n",
       "590             Procedure             Procedure      3   \n",
       "591             Procedure               Protein      1   \n",
       "593             Procedure         SmallMolecule      5   \n",
       "620               Protein               Protein      5   \n",
       "622               Protein         SmallMolecule      7   \n",
       "680         SmallMolecule         SmallMolecule     12   \n",
       "\n",
       "                                                  list  \n",
       "56   [Automat-viral-proteome(Trapi v1.3.0), Automat...  \n",
       "57                    [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "62   [Automat-viral-proteome(Trapi v1.3.0), Automat...  \n",
       "65   [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "69                    [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "71   [Automat-viral-proteome(Trapi v1.3.0), Automat...  \n",
       "72                    [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "74                    [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "75   [Automat-viral-proteome(Trapi v1.3.0), Automat...  \n",
       "77                    [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "114  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "121  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "124  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "128  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "132  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "133  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "136  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "137  [Automat-biolink(Trapi v1.3.0), Automat-viral-...  \n",
       "139  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "258  [Automat-biolink(Trapi v1.3.0), Ontology Looku...  \n",
       "262  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "266  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "269  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "270  [Automat-icees-kg(Trapi v1.3.0), BioThings SEM...  \n",
       "273  [Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....  \n",
       "274  [Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...  \n",
       "277  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "344  [Automat-biolink(Trapi v1.3.0), Automat-hetio(...  \n",
       "349  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "353  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "354  [Automat-icees-kg(Trapi v1.3.0), Automat-pharo...  \n",
       "356                   [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "357  [Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...  \n",
       "360  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "434  [Automat-biolink(Trapi v1.3.0), Automat-hetio(...  \n",
       "436  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "437  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "440  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "441  [Automat-icees-kg(Trapi v1.3.0), Automat-hmdb(...  \n",
       "443  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "492  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "493  [Automat-icees-kg(Trapi v1.3.0), BioThings SEM...  \n",
       "496  [Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....  \n",
       "497  [Automat-icees-kg(Trapi v1.3.0), Automat-ctd(T...  \n",
       "500  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "522  [Automat-icees-kg(Trapi v1.3.0), BioThings SEM...  \n",
       "525  [Automat-icees-kg(Trapi v1.3.0), BioThings SEM...  \n",
       "526  [Automat-icees-kg(Trapi v1.3.0), BioThings SEM...  \n",
       "528  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "590  [Automat-icees-kg(Trapi v1.3.0), COHD TRAPI 1....  \n",
       "591                   [Automat-icees-kg(Trapi v1.3.0)]  \n",
       "593  [Automat-icees-kg(Trapi v1.3.0), Automat-hetio...  \n",
       "620  [Automat-viral-proteome(Trapi v1.3.0), Automat...  \n",
       "622  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  \n",
       "680  [Automat-biolink(Trapi v1.3.0), Automat-icees-...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats_filt = api_stats.query(\"subject in @keep & object in @keep & subject <= object\")\n",
    "api_stats_filt.to_csv(\"results/api_stats_filt.tsv\", sep=\"\\t\")\n",
    "api_stats_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0eda096-761d-486b-8bee-a00585a0774f",
   "metadata": {},
   "source": [
    "## Export to graphml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "84d2d692-7498-4e95-8c2f-7ad227941905",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_graph(df2, filename):\n",
    "    G = nx.Graph()\n",
    "\n",
    "    node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "    for node_type in node_types:\n",
    "        G.add_node(node_type, label = add_spacing(node_type))\n",
    "\n",
    "    for index,row in df2.iterrows():\n",
    "        G.add_edge(row['subject'], row['object'], weight=row['count'])\n",
    "    \n",
    "    nx.write_graphml(G, filename, infer_numeric_types=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed33943d-4a08-4927-92e9-70df8f2dc190",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_spacing(str):\n",
    "    key = {\n",
    "        \"BiologicalProcess\":               \"Biological\\nProcess\",\n",
    "        \"ChemicalEntity\":                  \"Chemical\\nEntity\",\n",
    "        \"MolecularMixture\":                \"Molecular\\nMixture\", \n",
    "        \"PhysiologicalProcess\":            \"Physiological\\nProcess\",\n",
    "        \"SmallMolecule\":                   \"Small\\nMolecule\",\n",
    "        \"PhenotypicFeature\":               \"Phenotypic\\nFeature\",\n",
    "        'ChemicalExposure':                'Chemical\\nExposure',\n",
    "        'ClinicalAttribute':               'Clinical\\nAttribute',\n",
    "        'ClinicalIntervention':            'Clinical\\nIntervention',\n",
    "        'ComplexMolecularMixture':         'Complex\\nMolecular\\nMixture',\n",
    "        'EnvironmentalExposure':           'Environmental\\nExposure',\n",
    "        'InformationContentEntity':        'Information\\nContentEntity',\n",
    "        'MolecularMixture':                'Molecular\\nMixture',\n",
    "        'PhysiologicalProcess':            'Physiological\\nProcess',\n",
    "        'PopulationOfIndividualOrganisms': 'PopulationOf\\nIndividualOrganisms'\n",
    "    }\n",
    "    if str in key.keys():\n",
    "        return(key[str])\n",
    "    else:\n",
    "        return(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a53dc509-2c6c-4c70-ba6d-be046363e844",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "create_graph(api_stats_filt, \"results/api_stats_filt.graphml\")\n",
    "create_graph(predicate_stats_filt, \"results/predicate_stats_filt.graphml\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
