{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819c0140-24d4-4a83-9cdd-9f1f21a18d53",
   "metadata": {},
   "source": [
    "# BTE metakg visualization\n",
    "\n",
    "The goal of this notebook is to visualize the metaKG that BTE uses.  Similar to the [subway diagram](https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/diagrams/smartapi_metagraph.png) we've used before, but updated to the size and scale of the current metakg.\n",
    "\n",
    "This notebook takes as input an ndson file with the SmartAPI metakg (originally provided by Chunlei on 2023-03-01).\n",
    "\n",
    "Optimizations\n",
    "* remove less-commonly-used node types from subject/object\n",
    "* only count in one direction (`A-treats-B` gets merged with `B-treated_by-A`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "472db681-b76f-48e7-b9f6-86eb863d3dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biothings_client\n",
    "import json5\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3b92f98-9a9e-4890-bf19-a58ccb51bc48",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5ccf0d5-f69e-4256-9c4b-251b4f51efc0",
   "metadata": {},
   "source": [
    "### Option 1 -- Read in the Smart API ndjson file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5af40473-2748-4a6c-b16f-c5db1a830b31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>MolecularEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Occurrent</td>\n",
       "      <td>affects_activity_of</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'metadata'...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175543</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>causes</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175544</th>\n",
       "      <td>Transcript</td>\n",
       "      <td>CellLine</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175545</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175546</th>\n",
       "      <td>Device</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175547</th>\n",
       "      <td>Procedure</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>has_participant</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175548 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 subject             object  \\\n",
       "0       AnatomicalEntity  MolecularActivity   \n",
       "1       AnatomicalEntity    MolecularEntity   \n",
       "2       AnatomicalEntity         NamedThing   \n",
       "3       AnatomicalEntity  NucleicAcidEntity   \n",
       "4       AnatomicalEntity          Occurrent   \n",
       "...                  ...                ...   \n",
       "175543              Cell         Phenomenon   \n",
       "175544        Transcript           CellLine   \n",
       "175545           Disease  PhenotypicFeature   \n",
       "175546            Device            Vitamin   \n",
       "175547         Procedure          Procedure   \n",
       "\n",
       "                                 predicate  \\\n",
       "0                      affects_activity_of   \n",
       "1                      affects_activity_of   \n",
       "2                      affects_activity_of   \n",
       "3                      affects_activity_of   \n",
       "4                      affects_activity_of   \n",
       "...                                    ...   \n",
       "175543                              causes   \n",
       "175544           physically_interacts_with   \n",
       "175545  entity_positively_regulates_entity   \n",
       "175546  entity_negatively_regulates_entity   \n",
       "175547                     has_participant   \n",
       "\n",
       "                                                      api provided_by  \n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'metadata'...         NaN  \n",
       "...                                                   ...         ...  \n",
       "175543  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175544  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175545  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175546  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "175547  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...         NaN  \n",
       "\n",
       "[175548 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_json('data/smartapi_metakg_03012023.ndjson.gz', lines=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fcb411e-fede-45f3-a2c5-4990eb924e32",
   "metadata": {},
   "source": [
    "### Option 2 (preferred) -- Query the SmartAPI API\n",
    "\n",
    "This method takes longer, but it retrieves the most up-to-date data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5644e1a0-fc78-473f-adf2-0184168862fb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "c = biothings_client.get_client('metakg', url='https://dev.smart-api.info/api/metakg')\n",
    "c._query_endpoint=''\n",
    "a = c.query('*', fetch_all=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b8279669-8d85-4c00-8c26-71cda255bb73",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching 304687 metakg(s) . . .\n",
      "No more results to return.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GeneOrGeneProduct</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GeneProductMixin</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>MacromolecularComplexMixin</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304682</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>causes</td>\n",
       "      <td>Cell</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304683</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>CellLine</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304684</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304685</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>Device</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304686</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>has_participant</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>304687 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      api  \\\n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "...                                                   ...   \n",
       "304682  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304683  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304684  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304685  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304686  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                            object                           predicate  \\\n",
       "0                             Gene             entity_regulates_entity   \n",
       "1                GeneOrGeneProduct             entity_regulates_entity   \n",
       "2                 GeneProductMixin             entity_regulates_entity   \n",
       "3                    GenomicEntity             entity_regulates_entity   \n",
       "4       MacromolecularComplexMixin             entity_regulates_entity   \n",
       "...                            ...                                 ...   \n",
       "304682                  Phenomenon                              causes   \n",
       "304683                    CellLine           physically_interacts_with   \n",
       "304684           PhenotypicFeature  entity_positively_regulates_entity   \n",
       "304685                     Vitamin  entity_negatively_regulates_entity   \n",
       "304686                   Procedure                     has_participant   \n",
       "\n",
       "                           subject provided_by  \n",
       "0       MacromolecularMachineMixin         NaN  \n",
       "1       MacromolecularMachineMixin         NaN  \n",
       "2       MacromolecularMachineMixin         NaN  \n",
       "3       MacromolecularMachineMixin         NaN  \n",
       "4       MacromolecularMachineMixin         NaN  \n",
       "...                            ...         ...  \n",
       "304682                        Cell         NaN  \n",
       "304683                  Transcript         NaN  \n",
       "304684                     Disease         NaN  \n",
       "304685                      Device         NaN  \n",
       "304686                   Procedure         NaN  \n",
       "\n",
       "[304687 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(a)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e981e0d-f064-4f5b-b7a3-50929a1e58b7",
   "metadata": {},
   "source": [
    "parse out a couple lines for the API name and ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "479dce85-28b0-45e2-b494-7756a1b42bed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>4803457bdb4bfeeb63a88244830ece2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GeneOrGeneProduct</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>4803457bdb4bfeeb63a88244830ece2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GeneProductMixin</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>4803457bdb4bfeeb63a88244830ece2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>4803457bdb4bfeeb63a88244830ece2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '480...</td>\n",
       "      <td>MacromolecularComplexMixin</td>\n",
       "      <td>entity_regulates_entity</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>4803457bdb4bfeeb63a88244830ece2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304682</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>causes</td>\n",
       "      <td>Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304683</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>CellLine</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304684</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304685</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>Device</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304686</th>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>has_participant</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>304687 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      api  \\\n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'id': '480...   \n",
       "...                                                   ...   \n",
       "304682  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304683  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304684  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304685  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "304686  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                            object                           predicate  \\\n",
       "0                             Gene             entity_regulates_entity   \n",
       "1                GeneOrGeneProduct             entity_regulates_entity   \n",
       "2                 GeneProductMixin             entity_regulates_entity   \n",
       "3                    GenomicEntity             entity_regulates_entity   \n",
       "4       MacromolecularComplexMixin             entity_regulates_entity   \n",
       "...                            ...                                 ...   \n",
       "304682                  Phenomenon                              causes   \n",
       "304683                    CellLine           physically_interacts_with   \n",
       "304684           PhenotypicFeature  entity_positively_regulates_entity   \n",
       "304685                     Vitamin  entity_negatively_regulates_entity   \n",
       "304686                   Procedure                     has_participant   \n",
       "\n",
       "                           subject provided_by  \\\n",
       "0       MacromolecularMachineMixin         NaN   \n",
       "1       MacromolecularMachineMixin         NaN   \n",
       "2       MacromolecularMachineMixin         NaN   \n",
       "3       MacromolecularMachineMixin         NaN   \n",
       "4       MacromolecularMachineMixin         NaN   \n",
       "...                            ...         ...   \n",
       "304682                        Cell         NaN   \n",
       "304683                  Transcript         NaN   \n",
       "304684                     Disease         NaN   \n",
       "304685                      Device         NaN   \n",
       "304686                   Procedure         NaN   \n",
       "\n",
       "                                      api_name  \\\n",
       "0                                   CAM-KP API   \n",
       "1                                   CAM-KP API   \n",
       "2                                   CAM-KP API   \n",
       "3                                   CAM-KP API   \n",
       "4                                   CAM-KP API   \n",
       "...                                        ...   \n",
       "304682  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "304683  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "304684  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "304685  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "304686  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "\n",
       "                                  api_id  \n",
       "0       4803457bdb4bfeeb63a88244830ece2e  \n",
       "1       4803457bdb4bfeeb63a88244830ece2e  \n",
       "2       4803457bdb4bfeeb63a88244830ece2e  \n",
       "3       4803457bdb4bfeeb63a88244830ece2e  \n",
       "4       4803457bdb4bfeeb63a88244830ece2e  \n",
       "...                                  ...  \n",
       "304682  e248aefca0f469229e82cca80fbabc89  \n",
       "304683  e248aefca0f469229e82cca80fbabc89  \n",
       "304684  e248aefca0f469229e82cca80fbabc89  \n",
       "304685  e248aefca0f469229e82cca80fbabc89  \n",
       "304686  e248aefca0f469229e82cca80fbabc89  \n",
       "\n",
       "[304687 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(api_name = lambda x: pd.json_normalize(x['api'])['name'])\n",
    "df = df.assign(api_id = lambda x: pd.json_normalize(x['api'])['smartapi.id'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aafe106-7f9a-4713-84fe-59ba83195f1b",
   "metadata": {},
   "source": [
    "Read in the BTE config file that specifies currently-allowed APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "837e2003-2ac6-429b-a123-2715f1b93da0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'include': [{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       "  {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       "  {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       "  {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       "  {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "   'name': 'Ontology Lookup Service API'},\n",
       "  {'id': '38e9e5169a72aee3659c9ddba956790d',\n",
       "   'name': 'BioThings BindingDB API'},\n",
       "  {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "   'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       "  {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "   'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       "  {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       "  {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       "  {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       "  {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "   'name': 'BioThings EBIgene2phenotype API'},\n",
       "  {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "   'name': 'BioThings GO Biological Process API'},\n",
       "  {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "   'name': 'BioThings GO Cellular Component API'},\n",
       "  {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "   'name': 'BioThings GO Molecular Function API'},\n",
       "  {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       "  {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       "  {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       "  {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "   'name': 'BioThings MGIgene2phenotype API'},\n",
       "  {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       "  {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       "  {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       "  {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       "  {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       "  {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       "  {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       "  {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       "  {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "   'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       "  {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "   'name': 'Multiomics ClinicalTrials KP'},\n",
       "  {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "   'name': 'Multiomics EHR Risk KP API'},\n",
       "  {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "   'name': 'Multiomics Wellness KP API'},\n",
       "  {'id': '978fe380a147a8641caf72320862697b',\n",
       "   'name': 'Text Mining Targeted Association API'},\n",
       "  {'id': '25085b05fd1afcebb497724d147cfb44',\n",
       "   'name': 'Automat-biolink(Trapi v1.3.0)'},\n",
       "  {'id': '89aa98098ba4329aa49b43ff8d21ffbb',\n",
       "   'name': 'Automat-ctd(Trapi v1.3.0)'},\n",
       "  {'id': '463525b319d7e1ef8f705bce953d01bd',\n",
       "   'name': 'Automat-cord19(Trapi v1.3.0)'},\n",
       "  {'id': '539873b1f2f2eb913efaee411e09eaa7',\n",
       "   'name': 'Automat-drug-central(Trapi v1.3.0)'},\n",
       "  {'id': '272c2357763e4faf737f1cba94beaa1d',\n",
       "   'name': 'Automat-gtex(Trapi v1.3.0)'},\n",
       "  {'id': 'fb1f3780ad67b030cb0617363afa8f61',\n",
       "   'name': 'Automat-gwas-catalog(Trapi v1.3.0)'},\n",
       "  {'id': 'aacdd9e863bb659574d302f084084976',\n",
       "   'name': 'Automat-gtopdb(Trapi v1.3.0)'},\n",
       "  {'id': '7382f0fabffce3cc7f7b8b6358c69259',\n",
       "   'name': 'Automat-hgnc(Trapi v1.3.0)'},\n",
       "  {'id': 'ee2d2eae42ca30fe82946d3f42febaa0',\n",
       "   'name': 'Automat-hmdb(Trapi v1.3.0)'},\n",
       "  {'id': '830600da121a5accc955cbf62e60f802',\n",
       "   'name': 'Automat-hetio(Trapi v1.3.0)'},\n",
       "  {'id': 'f30339a6894a146a19a34974712ca2e3',\n",
       "   'name': 'Automat-human-goa(Trapi v1.3.0)'},\n",
       "  {'id': '44e7a1147ca8657f50af6bb25982762d',\n",
       "   'name': 'Automat-icees-kg(Trapi v1.3.0)'},\n",
       "  {'id': 'e0687431a9ff88344d20e83e0c99ee7d',\n",
       "   'name': 'Automat-intact(Trapi v1.3.0)'},\n",
       "  {'id': '4c6f9117581531161849e60ea906f0be',\n",
       "   'name': 'Automat-ontology-hierarchy(Trapi v1.3.0)'},\n",
       "  {'id': '2a879882329b000c7e7f08c2d71ccffd',\n",
       "   'name': 'Automat-panther(Trapi v1.3.0)'},\n",
       "  {'id': '0295640b4d060133e4296dda4c31da47',\n",
       "   'name': 'Automat-pharos(Trapi v1.3.0)'},\n",
       "  {'id': 'ef9027a7d2246c6540cc7b3ce202d89f',\n",
       "   'name': 'Automat-uberongraph(Trapi v1.3.0)'},\n",
       "  {'id': '79310edd01cae96f3d8495250e625886',\n",
       "   'name': 'Automat-viral-proteome(Trapi v1.3.0)'},\n",
       "  {'id': '51c178099fa2dc99b5d8fff8bf9f1a0d', 'name': 'COHD TRAPI 1.3'},\n",
       "  {'id': '855adaa128ce5aa58a091d99e520d396',\n",
       "   'name': 'Connections Hypothesis Provider API'}]}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_url = \"https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/src/config/apis.js\"\n",
    "r = requests.get(bte_config_url)\n",
    "str_bte_config = r.text\n",
    "#print(str_bte_config)\n",
    "str_bte_config = re.sub(\"exports.API_LIST = \",  \"\", str_bte_config)                       # remove variable assignment step\n",
    "str_bte_config = re.sub(\"\\s*//.*\",              \"\", str_bte_config)                       # remove commented lines\n",
    "str_bte_config = re.sub(r'^$\\n',                '', str_bte_config, flags=re.MULTILINE)   # remove blank lines\n",
    "str_bte_config = re.sub(r',\\s*exclude:[^\\]]*]', '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "str_bte_config = re.sub(r';$',                  '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "#print(str_bte_config)\n",
    "\n",
    "bte_config = json5.loads(str_bte_config)\n",
    "bte_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "689ae4af-34a1-4a5c-97e3-9f97fce9ef7e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['d22b657426375a5295e7da8a303b9893',\n",
       " '43af91b3d7cae43591083bff9d75c6dd',\n",
       " 'dca415f2d792976af9d642b7e73f7a41',\n",
       " '1f277e1563fcfd124bfae2cc3c4bcdec',\n",
       " '1c056ffc7ed0dd1229e71c4752239465',\n",
       " '38e9e5169a72aee3659c9ddba956790d',\n",
       " '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       " 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       " '00fb85fc776279163199e6c50f6ddfc6',\n",
       " 'e3edd325c76f2992a111b43a907a4870',\n",
       " 'a7f784626a426d054885a5f33f17d3f8',\n",
       " '1f47552dabd67351d4c625adb0a10d00',\n",
       " 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       " 'f339b28426e7bf72028f60feefcd7465',\n",
       " '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       " '316eab811fd9ef1097df98bcaa9f7361',\n",
       " 'a5b0ec6bfde5008984d4b6cde402d61f',\n",
       " '32f36164fabed5d3abe6c2fd899c9418',\n",
       " '77ed27f111262d0289ed4f4071faa619',\n",
       " 'edeb26858bd27d0322af93e7a9e08761',\n",
       " '03283cc2b21c077be6794e1704b1d230',\n",
       " '1d288b3a3caf75d541ffaae3aab386c8',\n",
       " 'ec6d76016ef40f284359d17fbf78df20',\n",
       " '8f08d1446e0bb9c2b323713ce83e2bd3',\n",
       " '671b45c0301c8624abbd26ae78449ca2',\n",
       " '59dce17363dce279d389100834e43648',\n",
       " '09c8782d9f4027712e65b95424adba79',\n",
       " 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       " '08a5ddcde71b4bf838327ef469076acd',\n",
       " 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       " '02af7d098ab304e80d6f4806c3527027',\n",
       " '978fe380a147a8641caf72320862697b',\n",
       " '25085b05fd1afcebb497724d147cfb44',\n",
       " '89aa98098ba4329aa49b43ff8d21ffbb',\n",
       " '463525b319d7e1ef8f705bce953d01bd',\n",
       " '539873b1f2f2eb913efaee411e09eaa7',\n",
       " '272c2357763e4faf737f1cba94beaa1d',\n",
       " 'fb1f3780ad67b030cb0617363afa8f61',\n",
       " 'aacdd9e863bb659574d302f084084976',\n",
       " '7382f0fabffce3cc7f7b8b6358c69259',\n",
       " 'ee2d2eae42ca30fe82946d3f42febaa0',\n",
       " '830600da121a5accc955cbf62e60f802',\n",
       " 'f30339a6894a146a19a34974712ca2e3',\n",
       " '44e7a1147ca8657f50af6bb25982762d',\n",
       " 'e0687431a9ff88344d20e83e0c99ee7d',\n",
       " '4c6f9117581531161849e60ea906f0be',\n",
       " '2a879882329b000c7e7f08c2d71ccffd',\n",
       " '0295640b4d060133e4296dda4c31da47',\n",
       " 'ef9027a7d2246c6540cc7b3ce202d89f',\n",
       " '79310edd01cae96f3d8495250e625886',\n",
       " '51c178099fa2dc99b5d8fff8bf9f1a0d',\n",
       " '855adaa128ce5aa58a091d99e520d396']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_ids = [ x['id'] for x in bte_config['include'] ]\n",
    "bte_config_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96ca64b-7295-4039-a8c7-9fe1c9912473",
   "metadata": {},
   "source": [
    "## Join SmartAPI data with BTE config IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7bf98b05-b49a-4308-88ec-eed57eba83a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37865</th>\n",
       "      <td>Gene</td>\n",
       "      <td>has_part</td>\n",
       "      <td>GeneFamily</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-hgnc(Trapi v1.3.0)</td>\n",
       "      <td>7382f0fabffce3cc7f7b8b6358c69259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43365</th>\n",
       "      <td>Disease</td>\n",
       "      <td>superclass_of</td>\n",
       "      <td>Disease</td>\n",
       "      <td>infores:disease-ontology</td>\n",
       "      <td>Ontology Lookup Service API</td>\n",
       "      <td>1c056ffc7ed0dd1229e71c4752239465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51379</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-uberongraph(Trapi v1.3.0)</td>\n",
       "      <td>ef9027a7d2246c6540cc7b3ce202d89f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51380</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-uberongraph(Trapi v1.3.0)</td>\n",
       "      <td>ef9027a7d2246c6540cc7b3ce202d89f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51381</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>overlaps</td>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Automat-uberongraph(Trapi v1.3.0)</td>\n",
       "      <td>ef9027a7d2246c6540cc7b3ce202d89f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259395</th>\n",
       "      <td>Gene</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259396</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259397</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259400</th>\n",
       "      <td>Disease</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259401</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4115 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   object          predicate            subject  \\\n",
       "37865                Gene           has_part         GeneFamily   \n",
       "43365             Disease      superclass_of            Disease   \n",
       "51379      ChemicalEntity        subclass_of            Protein   \n",
       "51380      ChemicalEntity         related_to            Protein   \n",
       "51381      ChemicalEntity           overlaps            Protein   \n",
       "...                   ...                ...                ...   \n",
       "259395               Gene  contribution_from  PhenotypicFeature   \n",
       "259396  PhenotypicFeature     contributes_to      SmallMolecule   \n",
       "259397      SmallMolecule  contribution_from  PhenotypicFeature   \n",
       "259400            Disease     contributes_to      SmallMolecule   \n",
       "259401      SmallMolecule  contribution_from            Disease   \n",
       "\n",
       "                     provided_by                              api_name  \\\n",
       "37865                        NaN            Automat-hgnc(Trapi v1.3.0)   \n",
       "43365   infores:disease-ontology           Ontology Lookup Service API   \n",
       "51379                        NaN     Automat-uberongraph(Trapi v1.3.0)   \n",
       "51380                        NaN     Automat-uberongraph(Trapi v1.3.0)   \n",
       "51381                        NaN     Automat-uberongraph(Trapi v1.3.0)   \n",
       "...                          ...                                   ...   \n",
       "259395                       NaN  Text Mining Targeted Association API   \n",
       "259396                       NaN  Text Mining Targeted Association API   \n",
       "259397                       NaN  Text Mining Targeted Association API   \n",
       "259400                       NaN  Text Mining Targeted Association API   \n",
       "259401                       NaN  Text Mining Targeted Association API   \n",
       "\n",
       "                                  api_id  \n",
       "37865   7382f0fabffce3cc7f7b8b6358c69259  \n",
       "43365   1c056ffc7ed0dd1229e71c4752239465  \n",
       "51379   ef9027a7d2246c6540cc7b3ce202d89f  \n",
       "51380   ef9027a7d2246c6540cc7b3ce202d89f  \n",
       "51381   ef9027a7d2246c6540cc7b3ce202d89f  \n",
       "...                                  ...  \n",
       "259395  978fe380a147a8641caf72320862697b  \n",
       "259396  978fe380a147a8641caf72320862697b  \n",
       "259397  978fe380a147a8641caf72320862697b  \n",
       "259400  978fe380a147a8641caf72320862697b  \n",
       "259401  978fe380a147a8641caf72320862697b  \n",
       "\n",
       "[4115 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte = df.query('api_id in @bte_config_ids').drop(columns=['api']).drop_duplicates()\n",
    "df_bte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6764f282-a1e6-4b3c-936f-11b4d3971792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_bte[['subject','object','predicate','api_name','api_id']].drop_duplicates().to_csv(\"results/bte_operations.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac384894-b9af-47c0-b0b5-e4920b30cf17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Automat-uberongraph(Trapi v1.3.0)           1368\n",
       "BioThings SEMMEDDB API                       932\n",
       "Automat-ctd(Trapi v1.3.0)                    509\n",
       "Automat-biolink(Trapi v1.3.0)                210\n",
       "Automat-hetio(Trapi v1.3.0)                  143\n",
       "Automat-ontology-hierarchy(Trapi v1.3.0)     134\n",
       "Automat-drug-central(Trapi v1.3.0)            93\n",
       "Automat-pharos(Trapi v1.3.0)                  79\n",
       "Automat-hmdb(Trapi v1.3.0)                    66\n",
       "Automat-human-goa(Trapi v1.3.0)               58\n",
       "Automat-icees-kg(Trapi v1.3.0)                55\n",
       "COHD TRAPI 1.3                                50\n",
       "Multiomics EHR Risk KP API                    44\n",
       "Automat-viral-proteome(Trapi v1.3.0)          35\n",
       "Automat-gtopdb(Trapi v1.3.0)                  33\n",
       "Automat-panther(Trapi v1.3.0)                 24\n",
       "BioLink API                                   21\n",
       "MyChem.info API                               21\n",
       "Text Mining Targeted Association API          20\n",
       "Automat-gwas-catalog(Trapi v1.3.0)            18\n",
       "Automat-gtex(Trapi v1.3.0)                    17\n",
       "Multiomics Wellness KP API                    17\n",
       "MyDisease.info API                            17\n",
       "MyVariant.info API                            16\n",
       "Multiomics BigGIM-DrugResponse KP API         14\n",
       "MyGene.info API                               13\n",
       "BioThings DGIdb API                           12\n",
       "Connections Hypothesis Provider API           12\n",
       "Automat-intact(Trapi v1.3.0)                  11\n",
       "BioThings GO Biological Process API           10\n",
       "BioThings UBERON API                           9\n",
       "BioThings IDISK API                            7\n",
       "BioThings PFOCR API                            6\n",
       "Automat-hgnc(Trapi v1.3.0)                     4\n",
       "BioThings EBIgene2phenotype API                4\n",
       "BioThings GO Molecular Function API            4\n",
       "BioThings MGIgene2phenotype API                4\n",
       "BioThings GO Cellular Component API            4\n",
       "BioThings GTRx API                             2\n",
       "Multiomics ClinicalTrials KP                   2\n",
       "BioThings BioPlanet Pathway-Disease API        2\n",
       "BioThings HPO API                              2\n",
       "BioThings DISEASES API                         2\n",
       "BioThings BioPlanet Pathway-Gene API           2\n",
       "BioThings BindingDB API                        2\n",
       "BioThings Rhea API                             2\n",
       "BioThings DDInter API                          1\n",
       "LitVar API                                     1\n",
       "QuickGO API                                    1\n",
       "EBI Proteins API                               1\n",
       "Ontology Lookup Service API                    1\n",
       "Name: api_name, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte['api_name'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398517c7-4afd-4e3c-b9e4-993dce33e07d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Summarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515b78f8-2606-4c88-9403-60e0fc2e34c6",
   "metadata": {},
   "source": [
    "### by subject, object; count # of APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a53da5f5-34dd-4f43-bdad-855bc4139fb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              subject                    object  count  \\\n",
      "0    AnatomicalEntity          AnatomicalEntity      4   \n",
      "1    AnatomicalEntity         BiologicalProcess      1   \n",
      "2    AnatomicalEntity                      Cell      3   \n",
      "3    AnatomicalEntity         CellularComponent      6   \n",
      "4    AnatomicalEntity            ChemicalEntity      1   \n",
      "5    AnatomicalEntity                   Disease      7   \n",
      "6    AnatomicalEntity                      Gene      2   \n",
      "7    AnatomicalEntity  GrossAnatomicalStructure      3   \n",
      "8    AnatomicalEntity         MolecularActivity      1   \n",
      "9    AnatomicalEntity          MolecularMixture      1   \n",
      "..                ...                       ...    ...   \n",
      "384     SmallMolecule          MolecularMixture     10   \n",
      "385     SmallMolecule       PathologicalProcess      1   \n",
      "386     SmallMolecule                   Pathway      2   \n",
      "387     SmallMolecule         PhenotypicFeature     11   \n",
      "388     SmallMolecule      PhysiologicalProcess      1   \n",
      "389     SmallMolecule               Polypeptide     10   \n",
      "390     SmallMolecule                 Procedure      3   \n",
      "391     SmallMolecule                   Protein      5   \n",
      "392     SmallMolecule             SmallMolecule     15   \n",
      "393         Treatment                   Disease      1   \n",
      "\n",
      "                                                  list  \n",
      "0    [Automat-uberongraph(Trapi v1.3.0), Automat-he...  \n",
      "1                  [Automat-uberongraph(Trapi v1.3.0)]  \n",
      "2    [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
      "3    [Automat-uberongraph(Trapi v1.3.0), Automat-pa...  \n",
      "4                  [Automat-uberongraph(Trapi v1.3.0)]  \n",
      "5    [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "6    [Automat-pharos(Trapi v1.3.0), Automat-hetio(T...  \n",
      "7    [Automat-uberongraph(Trapi v1.3.0), Automat-he...  \n",
      "8                  [Automat-uberongraph(Trapi v1.3.0)]  \n",
      "9                  [Automat-uberongraph(Trapi v1.3.0)]  \n",
      "..                                                 ...  \n",
      "384  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "385                           [BioThings SEMMEDDB API]  \n",
      "386  [Automat-uberongraph(Trapi v1.3.0), Automat-hm...  \n",
      "387  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "388                           [BioThings SEMMEDDB API]  \n",
      "389  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "390  [COHD TRAPI 1.3, Multiomics EHR Risk KP API, B...  \n",
      "391  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "392  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
      "393                     [Multiomics ClinicalTrials KP]  \n",
      "\n",
      "[394 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"api_name\"]]\n",
    "api_stats = df1.groupby(['subject','object'], group_keys=False)['api_name'].nunique().rename(\"count\").to_frame()\n",
    "api_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['api_name'].unique().apply(list)\n",
    "api_stats = api_stats.reset_index()\n",
    "\n",
    "with pd.option_context('display.min_rows', 20,\n",
    "                       'display.max_columns', None,\n",
    "                       'display.precision', 3,\n",
    "                       ):\n",
    "    print(api_stats)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d73842dd-d6a2-4309-b820-1594a85c44d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "api_stats.to_csv(\"results/api_stats.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abbb70cd-070a-4380-b9e2-63cb798d09f8",
   "metadata": {},
   "source": [
    "### by subject, object; count # of predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cc40d6f8-87ed-48b1-b8a1-d3c04a9d8893",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              subject                    object  count  \\\n",
      "0    AnatomicalEntity          AnatomicalEntity     17   \n",
      "1    AnatomicalEntity         BiologicalProcess     16   \n",
      "2    AnatomicalEntity                      Cell     15   \n",
      "3    AnatomicalEntity         CellularComponent      7   \n",
      "4    AnatomicalEntity            ChemicalEntity      4   \n",
      "5    AnatomicalEntity                   Disease     13   \n",
      "6    AnatomicalEntity                      Gene      4   \n",
      "7    AnatomicalEntity  GrossAnatomicalStructure     17   \n",
      "8    AnatomicalEntity         MolecularActivity      4   \n",
      "9    AnatomicalEntity          MolecularMixture      3   \n",
      "..                ...                       ...    ...   \n",
      "384     SmallMolecule          MolecularMixture      8   \n",
      "385     SmallMolecule       PathologicalProcess     10   \n",
      "386     SmallMolecule                   Pathway      2   \n",
      "387     SmallMolecule         PhenotypicFeature     20   \n",
      "388     SmallMolecule      PhysiologicalProcess      8   \n",
      "389     SmallMolecule               Polypeptide     16   \n",
      "390     SmallMolecule                 Procedure      6   \n",
      "391     SmallMolecule                   Protein     12   \n",
      "392     SmallMolecule             SmallMolecule     23   \n",
      "393         Treatment                   Disease      1   \n",
      "\n",
      "                                                  list  \n",
      "0    [subclass_of, superclass_of, has_part, part_of...  \n",
      "1    [related_to, regulated_by, is_output_of, parti...  \n",
      "2    [has_part, related_to, overlaps, coexists_with...  \n",
      "3    [superclass_of, subclass_of, has_part, related...  \n",
      "4           [related_to, overlaps, has_part, produces]  \n",
      "5    [related_to, affected_by, location_of, caused_...  \n",
      "6    [genetic_association, entity_positively_regula...  \n",
      "7    [overlaps, related_to, precedes, has_part, loc...  \n",
      "8    [contains_process, is_input_of, participates_i...  \n",
      "9                     [related_to, overlaps, has_part]  \n",
      "..                                                 ...  \n",
      "384  [subclass_of, related_to, overlaps, has_part, ...  \n",
      "385  [treats, causes, associated_with, prevents, af...  \n",
      "386                     [participates_in, is_input_of]  \n",
      "387  [affected_by, related_to, treats, causes_adver...  \n",
      "388  [causes, disrupts, affects, entity_positively_...  \n",
      "389  [subclass_of, superclass_of, related_to, overl...  \n",
      "390  [positively_correlated_with, negatively_correl...  \n",
      "391  [related_to, overlaps, part_of, subclass_of, i...  \n",
      "392  [subclass_of, superclass_of, related_to, overl...  \n",
      "393                                       [related_to]  \n",
      "\n",
      "[394 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"predicate\"]]\n",
    "predicate_stats = df1.groupby(['subject','object'], group_keys=False)['predicate'].nunique().rename(\"count\").to_frame()\n",
    "predicate_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['predicate'].unique().apply(list)\n",
    "predicate_stats = predicate_stats.reset_index()\n",
    "\n",
    "with pd.option_context('display.min_rows', 20,\n",
    "                       'display.max_columns', None,\n",
    "                       'display.precision', 3,\n",
    "                       ):\n",
    "    print(predicate_stats)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0a2da419-97e7-438d-a294-69fa0946081f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "predicate_stats.to_csv(\"results/predicate_stats.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e955923-8436-4453-b380-2f8b8bb9ce3f",
   "metadata": {},
   "source": [
    "## Filter by most common types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9104cf-516b-4457-9632-0ff9a42ff036",
   "metadata": {},
   "source": [
    "Filter to only include the most common types of entities.  Also, since we _mostly_ have the same info in both directions, only keep one direction to simplify visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b2b9bffb-2178-4266-9b12-845b75412e6e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Disease                     955\n",
       "SmallMolecule               824\n",
       "Gene                        805\n",
       "ChemicalEntity              629\n",
       "PhenotypicFeature           490\n",
       "BiologicalProcess           463\n",
       "Polypeptide                 463\n",
       "MolecularMixture            391\n",
       "Protein                     339\n",
       "Cell                        339\n",
       "CellularComponent           333\n",
       "AnatomicalEntity            331\n",
       "GrossAnatomicalStructure    316\n",
       "MolecularActivity           291\n",
       "Pathway                     203\n",
       "OrganismTaxon               201\n",
       "PathologicalProcess         182\n",
       "ChemicalMixture             164\n",
       "PhysiologicalProcess        138\n",
       "ComplexMolecularMixture     125\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ad7563bf-3c03-43ac-811c-186145df2acb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BiologicalProcess',\n",
       " 'Cell',\n",
       " 'ChemicalEntity',\n",
       " 'Disease',\n",
       " 'Gene',\n",
       " 'MolecularMixture',\n",
       " 'PhenotypicFeature',\n",
       " 'Polypeptide',\n",
       " 'Protein',\n",
       " 'SmallMolecule'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NUM_TYPES_TO_KEEP = 10\n",
    "\n",
    "keep = set(pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4b29e182-3e47-4e8f-9925-3eb0803566a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>23</td>\n",
       "      <td>[subclass_of, superclass_of, causes, caused_by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Cell</td>\n",
       "      <td>13</td>\n",
       "      <td>[related_to, has_participant, regulates, occur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>12</td>\n",
       "      <td>[has_input, has_participant, has_output, affec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>16</td>\n",
       "      <td>[causes, has_participant, subclass_of, related...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Gene</td>\n",
       "      <td>2</td>\n",
       "      <td>[has_part, has_participant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>7</td>\n",
       "      <td>[affects_transport_of, related_to, has_partici...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[affected_by, related_to, superclass_of, has_r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>8</td>\n",
       "      <td>[has_participant, related_to, affects_transpor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>20</td>\n",
       "      <td>[is_output_of, caused_by, related_to, has_capa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[has_input, has_participant, has_output, affec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Cell</td>\n",
       "      <td>11</td>\n",
       "      <td>[subclass_of, superclass_of, related_to, devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Cell</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>12</td>\n",
       "      <td>[expresses, related_to, overlaps, has_part, ha...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Disease</td>\n",
       "      <td>12</td>\n",
       "      <td>[related_to, affected_by, location_of, caused_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Gene</td>\n",
       "      <td>8</td>\n",
       "      <td>[genetic_association, interacts_with, has_part...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Cell</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>2</td>\n",
       "      <td>[has_biomarker, contribution_from]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Cell</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>2</td>\n",
       "      <td>[affected_by, related_to]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>9</td>\n",
       "      <td>[related_to, overlaps, has_part, location_of, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Protein</td>\n",
       "      <td>4</td>\n",
       "      <td>[expresses, related_to, overlaps, has_part]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Cell</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>14</td>\n",
       "      <td>[related_to, overlaps, has_part, has_biomarker...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>15</td>\n",
       "      <td>[subclass_of, superclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>22</td>\n",
       "      <td>[causes, caused_by, affected_by, related_to, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>38</td>\n",
       "      <td>[correlated_with, directly_interacts_with, act...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>8</td>\n",
       "      <td>[subclass_of, related_to, overlaps, has_part, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>16</td>\n",
       "      <td>[related_to, affected_by, superclass_of, cause...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>14</td>\n",
       "      <td>[superclass_of, subclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Protein</td>\n",
       "      <td>9</td>\n",
       "      <td>[superclass_of, related_to, overlaps, part_of,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[superclass_of, related_to, overlaps, part_of,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>43</td>\n",
       "      <td>[superclass_of, subclass_of, has_phenotype, ph...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>24</td>\n",
       "      <td>[related_to, causes, disrupts, has_participant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Disease</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>13</td>\n",
       "      <td>[affects, caused_by, causes, treated_by, has_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>25</td>\n",
       "      <td>[related_to, has_phenotype, subclass_of, corre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>15</td>\n",
       "      <td>[affects, treated_by, has_biomarker, contribut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Protein</td>\n",
       "      <td>6</td>\n",
       "      <td>[caused_by, affects, causes, related_to, produ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>27</td>\n",
       "      <td>[caused_by, affects, causes, has_participant, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>31</td>\n",
       "      <td>[subclass_of, superclass_of, interacts_with, o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Gene</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>32</td>\n",
       "      <td>[activity_decreased_by, activity_increased_by,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>11</td>\n",
       "      <td>[has_phenotype, genetic_association, related_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>21</td>\n",
       "      <td>[response_decreased_by, increases_uptake_of, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein</td>\n",
       "      <td>4</td>\n",
       "      <td>[entity_negatively_regulated_by_entity, intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>57</td>\n",
       "      <td>[correlated_with, activity_decreased_by, activ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>8</td>\n",
       "      <td>[subclass_of, superclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>14</td>\n",
       "      <td>[affected_by, related_to, superclass_of, treat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>6</td>\n",
       "      <td>[superclass_of, related_to, overlaps, has_part...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[subclass_of, interacts_with, increases_transp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>8</td>\n",
       "      <td>[superclass_of, related_to, overlaps, part_of,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>21</td>\n",
       "      <td>[subclass_of, superclass_of, has_part, part_of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>13</td>\n",
       "      <td>[treated_by, has_biomarker, related_to, affect...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[related_to, affects]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>20</td>\n",
       "      <td>[treated_by, has_contraindication, has_biomark...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>16</td>\n",
       "      <td>[subclass_of, superclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Protein</td>\n",
       "      <td>8</td>\n",
       "      <td>[related_to, overlaps, part_of, subclass_of, i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>16</td>\n",
       "      <td>[superclass_of, subclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>Protein</td>\n",
       "      <td>Protein</td>\n",
       "      <td>10</td>\n",
       "      <td>[subclass_of, superclass_of, derives_from, der...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>Protein</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>14</td>\n",
       "      <td>[related_to, overlaps, has_part, superclass_of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>23</td>\n",
       "      <td>[subclass_of, superclass_of, related_to, overl...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject             object  count  \\\n",
       "17   BiologicalProcess  BiologicalProcess     23   \n",
       "18   BiologicalProcess               Cell     13   \n",
       "20   BiologicalProcess     ChemicalEntity     12   \n",
       "23   BiologicalProcess            Disease     16   \n",
       "24   BiologicalProcess               Gene      2   \n",
       "27   BiologicalProcess   MolecularMixture      7   \n",
       "30   BiologicalProcess  PhenotypicFeature      4   \n",
       "31   BiologicalProcess        Polypeptide      8   \n",
       "32   BiologicalProcess            Protein     20   \n",
       "33   BiologicalProcess      SmallMolecule      9   \n",
       "36                Cell               Cell     11   \n",
       "38                Cell     ChemicalEntity     12   \n",
       "39                Cell            Disease     12   \n",
       "40                Cell               Gene      8   \n",
       "43                Cell   MolecularMixture      2   \n",
       "47                Cell  PhenotypicFeature      2   \n",
       "49                Cell        Polypeptide      9   \n",
       "51                Cell            Protein      4   \n",
       "52                Cell      SmallMolecule     14   \n",
       "78      ChemicalEntity     ChemicalEntity     15   \n",
       "81      ChemicalEntity            Disease     22   \n",
       "82      ChemicalEntity               Gene     38   \n",
       "85      ChemicalEntity   MolecularMixture      8   \n",
       "88      ChemicalEntity  PhenotypicFeature     16   \n",
       "90      ChemicalEntity        Polypeptide     14   \n",
       "92      ChemicalEntity            Protein      9   \n",
       "93      ChemicalEntity      SmallMolecule     15   \n",
       "125            Disease            Disease     43   \n",
       "128            Disease               Gene     24   \n",
       "131            Disease   MolecularMixture     13   \n",
       "135            Disease  PhenotypicFeature     25   \n",
       "137            Disease        Polypeptide     15   \n",
       "139            Disease            Protein      6   \n",
       "141            Disease      SmallMolecule     27   \n",
       "160               Gene               Gene     31   \n",
       "164               Gene   MolecularMixture     32   \n",
       "168               Gene  PhenotypicFeature     11   \n",
       "170               Gene        Polypeptide     21   \n",
       "171               Gene            Protein      4   \n",
       "173               Gene      SmallMolecule     57   \n",
       "228   MolecularMixture   MolecularMixture      8   \n",
       "230   MolecularMixture  PhenotypicFeature     14   \n",
       "231   MolecularMixture        Polypeptide      6   \n",
       "232   MolecularMixture            Protein      3   \n",
       "233   MolecularMixture      SmallMolecule      8   \n",
       "294  PhenotypicFeature  PhenotypicFeature     21   \n",
       "296  PhenotypicFeature        Polypeptide     13   \n",
       "298  PhenotypicFeature            Protein      2   \n",
       "300  PhenotypicFeature      SmallMolecule     20   \n",
       "331        Polypeptide        Polypeptide     16   \n",
       "333        Polypeptide            Protein      8   \n",
       "334        Polypeptide      SmallMolecule     16   \n",
       "366            Protein            Protein     10   \n",
       "367            Protein      SmallMolecule     14   \n",
       "392      SmallMolecule      SmallMolecule     23   \n",
       "\n",
       "                                                  list  \n",
       "17   [subclass_of, superclass_of, causes, caused_by...  \n",
       "18   [related_to, has_participant, regulates, occur...  \n",
       "20   [has_input, has_participant, has_output, affec...  \n",
       "23   [causes, has_participant, subclass_of, related...  \n",
       "24                         [has_part, has_participant]  \n",
       "27   [affects_transport_of, related_to, has_partici...  \n",
       "30   [affected_by, related_to, superclass_of, has_r...  \n",
       "31   [has_participant, related_to, affects_transpor...  \n",
       "32   [is_output_of, caused_by, related_to, has_capa...  \n",
       "33   [has_input, has_participant, has_output, affec...  \n",
       "36   [subclass_of, superclass_of, related_to, devel...  \n",
       "38   [expresses, related_to, overlaps, has_part, ha...  \n",
       "39   [related_to, affected_by, location_of, caused_...  \n",
       "40   [genetic_association, interacts_with, has_part...  \n",
       "43                  [has_biomarker, contribution_from]  \n",
       "47                           [affected_by, related_to]  \n",
       "49   [related_to, overlaps, has_part, location_of, ...  \n",
       "51         [expresses, related_to, overlaps, has_part]  \n",
       "52   [related_to, overlaps, has_part, has_biomarker...  \n",
       "78   [subclass_of, superclass_of, related_to, overl...  \n",
       "81   [causes, caused_by, affected_by, related_to, p...  \n",
       "82   [correlated_with, directly_interacts_with, act...  \n",
       "85   [subclass_of, related_to, overlaps, has_part, ...  \n",
       "88   [related_to, affected_by, superclass_of, cause...  \n",
       "90   [superclass_of, subclass_of, related_to, overl...  \n",
       "92   [superclass_of, related_to, overlaps, part_of,...  \n",
       "93   [superclass_of, related_to, overlaps, part_of,...  \n",
       "125  [superclass_of, subclass_of, has_phenotype, ph...  \n",
       "128  [related_to, causes, disrupts, has_participant...  \n",
       "131  [affects, caused_by, causes, treated_by, has_c...  \n",
       "135  [related_to, has_phenotype, subclass_of, corre...  \n",
       "137  [affects, treated_by, has_biomarker, contribut...  \n",
       "139  [caused_by, affects, causes, related_to, produ...  \n",
       "141  [caused_by, affects, causes, has_participant, ...  \n",
       "160  [subclass_of, superclass_of, interacts_with, o...  \n",
       "164  [activity_decreased_by, activity_increased_by,...  \n",
       "168  [has_phenotype, genetic_association, related_t...  \n",
       "170  [response_decreased_by, increases_uptake_of, a...  \n",
       "171  [entity_negatively_regulated_by_entity, intera...  \n",
       "173  [correlated_with, activity_decreased_by, activ...  \n",
       "228  [subclass_of, superclass_of, related_to, overl...  \n",
       "230  [affected_by, related_to, superclass_of, treat...  \n",
       "231  [superclass_of, related_to, overlaps, has_part...  \n",
       "232  [subclass_of, interacts_with, increases_transp...  \n",
       "233  [superclass_of, related_to, overlaps, part_of,...  \n",
       "294  [subclass_of, superclass_of, has_part, part_of...  \n",
       "296  [treated_by, has_biomarker, related_to, affect...  \n",
       "298                              [related_to, affects]  \n",
       "300  [treated_by, has_contraindication, has_biomark...  \n",
       "331  [subclass_of, superclass_of, related_to, overl...  \n",
       "333  [related_to, overlaps, part_of, subclass_of, i...  \n",
       "334  [superclass_of, subclass_of, related_to, overl...  \n",
       "366  [subclass_of, superclass_of, derives_from, der...  \n",
       "367  [related_to, overlaps, has_part, superclass_of...  \n",
       "392  [subclass_of, superclass_of, related_to, overl...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicate_stats_filt = predicate_stats.query(\"subject in @keep & object in @keep & subject <= object\")\n",
    "predicate_stats_filt.to_csv(\"results/predicate_stats_filt.tsv\", sep=\"\\t\")\n",
    "predicate_stats_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dabb9bd0-808b-4e46-a51c-ff15d25ac4dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Cell</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-ic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Gene</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-hetio(Trapi v1.3.0), MyGene.info API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-ic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-hu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-ic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Cell</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Cell</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Disease</td>\n",
       "      <td>6</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Gene</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-pharos(Trapi v1.3.0), BioThings SEMME...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Cell</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-ctd(Tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Cell</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), BioThings ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Cell</td>\n",
       "      <td>Protein</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Cell</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-hmdb(Trapi v1.3.0), Automat-pharos(Tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>8</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>6</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Protein</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>11</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>14</td>\n",
       "      <td>[Ontology Lookup Service API, Automat-uberongr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>16</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Disease</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>14</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Protein</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), BioThings ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>16</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>18</td>\n",
       "      <td>[Automat-hgnc(Trapi v1.3.0), Automat-hmdb(Trap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Gene</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-gtopdb(Trapi v1.3.0), Automat-pharos(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-pharos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>2</td>\n",
       "      <td>[Automat-ctd(Trapi v1.3.0), BioThings SEMMEDDB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>12</td>\n",
       "      <td>[Automat-hmdb(Trapi v1.3.0), Automat-gtopdb(Tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>7</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>Protein</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>MolecularMixture</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>12</td>\n",
       "      <td>[Automat-drug-central(Trapi v1.3.0), Automat-b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>4</td>\n",
       "      <td>[Automat-biolink(Trapi v1.3.0), Automat-uberon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Protein</td>\n",
       "      <td>1</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>11</td>\n",
       "      <td>[Automat-drug-central(Trapi v1.3.0), Automat-b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Protein</td>\n",
       "      <td>3</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-on...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>10</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>Protein</td>\n",
       "      <td>Protein</td>\n",
       "      <td>9</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>Protein</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>5</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-hm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[Automat-uberongraph(Trapi v1.3.0), Automat-dr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject             object  count  \\\n",
       "17   BiologicalProcess  BiologicalProcess      8   \n",
       "18   BiologicalProcess               Cell      1   \n",
       "20   BiologicalProcess     ChemicalEntity      2   \n",
       "23   BiologicalProcess            Disease     10   \n",
       "24   BiologicalProcess               Gene      2   \n",
       "27   BiologicalProcess   MolecularMixture      2   \n",
       "30   BiologicalProcess  PhenotypicFeature      7   \n",
       "31   BiologicalProcess        Polypeptide      1   \n",
       "32   BiologicalProcess            Protein      3   \n",
       "33   BiologicalProcess      SmallMolecule      2   \n",
       "36                Cell               Cell      4   \n",
       "38                Cell     ChemicalEntity      4   \n",
       "39                Cell            Disease      6   \n",
       "40                Cell               Gene      2   \n",
       "43                Cell   MolecularMixture      2   \n",
       "47                Cell  PhenotypicFeature      1   \n",
       "49                Cell        Polypeptide      2   \n",
       "51                Cell            Protein      1   \n",
       "52                Cell      SmallMolecule      4   \n",
       "78      ChemicalEntity     ChemicalEntity     10   \n",
       "81      ChemicalEntity            Disease      8   \n",
       "82      ChemicalEntity               Gene      5   \n",
       "85      ChemicalEntity   MolecularMixture     10   \n",
       "88      ChemicalEntity  PhenotypicFeature      8   \n",
       "90      ChemicalEntity        Polypeptide      6   \n",
       "92      ChemicalEntity            Protein     10   \n",
       "93      ChemicalEntity      SmallMolecule     11   \n",
       "125            Disease            Disease     14   \n",
       "128            Disease               Gene     16   \n",
       "131            Disease   MolecularMixture      7   \n",
       "135            Disease  PhenotypicFeature     14   \n",
       "137            Disease        Polypeptide      4   \n",
       "139            Disease            Protein      2   \n",
       "141            Disease      SmallMolecule     16   \n",
       "160               Gene               Gene     18   \n",
       "164               Gene   MolecularMixture      3   \n",
       "168               Gene  PhenotypicFeature      7   \n",
       "170               Gene        Polypeptide      2   \n",
       "171               Gene            Protein      1   \n",
       "173               Gene      SmallMolecule     12   \n",
       "228   MolecularMixture   MolecularMixture     10   \n",
       "230   MolecularMixture  PhenotypicFeature      7   \n",
       "231   MolecularMixture        Polypeptide      7   \n",
       "232   MolecularMixture            Protein      4   \n",
       "233   MolecularMixture      SmallMolecule     10   \n",
       "294  PhenotypicFeature  PhenotypicFeature     12   \n",
       "296  PhenotypicFeature        Polypeptide      4   \n",
       "298  PhenotypicFeature            Protein      1   \n",
       "300  PhenotypicFeature      SmallMolecule     11   \n",
       "331        Polypeptide        Polypeptide      5   \n",
       "333        Polypeptide            Protein      3   \n",
       "334        Polypeptide      SmallMolecule     10   \n",
       "366            Protein            Protein      9   \n",
       "367            Protein      SmallMolecule      5   \n",
       "392      SmallMolecule      SmallMolecule     15   \n",
       "\n",
       "                                                  list  \n",
       "17   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "18                 [Automat-uberongraph(Trapi v1.3.0)]  \n",
       "20   [Automat-uberongraph(Trapi v1.3.0), Automat-ic...  \n",
       "23   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "24      [Automat-hetio(Trapi v1.3.0), MyGene.info API]  \n",
       "27   [Automat-uberongraph(Trapi v1.3.0), Automat-ic...  \n",
       "30   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "31                 [Automat-uberongraph(Trapi v1.3.0)]  \n",
       "32   [Automat-uberongraph(Trapi v1.3.0), Automat-hu...  \n",
       "33   [Automat-uberongraph(Trapi v1.3.0), Automat-ic...  \n",
       "36   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "38   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "39   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "40   [Automat-pharos(Trapi v1.3.0), BioThings SEMME...  \n",
       "43   [Automat-biolink(Trapi v1.3.0), Automat-ctd(Tr...  \n",
       "47                 [Automat-uberongraph(Trapi v1.3.0)]  \n",
       "49   [Automat-uberongraph(Trapi v1.3.0), BioThings ...  \n",
       "51                 [Automat-uberongraph(Trapi v1.3.0)]  \n",
       "52   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "78   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "81   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "82   [Automat-hmdb(Trapi v1.3.0), Automat-pharos(Tr...  \n",
       "85   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "88   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "90   [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "92   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "93   [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "125  [Ontology Lookup Service API, Automat-uberongr...  \n",
       "128  [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "131  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "135  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "137  [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "139  [Automat-uberongraph(Trapi v1.3.0), BioThings ...  \n",
       "141  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "160  [Automat-hgnc(Trapi v1.3.0), Automat-hmdb(Trap...  \n",
       "164  [Automat-gtopdb(Trapi v1.3.0), Automat-pharos(...  \n",
       "168  [Automat-biolink(Trapi v1.3.0), Automat-pharos...  \n",
       "170  [Automat-ctd(Trapi v1.3.0), BioThings SEMMEDDB...  \n",
       "171                           [BioThings SEMMEDDB API]  \n",
       "173  [Automat-hmdb(Trapi v1.3.0), Automat-gtopdb(Tr...  \n",
       "228  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "230  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "231  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "232  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "233  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "294  [Automat-drug-central(Trapi v1.3.0), Automat-b...  \n",
       "296  [Automat-biolink(Trapi v1.3.0), Automat-uberon...  \n",
       "298                [Automat-uberongraph(Trapi v1.3.0)]  \n",
       "300  [Automat-drug-central(Trapi v1.3.0), Automat-b...  \n",
       "331  [Automat-uberongraph(Trapi v1.3.0), Automat-bi...  \n",
       "333  [Automat-uberongraph(Trapi v1.3.0), Automat-on...  \n",
       "334  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "366  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  \n",
       "367  [Automat-uberongraph(Trapi v1.3.0), Automat-hm...  \n",
       "392  [Automat-uberongraph(Trapi v1.3.0), Automat-dr...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats_filt = api_stats.query(\"subject in @keep & object in @keep & subject <= object\")\n",
    "api_stats_filt.to_csv(\"results/api_stats_filt.tsv\", sep=\"\\t\")\n",
    "api_stats_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0eda096-761d-486b-8bee-a00585a0774f",
   "metadata": {},
   "source": [
    "## Export to graphml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "84d2d692-7498-4e95-8c2f-7ad227941905",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_graph(df2, filename):\n",
    "    G = nx.Graph()\n",
    "\n",
    "    node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "    for node_type in node_types:\n",
    "        G.add_node(node_type, label = add_spacing(node_type))\n",
    "\n",
    "    for index,row in df2.iterrows():\n",
    "        G.add_edge(row['subject'], row['object'], weight=row['count'])\n",
    "    \n",
    "    nx.write_graphml(G, filename, infer_numeric_types=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ed33943d-4a08-4927-92e9-70df8f2dc190",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_spacing(str):\n",
    "    key = {\n",
    "        \"BiologicalProcess\":               \"Biological\\nProcess\",\n",
    "        \"ChemicalEntity\":                  \"Chemical\\nEntity\",\n",
    "        \"MolecularMixture\":                \"Molecular\\nMixture\", \n",
    "        \"PhysiologicalProcess\":            \"Physiological\\nProcess\",\n",
    "        \"SmallMolecule\":                   \"Small\\nMolecule\",\n",
    "        \"PhenotypicFeature\":               \"Phenotypic\\nFeature\",\n",
    "        'ChemicalExposure':                'Chemical\\nExposure',\n",
    "        'ClinicalAttribute':               'Clinical\\nAttribute',\n",
    "        'ClinicalIntervention':            'Clinical\\nIntervention',\n",
    "        'ComplexMolecularMixture':         'Complex\\nMolecular\\nMixture',\n",
    "        'EnvironmentalExposure':           'Environmental\\nExposure',\n",
    "        'InformationContentEntity':        'Information\\nContentEntity',\n",
    "        'MolecularMixture':                'Molecular\\nMixture',\n",
    "        'PhysiologicalProcess':            'Physiological\\nProcess',\n",
    "        'PopulationOfIndividualOrganisms': 'PopulationOf\\nIndividualOrganisms'\n",
    "    }\n",
    "    if str in key.keys():\n",
    "        return(key[str])\n",
    "    else:\n",
    "        return(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a53dc509-2c6c-4c70-ba6d-be046363e844",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "create_graph(api_stats_filt, \"results/api_stats_filt.graphml\")\n",
    "create_graph(predicate_stats_filt, \"results/predicate_stats_filt.graphml\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
