{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819c0140-24d4-4a83-9cdd-9f1f21a18d53",
   "metadata": {
    "tags": []
   },
   "source": [
    "# BTE metakg visualization\n",
    "\n",
    "The goal of this notebook is to visualize the metaKG that BTE uses.  Similar to the [subway diagram](https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/diagrams/smartapi_metagraph.png) we've used before, but updated to the size and scale of the current metakg.\n",
    "\n",
    "Optimizations\n",
    "* remove less-commonly-used node types from subject/object\n",
    "* only count in one direction (`A-treats-B` gets merged with `B-treated_by-A`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "472db681-b76f-48e7-b9f6-86eb863d3dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biothings_client\n",
    "import json5\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3b92f98-9a9e-4890-bf19-a58ccb51bc48",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in SmartAPI data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5ccf0d5-f69e-4256-9c4b-251b4f51efc0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Option 1 -- Read in the Smart API ndjson file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5af40473-2748-4a6c-b16f-c5db1a830b31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api</th>\n",
       "      <th>bte</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>superclass_of</td>\n",
       "      <td>infores:disease-ontology</td>\n",
       "      <td>{'name': 'Ontology Lookup Service API', 'smart...</td>\n",
       "      <td>{'query_operation': {'params': {'id': '{{ quer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SequenceVariant</td>\n",
       "      <td>Gene</td>\n",
       "      <td>is_sequence_variant_of</td>\n",
       "      <td>infores:dbsnp</td>\n",
       "      <td>{'name': 'LitVar API', 'smartapi': {'metadata'...</td>\n",
       "      <td>{'query_operation': {'params': {'variantid': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneOrGeneProduct</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367201</th>\n",
       "      <td>GeographicLocation</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367202</th>\n",
       "      <td>ClinicalIntervention</td>\n",
       "      <td>Activity</td>\n",
       "      <td>related_to</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367203</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367204</th>\n",
       "      <td>NamedThing</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367205</th>\n",
       "      <td>Agent</td>\n",
       "      <td>Agent</td>\n",
       "      <td>treats</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367206 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     subject            object               predicate  \\\n",
       "0                    Disease           Disease           superclass_of   \n",
       "1            SequenceVariant              Gene  is_sequence_variant_of   \n",
       "2          GeneOrGeneProduct    ChemicalEntity                has_part   \n",
       "3          NucleicAcidEntity    ChemicalEntity                has_part   \n",
       "4             ChemicalEntity    ChemicalEntity                has_part   \n",
       "...                      ...               ...                     ...   \n",
       "367201    GeographicLocation         Procedure            derives_from   \n",
       "367202  ClinicalIntervention          Activity              related_to   \n",
       "367203           Polypeptide  AnatomicalEntity           coexists_with   \n",
       "367204            NamedThing        Phenomenon                has_part   \n",
       "367205                 Agent             Agent                  treats   \n",
       "\n",
       "                     provided_by  \\\n",
       "0       infores:disease-ontology   \n",
       "1                  infores:dbsnp   \n",
       "2                            NaN   \n",
       "3                            NaN   \n",
       "4                            NaN   \n",
       "...                          ...   \n",
       "367201                       NaN   \n",
       "367202                       NaN   \n",
       "367203                       NaN   \n",
       "367204                       NaN   \n",
       "367205                       NaN   \n",
       "\n",
       "                                                      api  \\\n",
       "0       {'name': 'Ontology Lookup Service API', 'smart...   \n",
       "1       {'name': 'LitVar API', 'smartapi': {'metadata'...   \n",
       "2       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "3       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "4       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "...                                                   ...   \n",
       "367201  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367202  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367203  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367204  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367205  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                                                      bte  \n",
       "0       {'query_operation': {'params': {'id': '{{ quer...  \n",
       "1       {'query_operation': {'params': {'variantid': '...  \n",
       "2       {'query_operation': {'path': '/query', 'method...  \n",
       "3       {'query_operation': {'path': '/query', 'method...  \n",
       "4       {'query_operation': {'path': '/query', 'method...  \n",
       "...                                                   ...  \n",
       "367201  {'query_operation': {'path': '/query', 'method...  \n",
       "367202  {'query_operation': {'path': '/query', 'method...  \n",
       "367203  {'query_operation': {'path': '/query', 'method...  \n",
       "367204  {'query_operation': {'path': '/query', 'method...  \n",
       "367205  {'query_operation': {'path': '/query', 'method...  \n",
       "\n",
       "[367206 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# older input file from Chunlei\n",
    "#df = pd.read_json('data/smartapi_metakg_20230413.ndjson.gz', lines=True)\n",
    "\n",
    "# newer input file from Chunlei -- preprocessed using following command:\n",
    "#      gzip -cd smartapi_metakg_20230413.ndjson.gz | jq -c ' ._source' | gzip > smartapi_metakg_20230413b.ndjson.gz\n",
    "df = pd.read_json('data/smartapi_metakg_20230413b.ndjson.gz', lines=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fcb411e-fede-45f3-a2c5-4990eb924e32",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Option 2 (preferred) -- Query the SmartAPI API\n",
    "\n",
    "This method takes longer, but it retrieves the most up-to-date data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5644e1a0-fc78-473f-adf2-0184168862fb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "c = biothings_client.get_client('metakg', url='https://smart-api.info/api/metakg')\n",
    "c._query_endpoint=''\n",
    "a = c.query('*', fetch_all=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b8279669-8d85-4c00-8c26-71cda255bb73",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching 183941 metakg(s) . . .\n",
      "No more results to return.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>api</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54derYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6IderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6YderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6oderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64derYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183936</th>\n",
       "      <td>Z4pgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Agent</td>\n",
       "      <td>affects</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183937</th>\n",
       "      <td>aIpgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>directly_physically_interacts_with</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183938</th>\n",
       "      <td>aYpgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>IndividualOrganism</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>MolecularEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183939</th>\n",
       "      <td>aopgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Activity</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>GeneticInheritance</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183940</th>\n",
       "      <td>a4pgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>RNAProduct</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>183941 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         _id  _score  \\\n",
       "0       54derYkBXoKXcforKZub     1.0   \n",
       "1       6IderYkBXoKXcforKZub     1.0   \n",
       "2       6YderYkBXoKXcforKZub     1.0   \n",
       "3       6oderYkBXoKXcforKZub     1.0   \n",
       "4       64derYkBXoKXcforKZub     1.0   \n",
       "...                      ...     ...   \n",
       "183936  Z4pgrYkBXoKXcforP2oX     1.0   \n",
       "183937  aIpgrYkBXoKXcforP2oX     1.0   \n",
       "183938  aYpgrYkBXoKXcforP2oX     1.0   \n",
       "183939  aopgrYkBXoKXcforP2oX     1.0   \n",
       "183940  a4pgrYkBXoKXcforP2oX     1.0   \n",
       "\n",
       "                                                      api              object  \\\n",
       "0       {'name': 'CTD API', 'smartapi': {'id': '021261...                Gene   \n",
       "1       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "2       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "3       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "4       {'name': 'CTD API', 'smartapi': {'id': '021261...   BiologicalProcess   \n",
       "...                                                   ...                 ...   \n",
       "183936  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...               Agent   \n",
       "183937  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...                Gene   \n",
       "183938  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...  IndividualOrganism   \n",
       "183939  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...            Activity   \n",
       "183940  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...             Disease   \n",
       "\n",
       "                                 predicate                     subject  \\\n",
       "0                               related_to               SmallMolecule   \n",
       "1                               related_to                        Gene   \n",
       "2                               related_to                     Disease   \n",
       "3                               related_to                     Disease   \n",
       "4                               related_to               SmallMolecule   \n",
       "...                                    ...                         ...   \n",
       "183936                             affects  DiseaseOrPhenotypicFeature   \n",
       "183937  directly_physically_interacts_with                        Gene   \n",
       "183938                            disrupts             MolecularEntity   \n",
       "183939                           occurs_in          GeneticInheritance   \n",
       "183940                              treats                  RNAProduct   \n",
       "\n",
       "       provided_by  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "...            ...  \n",
       "183936         NaN  \n",
       "183937         NaN  \n",
       "183938         NaN  \n",
       "183939         NaN  \n",
       "183940         NaN  \n",
       "\n",
       "[183941 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(a)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2158f491-600f-48d6-bac4-9080ae188d8a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Post-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e981e0d-f064-4f5b-b7a3-50929a1e58b7",
   "metadata": {},
   "source": [
    "parse out a couple lines for the API name and ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "479dce85-28b0-45e2-b494-7756a1b42bed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>api</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54derYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6IderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6YderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6oderYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64derYkBXoKXcforKZub</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CTD API', 'smartapi': {'id': '021261...</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183936</th>\n",
       "      <td>Z4pgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Agent</td>\n",
       "      <td>affects</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.4.0</td>\n",
       "      <td>283042abacfe3c6bcdc924c4a226ff98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183937</th>\n",
       "      <td>aIpgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>directly_physically_interacts_with</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.4.0</td>\n",
       "      <td>283042abacfe3c6bcdc924c4a226ff98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183938</th>\n",
       "      <td>aYpgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>IndividualOrganism</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>MolecularEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.4.0</td>\n",
       "      <td>283042abacfe3c6bcdc924c4a226ff98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183939</th>\n",
       "      <td>aopgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Activity</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>GeneticInheritance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.4.0</td>\n",
       "      <td>283042abacfe3c6bcdc924c4a226ff98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183940</th>\n",
       "      <td>a4pgrYkBXoKXcforP2oX</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.4...</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>RNAProduct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.4.0</td>\n",
       "      <td>283042abacfe3c6bcdc924c4a226ff98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>183941 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         _id  _score  \\\n",
       "0       54derYkBXoKXcforKZub     1.0   \n",
       "1       6IderYkBXoKXcforKZub     1.0   \n",
       "2       6YderYkBXoKXcforKZub     1.0   \n",
       "3       6oderYkBXoKXcforKZub     1.0   \n",
       "4       64derYkBXoKXcforKZub     1.0   \n",
       "...                      ...     ...   \n",
       "183936  Z4pgrYkBXoKXcforP2oX     1.0   \n",
       "183937  aIpgrYkBXoKXcforP2oX     1.0   \n",
       "183938  aYpgrYkBXoKXcforP2oX     1.0   \n",
       "183939  aopgrYkBXoKXcforP2oX     1.0   \n",
       "183940  a4pgrYkBXoKXcforP2oX     1.0   \n",
       "\n",
       "                                                      api              object  \\\n",
       "0       {'name': 'CTD API', 'smartapi': {'id': '021261...                Gene   \n",
       "1       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "2       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "3       {'name': 'CTD API', 'smartapi': {'id': '021261...       SmallMolecule   \n",
       "4       {'name': 'CTD API', 'smartapi': {'id': '021261...   BiologicalProcess   \n",
       "...                                                   ...                 ...   \n",
       "183936  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...               Agent   \n",
       "183937  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...                Gene   \n",
       "183938  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...  IndividualOrganism   \n",
       "183939  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...            Activity   \n",
       "183940  {'name': 'ARAX Translator Reasoner - TRAPI 1.4...             Disease   \n",
       "\n",
       "                                 predicate                     subject  \\\n",
       "0                               related_to               SmallMolecule   \n",
       "1                               related_to                        Gene   \n",
       "2                               related_to                     Disease   \n",
       "3                               related_to                     Disease   \n",
       "4                               related_to               SmallMolecule   \n",
       "...                                    ...                         ...   \n",
       "183936                             affects  DiseaseOrPhenotypicFeature   \n",
       "183937  directly_physically_interacts_with                        Gene   \n",
       "183938                            disrupts             MolecularEntity   \n",
       "183939                           occurs_in          GeneticInheritance   \n",
       "183940                              treats                  RNAProduct   \n",
       "\n",
       "       provided_by                                api_name  \\\n",
       "0              NaN                                 CTD API   \n",
       "1              NaN                                 CTD API   \n",
       "2              NaN                                 CTD API   \n",
       "3              NaN                                 CTD API   \n",
       "4              NaN                                 CTD API   \n",
       "...            ...                                     ...   \n",
       "183936         NaN  ARAX Translator Reasoner - TRAPI 1.4.0   \n",
       "183937         NaN  ARAX Translator Reasoner - TRAPI 1.4.0   \n",
       "183938         NaN  ARAX Translator Reasoner - TRAPI 1.4.0   \n",
       "183939         NaN  ARAX Translator Reasoner - TRAPI 1.4.0   \n",
       "183940         NaN  ARAX Translator Reasoner - TRAPI 1.4.0   \n",
       "\n",
       "                                  api_id  \n",
       "0       0212611d1c670f9107baf00b77f0889a  \n",
       "1       0212611d1c670f9107baf00b77f0889a  \n",
       "2       0212611d1c670f9107baf00b77f0889a  \n",
       "3       0212611d1c670f9107baf00b77f0889a  \n",
       "4       0212611d1c670f9107baf00b77f0889a  \n",
       "...                                  ...  \n",
       "183936  283042abacfe3c6bcdc924c4a226ff98  \n",
       "183937  283042abacfe3c6bcdc924c4a226ff98  \n",
       "183938  283042abacfe3c6bcdc924c4a226ff98  \n",
       "183939  283042abacfe3c6bcdc924c4a226ff98  \n",
       "183940  283042abacfe3c6bcdc924c4a226ff98  \n",
       "\n",
       "[183941 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(api_name = lambda x: pd.json_normalize(x['api'])['name'])\n",
    "df = df.assign(api_id = lambda x: pd.json_normalize(x['api'])['smartapi.id'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aafe106-7f9a-4713-84fe-59ba83195f1b",
   "metadata": {},
   "source": [
    "## Read in the BTE config file that specifies currently-allowed APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "837e2003-2ac6-429b-a123-2715f1b93da0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       " {'id': '0212611d1c670f9107baf00b77f0889a',\n",
       "  'name': 'CTD API',\n",
       "  'primarySource': True},\n",
       " {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       " {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       " {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       " {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "  'name': 'Ontology Lookup Service API'},\n",
       " {'id': '38e9e5169a72aee3659c9ddba956790d', 'name': 'BioThings BindingDB API'},\n",
       " {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "  'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       " {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "  'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       " {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       " {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       " {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       " {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "  'name': 'BioThings EBIgene2phenotype API'},\n",
       " {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "  'name': 'BioThings GO Biological Process API'},\n",
       " {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "  'name': 'BioThings GO Cellular Component API'},\n",
       " {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "  'name': 'BioThings GO Molecular Function API'},\n",
       " {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       " {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       " {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       " {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "  'name': 'BioThings MGIgene2phenotype API'},\n",
       " {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       " {'id': 'b772ebfbfa536bba37764d7fddb11d6f',\n",
       "  'name': 'BioThings RARe-SOURCE API'},\n",
       " {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       " {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       " {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       " {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       " {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       " {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       " {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       " {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "  'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       " {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "  'name': 'Multiomics ClinicalTrials KP'},\n",
       " {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "  'name': 'Multiomics EHR Risk KP API'},\n",
       " {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "  'name': 'Multiomics Wellness KP API'},\n",
       " {'id': '978fe380a147a8641caf72320862697b',\n",
       "  'name': 'Text Mining Targeted Association API'},\n",
       " {'id': 'ef0656900ff73f861611bcad87a94bce',\n",
       "  'name': 'Automat-biolink(Trapi v1.4.0)'},\n",
       " {'id': '97da45e75266b021fae885735befad07',\n",
       "  'name': 'Automat-ctd(Trapi v1.4.0)'},\n",
       " {'id': 'a80b9c70e756453d1ce8971b59fe1778',\n",
       "  'name': 'Automat-drug-central(Trapi v1.4.0)'},\n",
       " {'id': '2575e053d0a631433b447995e1bc9602',\n",
       "  'name': 'Automat-gtex(Trapi v1.4.0)'},\n",
       " {'id': '387f7a2c21656ddfcce5ccf9ea459049',\n",
       "  'name': 'Automat-gtopdb(Trapi v1.4.0)'},\n",
       " {'id': 'cd9fc0ca8cc6d9f56bd56a34766de791',\n",
       "  'name': 'Automat-gwas-catalog(Trapi v1.4.0)'},\n",
       " {'id': '8a1e2c2eade9fe3a932ba1dbb7f85688',\n",
       "  'name': 'Automat-hetio(Trapi v1.4.0)'},\n",
       " {'id': '067d3a847117c6f42896cc8cd140a704',\n",
       "  'name': 'Automat-hgnc(Trapi v1.4.0)'},\n",
       " {'id': '0658e8749b9601a5faba5157ba12eb06',\n",
       "  'name': 'Automat-hmdb(Trapi v1.4.0)'},\n",
       " {'id': '43cf256c660cc5bdeac23fdd3063d474',\n",
       "  'name': 'Automat-human-goa(Trapi v1.4.0)'},\n",
       " {'id': '76a164ff43e7ab39a5b98a782f6361bf',\n",
       "  'name': 'Automat-icees-kg(Trapi v1.4.0)'},\n",
       " {'id': '0b0a4d48ccd9ad2fd34ee53c34f87e94',\n",
       "  'name': 'Automat-intact(Trapi v1.4.0)'},\n",
       " {'id': '26ca4939d437c411bcb65b85a9dc2b99',\n",
       "  'name': 'Automat-panther(Trapi v1.4.0)'},\n",
       " {'id': '1c71f68839a44b1b857e79ae7f7e3381',\n",
       "  'name': 'Automat-pharos(Trapi v1.4.0)'},\n",
       " {'id': '465ff6de7ddf35ca8b2df6c0b01e6554',\n",
       "  'name': 'Automat-viral-proteome(Trapi v1.4.0)'},\n",
       " {'id': 'af364143267ad5235bf78c1511223875', 'name': 'COHD TRAPI'},\n",
       " {'id': '23f770568b92b7a82063989b3ddd9706',\n",
       "  'name': 'Connections Hypothesis Provider API'}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_url = \"https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/src/config/apis.js\"\n",
    "r = requests.get(bte_config_url)\n",
    "str_bte_config = r.text\n",
    "#print(str_bte_config)\n",
    "str_bte_config = re.sub(\"exports.API_LIST = \",  \"\", str_bte_config)                       # remove variable assignment step\n",
    "str_bte_config = re.sub(\"\\s*//.*\",              \"\", str_bte_config)                       # remove commented lines\n",
    "str_bte_config = re.sub(r'^$\\n',                '', str_bte_config, flags=re.MULTILINE)   # remove blank lines\n",
    "str_bte_config = re.sub(r',\\s*exclude:[^\\]]*]', '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "str_bte_config = re.sub(r';$',                  '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "#print(str_bte_config)\n",
    "\n",
    "bte_config = json5.loads(str_bte_config)['include']\n",
    "bte_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "614a47e4-d78a-4275-bc61-051cc0e2ac27",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       " {'id': '0212611d1c670f9107baf00b77f0889a',\n",
       "  'name': 'CTD API',\n",
       "  'primarySource': True},\n",
       " {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       " {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       " {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       " {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "  'name': 'Ontology Lookup Service API'},\n",
       " {'id': '38e9e5169a72aee3659c9ddba956790d', 'name': 'BioThings BindingDB API'},\n",
       " {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "  'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       " {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "  'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       " {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       " {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       " {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       " {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "  'name': 'BioThings EBIgene2phenotype API'},\n",
       " {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "  'name': 'BioThings GO Biological Process API'},\n",
       " {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "  'name': 'BioThings GO Cellular Component API'},\n",
       " {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "  'name': 'BioThings GO Molecular Function API'},\n",
       " {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       " {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       " {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       " {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "  'name': 'BioThings MGIgene2phenotype API'},\n",
       " {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       " {'id': 'b772ebfbfa536bba37764d7fddb11d6f',\n",
       "  'name': 'BioThings RARe-SOURCE API'},\n",
       " {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       " {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       " {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       " {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       " {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       " {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       " {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       " {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "  'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       " {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "  'name': 'Multiomics ClinicalTrials KP'},\n",
       " {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "  'name': 'Multiomics EHR Risk KP API'},\n",
       " {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "  'name': 'Multiomics Wellness KP API'},\n",
       " {'id': '978fe380a147a8641caf72320862697b',\n",
       "  'name': 'Text Mining Targeted Association API'}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_notrapi = [ item for item in bte_config if not re.search(\"trapi\", item['name'], re.IGNORECASE)]\n",
    "\n",
    "# temporary hack on next line pending merge of https://github.com/biothings/biothings_explorer/pull/624\n",
    "bte_config_notrapi = [ item for item in bte_config_notrapi if item['name'] != \"Connections Hypothesis Provider API\" ]\n",
    "\n",
    "bte_config_notrapi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "689ae4af-34a1-4a5c-97e3-9f97fce9ef7e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34\n",
      "['d22b657426375a5295e7da8a303b9893', '0212611d1c670f9107baf00b77f0889a', '43af91b3d7cae43591083bff9d75c6dd', 'dca415f2d792976af9d642b7e73f7a41', '1f277e1563fcfd124bfae2cc3c4bcdec', '1c056ffc7ed0dd1229e71c4752239465', '38e9e5169a72aee3659c9ddba956790d', '55a223c6c6e0291dbd05f2faf27d16f4', 'b99c6dd64abcefe87dcd0a51c249ee6d', '00fb85fc776279163199e6c50f6ddfc6', 'e3edd325c76f2992a111b43a907a4870', 'a7f784626a426d054885a5f33f17d3f8', '1f47552dabd67351d4c625adb0a10d00', 'cc857d5b7c8b7609b5bbb38ff990bfff', 'f339b28426e7bf72028f60feefcd7465', '34bad236d77bea0a0ee6c6cba5be54a6', '316eab811fd9ef1097df98bcaa9f7361', 'a5b0ec6bfde5008984d4b6cde402d61f', '32f36164fabed5d3abe6c2fd899c9418', '77ed27f111262d0289ed4f4071faa619', 'edeb26858bd27d0322af93e7a9e08761', 'b772ebfbfa536bba37764d7fddb11d6f', '03283cc2b21c077be6794e1704b1d230', '1d288b3a3caf75d541ffaae3aab386c8', 'ec6d76016ef40f284359d17fbf78df20', '8f08d1446e0bb9c2b323713ce83e2bd3', '671b45c0301c8624abbd26ae78449ca2', '59dce17363dce279d389100834e43648', '09c8782d9f4027712e65b95424adba79', 'adf20dd6ff23dfe18e8e012bde686e31', '08a5ddcde71b4bf838327ef469076acd', 'd86a24f6027ffe778f84ba10a7a1861a', '02af7d098ab304e80d6f4806c3527027', '978fe380a147a8641caf72320862697b']\n"
     ]
    }
   ],
   "source": [
    "# uncomment one of the following two lines to look at all APIs, or just x-bte APIs\n",
    "#bte_config_ids = [ x['id'] for x in bte_config ]\n",
    "bte_config_ids = [ x['id'] for x in bte_config_notrapi ]\n",
    "\n",
    "print(len(bte_config_ids))\n",
    "print(bte_config_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96ca64b-7295-4039-a8c7-9fe1c9912473",
   "metadata": {},
   "source": [
    "## Join SmartAPI data with BTE config IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "462ed5f8-c015-438a-886c-650c4fdaa2fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>161530</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161531</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>condition_associated_with_gene</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161532</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>actively_involves</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161533</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161534</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SequenceVariant</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161568</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161550</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161554</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161566</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>contributes_to</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161552</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>treats</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1033 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              subject             object                       predicate  \\\n",
       "161530        Disease               Gene                       caused_by   \n",
       "161531        Disease               Gene  condition_associated_with_gene   \n",
       "161532        Disease            Pathway               actively_involves   \n",
       "161533        Disease  PhenotypicFeature                   has_phenotype   \n",
       "161534        Disease    SequenceVariant               contribution_from   \n",
       "...               ...                ...                             ...   \n",
       "161568  SmallMolecule            Disease                  contributes_to   \n",
       "161550  SmallMolecule            Disease                          treats   \n",
       "161554  SmallMolecule               Gene                         affects   \n",
       "161566  SmallMolecule  PhenotypicFeature                  contributes_to   \n",
       "161552  SmallMolecule  PhenotypicFeature                          treats   \n",
       "\n",
       "                                    api_name                            api_id  \n",
       "161530                           BioLink API  d22b657426375a5295e7da8a303b9893  \n",
       "161531                           BioLink API  d22b657426375a5295e7da8a303b9893  \n",
       "161532                           BioLink API  d22b657426375a5295e7da8a303b9893  \n",
       "161533                           BioLink API  d22b657426375a5295e7da8a303b9893  \n",
       "161534                           BioLink API  d22b657426375a5295e7da8a303b9893  \n",
       "...                                      ...                               ...  \n",
       "161568  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "161550  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "161554  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "161566  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "161552  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "\n",
       "[1033 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte = df.query('api_id in @bte_config_ids')[['subject','object','predicate','api_name','api_id']].drop_duplicates().sort_values(['api_name','subject','object','predicate'])\n",
    "df_bte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6764f282-a1e6-4b3c-936f-11b4d3971792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_bte.to_csv(\"results/bte_operations.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ce9ff2c-93df-4af2-bcd2-039f7f6a97e8",
   "metadata": {},
   "source": [
    "Note: `bte_operations.tsv` is included in the paper as a Supplemental Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ac384894-b9af-47c0-b0b5-e4920b30cf17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BioThings SEMMEDDB API                     826\n",
       "BioLink API                                 21\n",
       "Text Mining Targeted Association API        20\n",
       "MyDisease.info API                          17\n",
       "Multiomics EHR Risk KP API                  14\n",
       "Multiomics BigGIM-DrugResponse KP API       14\n",
       "MyChem.info API                             12\n",
       "MyGene.info API                             11\n",
       "MyVariant.info API                          10\n",
       "BioThings UBERON API                         9\n",
       "Multiomics Wellness KP API                   9\n",
       "CTD API                                      8\n",
       "BioThings IDISK API                          7\n",
       "BioThings PFOCR API                          6\n",
       "BioThings GO Biological Process API          6\n",
       "BioThings DGIdb API                          6\n",
       "BioThings GO Molecular Function API          4\n",
       "BioThings MGIgene2phenotype API              4\n",
       "BioThings GO Cellular Component API          4\n",
       "Multiomics ClinicalTrials KP                 2\n",
       "BioThings BioPlanet Pathway-Gene API         2\n",
       "BioThings BioPlanet Pathway-Disease API      2\n",
       "BioThings BindingDB API                      2\n",
       "BioThings RARe-SOURCE API                    2\n",
       "BioThings HPO API                            2\n",
       "BioThings GTRx API                           2\n",
       "BioThings EBIgene2phenotype API              2\n",
       "BioThings DISEASES API                       2\n",
       "BioThings Rhea API                           2\n",
       "LitVar API                                   1\n",
       "EBI Proteins API                             1\n",
       "BioThings DDInter API                        1\n",
       "Ontology Lookup Service API                  1\n",
       "QuickGO API                                  1\n",
       "Name: api_name, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte['api_name'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398517c7-4afd-4e3c-b9e4-993dce33e07d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Summarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515b78f8-2606-4c88-9403-60e0fc2e34c6",
   "metadata": {},
   "source": [
    "### by subject, object; count # of APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bde741c7-5a9d-4644-8f3c-13f86a0474f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>api_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>161530</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioLink API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161532</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>BioLink API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161533</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>BioLink API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161534</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SequenceVariant</td>\n",
       "      <td>BioLink API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161536</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>BioLink API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161573</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161567</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161568</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161554</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161566</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject             object  \\\n",
       "161530            Disease               Gene   \n",
       "161532            Disease            Pathway   \n",
       "161533            Disease  PhenotypicFeature   \n",
       "161534            Disease    SequenceVariant   \n",
       "161536               Gene            Disease   \n",
       "...                   ...                ...   \n",
       "161573  PhenotypicFeature               Gene   \n",
       "161567  PhenotypicFeature      SmallMolecule   \n",
       "161568      SmallMolecule            Disease   \n",
       "161554      SmallMolecule               Gene   \n",
       "161566      SmallMolecule  PhenotypicFeature   \n",
       "\n",
       "                                    api_name  \n",
       "161530                           BioLink API  \n",
       "161532                           BioLink API  \n",
       "161533                           BioLink API  \n",
       "161534                           BioLink API  \n",
       "161536                           BioLink API  \n",
       "...                                      ...  \n",
       "161573  Text Mining Targeted Association API  \n",
       "161567  Text Mining Targeted Association API  \n",
       "161568  Text Mining Targeted Association API  \n",
       "161554  Text Mining Targeted Association API  \n",
       "161566  Text Mining Targeted Association API  \n",
       "\n",
       "[326 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"api_name\"]].drop_duplicates()\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a53da5f5-34dd-4f43-bdad-855bc4139fb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[BioLink API, BioThings DISEASES API, BioThing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>12</td>\n",
       "      <td>[BioLink API, BioThings DISEASES API, BioThing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[BioThings GTRx API, BioThings IDISK API, BioT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>8</td>\n",
       "      <td>[BioThings BindingDB API, BioThings DGIdb API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>8</td>\n",
       "      <td>[BioThings BindingDB API, BioThings DGIdb API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>8</td>\n",
       "      <td>[BioThings GTRx API, BioThings IDISK API, BioT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>[BioLink API, BioThings SEMMEDDB API, Multiomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, BioThings SEMMEDDB API, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings SEMMEDDB API, Multiomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings MGIgene2phenotype API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings SEMMEDDB API, Multiomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings SEMMEDDB API, Multiomics EHR Risk K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings Rhea API, BioThings SEMMEDDB API, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, BioThings SEMMEDDB API, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Gene</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings MGIgene2phenotype API,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject             object  count  \\\n",
       "50             Disease               Gene     13   \n",
       "72                Gene            Disease     12   \n",
       "61             Disease      SmallMolecule      9   \n",
       "197      SmallMolecule               Gene      8   \n",
       "83                Gene      SmallMolecule      8   \n",
       "196      SmallMolecule            Disease      8   \n",
       "73                Gene               Gene      5   \n",
       "135  PhenotypicFeature      SmallMolecule      4   \n",
       "55             Disease  PhenotypicFeature      4   \n",
       "78                Gene  PhenotypicFeature      4   \n",
       "126  PhenotypicFeature            Disease      4   \n",
       "47             Disease            Disease      4   \n",
       "199      SmallMolecule  MolecularActivity      4   \n",
       "201      SmallMolecule  PhenotypicFeature      4   \n",
       "127  PhenotypicFeature               Gene      4   \n",
       "\n",
       "                                                  list  \n",
       "50   [BioLink API, BioThings DISEASES API, BioThing...  \n",
       "72   [BioLink API, BioThings DISEASES API, BioThing...  \n",
       "61   [BioThings GTRx API, BioThings IDISK API, BioT...  \n",
       "197  [BioThings BindingDB API, BioThings DGIdb API,...  \n",
       "83   [BioThings BindingDB API, BioThings DGIdb API,...  \n",
       "196  [BioThings GTRx API, BioThings IDISK API, BioT...  \n",
       "73   [BioLink API, BioThings SEMMEDDB API, Multiomi...  \n",
       "135  [BioThings IDISK API, BioThings SEMMEDDB API, ...  \n",
       "55   [BioLink API, BioThings SEMMEDDB API, Multiomi...  \n",
       "78   [BioLink API, BioThings MGIgene2phenotype API,...  \n",
       "126  [BioLink API, BioThings SEMMEDDB API, Multiomi...  \n",
       "47   [BioThings SEMMEDDB API, Multiomics EHR Risk K...  \n",
       "199  [BioThings Rhea API, BioThings SEMMEDDB API, C...  \n",
       "201  [BioThings IDISK API, BioThings SEMMEDDB API, ...  \n",
       "127  [BioLink API, BioThings MGIgene2phenotype API,...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats = df1.groupby(['subject','object'], group_keys=False)['api_name'].nunique().rename(\"count\").to_frame()\n",
    "api_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['api_name'].unique().apply(list)\n",
    "api_stats = api_stats.reset_index().sort_values(by=['count'],ascending=False)\n",
    "\n",
    "api_stats.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d73842dd-d6a2-4309-b820-1594a85c44d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "api_stats.to_csv(\"results/api_stats.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abbb70cd-070a-4380-b9e2-63cb798d09f8",
   "metadata": {},
   "source": [
    "### by subject, object; count # of predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cc40d6f8-87ed-48b1-b8a1-d3c04a9d8893",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>26</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>15</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>15</td>\n",
       "      <td>[interacts_with, affected_by, affects, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[treated_by, adverse_event_of, occurs_together...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>15</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>15</td>\n",
       "      <td>[treats, has_adverse_event, occurs_together_in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[caused_by, condition_associated_with_gene, re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[physically_interacts_with, affects, interacts...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>13</td>\n",
       "      <td>[causes, gene_associated_with_condition, relat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>13</td>\n",
       "      <td>[physically_interacts_with, affected_by, inter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>11</td>\n",
       "      <td>[interacts_with, affected_by, affects, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>11</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Disease</td>\n",
       "      <td>10</td>\n",
       "      <td>[affects, causes, disrupts, exacerbates, part_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[affected_by, caused_by, disrupted_by, has_par...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 subject               object  count  \\\n",
       "47               Disease              Disease     26   \n",
       "53               Disease  PathologicalProcess     15   \n",
       "73                  Gene                 Gene     15   \n",
       "61               Disease        SmallMolecule     15   \n",
       "111  PathologicalProcess              Disease     15   \n",
       "196        SmallMolecule              Disease     15   \n",
       "50               Disease                 Gene     13   \n",
       "197        SmallMolecule                 Gene     13   \n",
       "72                  Gene              Disease     13   \n",
       "83                  Gene        SmallMolecule     13   \n",
       "206        SmallMolecule        SmallMolecule     11   \n",
       "116  PathologicalProcess  PathologicalProcess     11   \n",
       "154          Polypeptide              Disease     10   \n",
       "57               Disease          Polypeptide     10   \n",
       "161          Polypeptide          Polypeptide     10   \n",
       "\n",
       "                                                  list  \n",
       "47   [affected_by, affects, caused_by, causes, coex...  \n",
       "53   [affected_by, affects, caused_by, causes, coex...  \n",
       "73   [interacts_with, affected_by, affects, coexist...  \n",
       "61   [treated_by, adverse_event_of, occurs_together...  \n",
       "111  [affected_by, affects, caused_by, causes, coex...  \n",
       "196  [treats, has_adverse_event, occurs_together_in...  \n",
       "50   [caused_by, condition_associated_with_gene, re...  \n",
       "197  [physically_interacts_with, affects, interacts...  \n",
       "72   [causes, gene_associated_with_condition, relat...  \n",
       "83   [physically_interacts_with, affected_by, inter...  \n",
       "206  [interacts_with, affected_by, affects, coexist...  \n",
       "116  [affected_by, affects, caused_by, causes, coex...  \n",
       "154  [affects, causes, disrupts, exacerbates, part_...  \n",
       "57   [affected_by, caused_by, disrupted_by, has_par...  \n",
       "161  [affected_by, affects, coexists_with, derives_...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"predicate\"]].drop_duplicates()\n",
    "predicate_stats = df1.groupby(['subject','object'], group_keys=False)['predicate'].nunique().rename(\"count\").to_frame()\n",
    "predicate_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['predicate'].unique().apply(list)\n",
    "predicate_stats = predicate_stats.reset_index().sort_values(by=['count'],ascending=False)\n",
    "\n",
    "predicate_stats.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0a2da419-97e7-438d-a294-69fa0946081f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "predicate_stats.to_csv(\"results/predicate_stats.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e955923-8436-4453-b380-2f8b8bb9ce3f",
   "metadata": {},
   "source": [
    "## Filter by most common types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9104cf-516b-4457-9632-0ff9a42ff036",
   "metadata": {},
   "source": [
    "Filter to only include the most common types of entities.  Also, since we _mostly_ have the same info in both directions, only keep one direction to simplify visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b2b9bffb-2178-4266-9b12-845b75412e6e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Disease                     325\n",
       "Gene                        256\n",
       "SmallMolecule               250\n",
       "Polypeptide                 182\n",
       "PathologicalProcess         174\n",
       "ChemicalEntity              132\n",
       "PhenotypicFeature           127\n",
       "PhysiologicalProcess        124\n",
       "Procedure                    80\n",
       "Cell                         72\n",
       "CellularComponent            71\n",
       "GrossAnatomicalStructure     70\n",
       "MolecularActivity            62\n",
       "Protein                      46\n",
       "AnatomicalEntity             18\n",
       "Drug                         16\n",
       "SequenceVariant              15\n",
       "BiologicalProcess            15\n",
       "Food                         14\n",
       "Pathway                       9\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ad7563bf-3c03-43ac-811c-186145df2acb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ChemicalEntity',\n",
       " 'Disease',\n",
       " 'Gene',\n",
       " 'PathologicalProcess',\n",
       " 'PhenotypicFeature',\n",
       " 'PhysiologicalProcess',\n",
       " 'Polypeptide',\n",
       " 'SmallMolecule'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NUM_TYPES_TO_KEEP = 8\n",
    "\n",
    "keep = set(pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4b29e182-3e47-4e8f-9925-3eb0803566a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>26</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>15</td>\n",
       "      <td>[interacts_with, affected_by, affects, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[treated_by, adverse_event_of, occurs_together...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>15</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[caused_by, condition_associated_with_gene, re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>13</td>\n",
       "      <td>[physically_interacts_with, affected_by, inter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>11</td>\n",
       "      <td>[interacts_with, affected_by, affects, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>11</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>10</td>\n",
       "      <td>[adverse_event_of, affected_by, caused_by, dis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[affected_by, caused_by, disrupted_by, has_par...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>9</td>\n",
       "      <td>[has_phenotype, caused_by, causes, coexists_wi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, affects, causes, coexists_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, caused_by, disrupted_by, has_pre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, caused_by, disrupted_by, has_pre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>8</td>\n",
       "      <td>[affects, causes, disrupts, predisposes, preve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>7</td>\n",
       "      <td>[affected_by, affects, caused_by, causes, coex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>7</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[gene_associated_with_condition, related_to, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>7</td>\n",
       "      <td>[affects, causes, disrupts, predisposes, preve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>7</td>\n",
       "      <td>[affected_by, caused_by, disrupted_by, has_pre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[subclass_of, superclass_of, caused_by, causes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>7</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>6</td>\n",
       "      <td>[affects, causes, disrupts, predisposes, preve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>6</td>\n",
       "      <td>[affected_by, affects, caused_by, disrupted_by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>6</td>\n",
       "      <td>[affected_by, affects, caused_by, disrupted_by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>6</td>\n",
       "      <td>[affected_by, affects, causes, disrupts, manif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>6</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>5</td>\n",
       "      <td>[affected_by, affects, coexists_with, derives_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>4</td>\n",
       "      <td>[affects, causes, disrupts, produced_by]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[causes, coexists_with, has_manifestation, pre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[affects, causes, predisposes, related_to]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>4</td>\n",
       "      <td>[affected_by, affects, causes, disrupts]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>3</td>\n",
       "      <td>[causes, manifestation_of, related_to]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "47                Disease               Disease     26   \n",
       "73                   Gene                  Gene     15   \n",
       "61                Disease         SmallMolecule     15   \n",
       "53                Disease   PathologicalProcess     15   \n",
       "50                Disease                  Gene     13   \n",
       "83                   Gene         SmallMolecule     13   \n",
       "206         SmallMolecule         SmallMolecule     11   \n",
       "116   PathologicalProcess   PathologicalProcess     11   \n",
       "161           Polypeptide           Polypeptide     10   \n",
       "135     PhenotypicFeature         SmallMolecule     10   \n",
       "57                Disease           Polypeptide     10   \n",
       "164           Polypeptide         SmallMolecule      9   \n",
       "55                Disease     PhenotypicFeature      9   \n",
       "56                Disease  PhysiologicalProcess      9   \n",
       "122   PathologicalProcess         SmallMolecule      9   \n",
       "80                   Gene           Polypeptide      9   \n",
       "119   PathologicalProcess           Polypeptide      9   \n",
       "31         ChemicalEntity               Disease      8   \n",
       "146  PhysiologicalProcess  PhysiologicalProcess      7   \n",
       "38         ChemicalEntity           Polypeptide      7   \n",
       "78                   Gene     PhenotypicFeature      7   \n",
       "35         ChemicalEntity   PathologicalProcess      7   \n",
       "132     PhenotypicFeature           Polypeptide      7   \n",
       "130     PhenotypicFeature     PhenotypicFeature      7   \n",
       "32         ChemicalEntity                  Gene      7   \n",
       "76                   Gene   PathologicalProcess      6   \n",
       "150  PhysiologicalProcess         SmallMolecule      6   \n",
       "147  PhysiologicalProcess           Polypeptide      6   \n",
       "118   PathologicalProcess  PhysiologicalProcess      6   \n",
       "30         ChemicalEntity        ChemicalEntity      6   \n",
       "40         ChemicalEntity         SmallMolecule      5   \n",
       "37         ChemicalEntity  PhysiologicalProcess      4   \n",
       "117   PathologicalProcess     PhenotypicFeature      4   \n",
       "36         ChemicalEntity     PhenotypicFeature      4   \n",
       "79                   Gene  PhysiologicalProcess      4   \n",
       "131     PhenotypicFeature  PhysiologicalProcess      3   \n",
       "\n",
       "                                                  list  \n",
       "47   [affected_by, affects, caused_by, causes, coex...  \n",
       "73   [interacts_with, affected_by, affects, coexist...  \n",
       "61   [treated_by, adverse_event_of, occurs_together...  \n",
       "53   [affected_by, affects, caused_by, causes, coex...  \n",
       "50   [caused_by, condition_associated_with_gene, re...  \n",
       "83   [physically_interacts_with, affected_by, inter...  \n",
       "206  [interacts_with, affected_by, affects, coexist...  \n",
       "116  [affected_by, affects, caused_by, causes, coex...  \n",
       "161  [affected_by, affects, coexists_with, derives_...  \n",
       "135  [adverse_event_of, affected_by, caused_by, dis...  \n",
       "57   [affected_by, caused_by, disrupted_by, has_par...  \n",
       "164  [affected_by, affects, coexists_with, derives_...  \n",
       "55   [has_phenotype, caused_by, causes, coexists_wi...  \n",
       "56   [affected_by, affects, causes, coexists_with, ...  \n",
       "122  [affected_by, caused_by, disrupted_by, has_pre...  \n",
       "80   [affected_by, affects, coexists_with, derives_...  \n",
       "119  [affected_by, caused_by, disrupted_by, has_pre...  \n",
       "31   [affects, causes, disrupts, predisposes, preve...  \n",
       "146  [affected_by, affects, caused_by, causes, coex...  \n",
       "38   [affected_by, affects, coexists_with, derives_...  \n",
       "78   [gene_associated_with_condition, related_to, a...  \n",
       "35   [affects, causes, disrupts, predisposes, preve...  \n",
       "132  [affected_by, caused_by, disrupted_by, has_pre...  \n",
       "130  [subclass_of, superclass_of, caused_by, causes...  \n",
       "32   [affected_by, affects, coexists_with, derives_...  \n",
       "76   [affects, causes, disrupts, predisposes, preve...  \n",
       "150  [affected_by, affects, caused_by, disrupted_by...  \n",
       "147  [affected_by, affects, caused_by, disrupted_by...  \n",
       "118  [affected_by, affects, causes, disrupts, manif...  \n",
       "30   [affected_by, affects, coexists_with, derives_...  \n",
       "40   [affected_by, affects, coexists_with, derives_...  \n",
       "37            [affects, causes, disrupts, produced_by]  \n",
       "117  [causes, coexists_with, has_manifestation, pre...  \n",
       "36          [affects, causes, predisposes, related_to]  \n",
       "79            [affected_by, affects, causes, disrupts]  \n",
       "131             [causes, manifestation_of, related_to]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicate_stats_filt = predicate_stats.query(\"subject in @keep & object in @keep & subject <= object\").sort_values(by=['count'], ascending=False)\n",
    "predicate_stats_filt.to_csv(\"results/predicate_stats_filt.tsv\", sep=\"\\t\")\n",
    "predicate_stats_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dabb9bd0-808b-4e46-a51c-ff15d25ac4dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[BioLink API, BioThings DISEASES API, BioThing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[BioThings GTRx API, BioThings IDISK API, BioT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>8</td>\n",
       "      <td>[BioThings BindingDB API, BioThings DGIdb API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>[BioLink API, BioThings SEMMEDDB API, Multiomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, BioThings SEMMEDDB API, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings SEMMEDDB API, Multiomi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioLink API, BioThings MGIgene2phenotype API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings SEMMEDDB API, Multiomics EHR Risk K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings DDInter API, BioThings IDISK API, B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>3</td>\n",
       "      <td>[BioThings HPO API, BioThings SEMMEDDB API, Mu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "50                Disease                  Gene     13   \n",
       "61                Disease         SmallMolecule      9   \n",
       "83                   Gene         SmallMolecule      8   \n",
       "73                   Gene                  Gene      5   \n",
       "135     PhenotypicFeature         SmallMolecule      4   \n",
       "55                Disease     PhenotypicFeature      4   \n",
       "78                   Gene     PhenotypicFeature      4   \n",
       "47                Disease               Disease      4   \n",
       "206         SmallMolecule         SmallMolecule      4   \n",
       "130     PhenotypicFeature     PhenotypicFeature      3   \n",
       "40         ChemicalEntity         SmallMolecule      1   \n",
       "32         ChemicalEntity                  Gene      1   \n",
       "35         ChemicalEntity   PathologicalProcess      1   \n",
       "37         ChemicalEntity  PhysiologicalProcess      1   \n",
       "38         ChemicalEntity           Polypeptide      1   \n",
       "76                   Gene   PathologicalProcess      1   \n",
       "53                Disease   PathologicalProcess      1   \n",
       "56                Disease  PhysiologicalProcess      1   \n",
       "57                Disease           Polypeptide      1   \n",
       "30         ChemicalEntity        ChemicalEntity      1   \n",
       "79                   Gene  PhysiologicalProcess      1   \n",
       "31         ChemicalEntity               Disease      1   \n",
       "150  PhysiologicalProcess         SmallMolecule      1   \n",
       "36         ChemicalEntity     PhenotypicFeature      1   \n",
       "116   PathologicalProcess   PathologicalProcess      1   \n",
       "164           Polypeptide         SmallMolecule      1   \n",
       "161           Polypeptide           Polypeptide      1   \n",
       "132     PhenotypicFeature           Polypeptide      1   \n",
       "131     PhenotypicFeature  PhysiologicalProcess      1   \n",
       "147  PhysiologicalProcess           Polypeptide      1   \n",
       "117   PathologicalProcess     PhenotypicFeature      1   \n",
       "118   PathologicalProcess  PhysiologicalProcess      1   \n",
       "146  PhysiologicalProcess  PhysiologicalProcess      1   \n",
       "119   PathologicalProcess           Polypeptide      1   \n",
       "122   PathologicalProcess         SmallMolecule      1   \n",
       "80                   Gene           Polypeptide      1   \n",
       "\n",
       "                                                  list  \n",
       "50   [BioLink API, BioThings DISEASES API, BioThing...  \n",
       "61   [BioThings GTRx API, BioThings IDISK API, BioT...  \n",
       "83   [BioThings BindingDB API, BioThings DGIdb API,...  \n",
       "73   [BioLink API, BioThings SEMMEDDB API, Multiomi...  \n",
       "135  [BioThings IDISK API, BioThings SEMMEDDB API, ...  \n",
       "55   [BioLink API, BioThings SEMMEDDB API, Multiomi...  \n",
       "78   [BioLink API, BioThings MGIgene2phenotype API,...  \n",
       "47   [BioThings SEMMEDDB API, Multiomics EHR Risk K...  \n",
       "206  [BioThings DDInter API, BioThings IDISK API, B...  \n",
       "130  [BioThings HPO API, BioThings SEMMEDDB API, Mu...  \n",
       "40                            [BioThings SEMMEDDB API]  \n",
       "32                            [BioThings SEMMEDDB API]  \n",
       "35                            [BioThings SEMMEDDB API]  \n",
       "37                            [BioThings SEMMEDDB API]  \n",
       "38                            [BioThings SEMMEDDB API]  \n",
       "76                            [BioThings SEMMEDDB API]  \n",
       "53                            [BioThings SEMMEDDB API]  \n",
       "56                            [BioThings SEMMEDDB API]  \n",
       "57                            [BioThings SEMMEDDB API]  \n",
       "30                            [BioThings SEMMEDDB API]  \n",
       "79                            [BioThings SEMMEDDB API]  \n",
       "31                            [BioThings SEMMEDDB API]  \n",
       "150                           [BioThings SEMMEDDB API]  \n",
       "36                            [BioThings SEMMEDDB API]  \n",
       "116                           [BioThings SEMMEDDB API]  \n",
       "164                           [BioThings SEMMEDDB API]  \n",
       "161                           [BioThings SEMMEDDB API]  \n",
       "132                           [BioThings SEMMEDDB API]  \n",
       "131                           [BioThings SEMMEDDB API]  \n",
       "147                           [BioThings SEMMEDDB API]  \n",
       "117                           [BioThings SEMMEDDB API]  \n",
       "118                           [BioThings SEMMEDDB API]  \n",
       "146                           [BioThings SEMMEDDB API]  \n",
       "119                           [BioThings SEMMEDDB API]  \n",
       "122                           [BioThings SEMMEDDB API]  \n",
       "80                            [BioThings SEMMEDDB API]  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats_filt = api_stats.query(\"subject in @keep & object in @keep & subject <= object\").sort_values(by=['count'], ascending=False)\n",
    "api_stats_filt.to_csv(\"results/api_stats_filt.tsv\", sep=\"\\t\")\n",
    "api_stats_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0eda096-761d-486b-8bee-a00585a0774f",
   "metadata": {},
   "source": [
    "## Export to graphml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "84d2d692-7498-4e95-8c2f-7ad227941905",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_graph(df2, filename):\n",
    "    G = nx.Graph()\n",
    "\n",
    "    node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "    for node_type in node_types:\n",
    "        G.add_node(node_type, label = add_spacing(node_type))\n",
    "\n",
    "    for index,row in df2.iterrows():\n",
    "        if 'count' in row.keys():\n",
    "            G.add_edge(row['subject'], row['object'], weight=row['count'])\n",
    "        else:\n",
    "            G.add_edge(row['subject'], row['object'])\n",
    "    \n",
    "    nx.write_graphml(G, filename, infer_numeric_types=True)\n",
    "    return(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ed33943d-4a08-4927-92e9-70df8f2dc190",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_spacing(str):\n",
    "    key = {\n",
    "        \"BiologicalProcess\":               \"Biological\\nProcess\",\n",
    "        \"ChemicalEntity\":                  \"Chemical\\nEntity\",\n",
    "        \"MolecularMixture\":                \"Molecular\\nMixture\", \n",
    "        \"PhysiologicalProcess\":            \"Physiological\\nProcess\",\n",
    "        \"SmallMolecule\":                   \"Small\\nMolecule\",\n",
    "        \"PhenotypicFeature\":               \"Phenotypic\\nFeature\",\n",
    "        'ChemicalExposure':                'Chemical\\nExposure',\n",
    "        'ClinicalAttribute':               'Clinical\\nAttribute',\n",
    "        'ClinicalIntervention':            'Clinical\\nIntervention',\n",
    "        'ComplexMolecularMixture':         'Complex\\nMolecular\\nMixture',\n",
    "        'EnvironmentalExposure':           'Environmental\\nExposure',\n",
    "        'InformationContentEntity':        'Information\\nContentEntity',\n",
    "        'MolecularMixture':                'Molecular\\nMixture',\n",
    "        'PhysiologicalProcess':            'Physiological\\nProcess',\n",
    "        'PopulationOfIndividualOrganisms': 'PopulationOf\\nIndividualOrganisms',\n",
    "        'GrossAnatomicalStructure':        'Gross\\nAnatomical\\nStructure',\n",
    "        'PathologicalProcess':             'Pathological\\nProcess'\n",
    "    }\n",
    "    if str in key.keys():\n",
    "        return(key[str])\n",
    "    else:\n",
    "        return(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a53dc509-2c6c-4c70-ba6d-be046363e844",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "g = create_graph(api_stats_filt, \"results/api_stats_filt.graphml\")\n",
    "g = create_graph(predicate_stats_filt, \"results/predicate_stats_filt.graphml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "eb51ec31-8eaa-4c79-9f9f-fc42b51efcac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EdgeView([('Gene', 'Gene'), ('Gene', 'Disease'), ('Gene', 'SmallMolecule'), ('Gene', 'Polypeptide'), ('Gene', 'PhenotypicFeature'), ('Gene', 'ChemicalEntity'), ('Gene', 'PathologicalProcess'), ('Gene', 'PhysiologicalProcess'), ('Polypeptide', 'Polypeptide'), ('Polypeptide', 'Disease'), ('Polypeptide', 'SmallMolecule'), ('Polypeptide', 'PathologicalProcess'), ('Polypeptide', 'ChemicalEntity'), ('Polypeptide', 'PhenotypicFeature'), ('Polypeptide', 'PhysiologicalProcess'), ('SmallMolecule', 'Disease'), ('SmallMolecule', 'SmallMolecule'), ('SmallMolecule', 'PhenotypicFeature'), ('SmallMolecule', 'PathologicalProcess'), ('SmallMolecule', 'PhysiologicalProcess'), ('SmallMolecule', 'ChemicalEntity'), ('PhysiologicalProcess', 'Disease'), ('PhysiologicalProcess', 'PhysiologicalProcess'), ('PhysiologicalProcess', 'PathologicalProcess'), ('PhysiologicalProcess', 'ChemicalEntity'), ('PhysiologicalProcess', 'PhenotypicFeature'), ('PhenotypicFeature', 'Disease'), ('PhenotypicFeature', 'PhenotypicFeature'), ('PhenotypicFeature', 'PathologicalProcess'), ('PhenotypicFeature', 'ChemicalEntity'), ('Disease', 'Disease'), ('Disease', 'PathologicalProcess'), ('Disease', 'ChemicalEntity'), ('PathologicalProcess', 'PathologicalProcess'), ('PathologicalProcess', 'ChemicalEntity'), ('ChemicalEntity', 'ChemicalEntity')])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c668c6d-e27e-49b9-80a1-e4da3e05a620",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Graph' object has no attribute 'all_simple_paths'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[25], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m G \u001b[38;5;241m=\u001b[39m nx\u001b[38;5;241m.\u001b[39mGraph()\n\u001b[0;32m----> 2\u001b[0m \u001b[43mG\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mall_simple_paths\u001b[49m(g, source\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mChemicalEntity\u001b[39m\u001b[38;5;124m\"\u001b[39m, target\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDisease\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Graph' object has no attribute 'all_simple_paths'"
     ]
    }
   ],
   "source": [
    "G = nx.Graph()\n",
    "G.all_simple_paths(g, source=\"ChemicalEntity\", target=\"Disease\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790e6e61-eaac-4719-8e0d-a0315eaf9f62",
   "metadata": {},
   "source": [
    "## Restricted subway diagram\n",
    "\n",
    "Ultimately wasn't able to get a usable figure out of this, but keeping this section in case I want to experiment with this again later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12e2f4d-184c-4380-b87e-d4c381bb5675",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "NUM_TYPES_TO_KEEP = 5\n",
    "\n",
    "keep = set(pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d2a8dcf-4d78-4a6c-bf51-783796dd6299",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_subway = df_bte.query('object in @keep & subject in @keep & subject < object').drop(columns=['predicate']).drop_duplicates()\n",
    "df_subway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ab3bc7a-1673-411f-bd86-c5d8899eb61a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "G = nx.MultiGraph()\n",
    "\n",
    "df2 = df_subway\n",
    "\n",
    "node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "for node_type in node_types:\n",
    "    G.add_node(node_type, label = add_spacing(node_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e344e620-d5c4-44ac-bf8a-e8efaa2eec7f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "idcount = 0\n",
    "for index,row in df2.iterrows():\n",
    "    G.add_edge(row['subject'], row['object'], api_name=row['api_name'], key=idcount)\n",
    "    idcount = idcount + 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb3f8b64-c55b-476a-8386-e26be9c527b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "nx.write_graphml(G, \"results/subway.graphml\", infer_numeric_types=True,named_key_ids=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "522e69c3-a1e5-4f74-abd5-2aad0839f79d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
