{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819c0140-24d4-4a83-9cdd-9f1f21a18d53",
   "metadata": {
    "tags": []
   },
   "source": [
    "# BTE metakg visualization\n",
    "\n",
    "The goal of this notebook is to visualize the metaKG that BTE uses.  Similar to the [subway diagram](https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/diagrams/smartapi_metagraph.png) we've used before, but updated to the size and scale of the current metakg.\n",
    "\n",
    "Optimizations\n",
    "* remove less-commonly-used node types from subject/object\n",
    "* only count in one direction (`A-treats-B` gets merged with `B-treated_by-A`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "472db681-b76f-48e7-b9f6-86eb863d3dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biothings_client\n",
    "import json5\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3b92f98-9a9e-4890-bf19-a58ccb51bc48",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in SmartAPI data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5ccf0d5-f69e-4256-9c4b-251b4f51efc0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Option 1 -- Read in the Smart API ndjson file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5af40473-2748-4a6c-b16f-c5db1a830b31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api</th>\n",
       "      <th>bte</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>superclass_of</td>\n",
       "      <td>infores:disease-ontology</td>\n",
       "      <td>{'name': 'Ontology Lookup Service API', 'smart...</td>\n",
       "      <td>{'query_operation': {'params': {'id': '{{ quer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SequenceVariant</td>\n",
       "      <td>Gene</td>\n",
       "      <td>is_sequence_variant_of</td>\n",
       "      <td>infores:dbsnp</td>\n",
       "      <td>{'name': 'LitVar API', 'smartapi': {'metadata'...</td>\n",
       "      <td>{'query_operation': {'params': {'variantid': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneOrGeneProduct</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'Ontology-KP API', 'smartapi': {'meta...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367201</th>\n",
       "      <td>GeographicLocation</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367202</th>\n",
       "      <td>ClinicalIntervention</td>\n",
       "      <td>Activity</td>\n",
       "      <td>related_to</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367203</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367204</th>\n",
       "      <td>NamedThing</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367205</th>\n",
       "      <td>Agent</td>\n",
       "      <td>Agent</td>\n",
       "      <td>treats</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'path': '/query', 'method...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367206 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     subject            object               predicate  \\\n",
       "0                    Disease           Disease           superclass_of   \n",
       "1            SequenceVariant              Gene  is_sequence_variant_of   \n",
       "2          GeneOrGeneProduct    ChemicalEntity                has_part   \n",
       "3          NucleicAcidEntity    ChemicalEntity                has_part   \n",
       "4             ChemicalEntity    ChemicalEntity                has_part   \n",
       "...                      ...               ...                     ...   \n",
       "367201    GeographicLocation         Procedure            derives_from   \n",
       "367202  ClinicalIntervention          Activity              related_to   \n",
       "367203           Polypeptide  AnatomicalEntity           coexists_with   \n",
       "367204            NamedThing        Phenomenon                has_part   \n",
       "367205                 Agent             Agent                  treats   \n",
       "\n",
       "                     provided_by  \\\n",
       "0       infores:disease-ontology   \n",
       "1                  infores:dbsnp   \n",
       "2                            NaN   \n",
       "3                            NaN   \n",
       "4                            NaN   \n",
       "...                          ...   \n",
       "367201                       NaN   \n",
       "367202                       NaN   \n",
       "367203                       NaN   \n",
       "367204                       NaN   \n",
       "367205                       NaN   \n",
       "\n",
       "                                                      api  \\\n",
       "0       {'name': 'Ontology Lookup Service API', 'smart...   \n",
       "1       {'name': 'LitVar API', 'smartapi': {'metadata'...   \n",
       "2       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "3       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "4       {'name': 'Ontology-KP API', 'smartapi': {'meta...   \n",
       "...                                                   ...   \n",
       "367201  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367202  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367203  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367204  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "367205  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                                                      bte  \n",
       "0       {'query_operation': {'params': {'id': '{{ quer...  \n",
       "1       {'query_operation': {'params': {'variantid': '...  \n",
       "2       {'query_operation': {'path': '/query', 'method...  \n",
       "3       {'query_operation': {'path': '/query', 'method...  \n",
       "4       {'query_operation': {'path': '/query', 'method...  \n",
       "...                                                   ...  \n",
       "367201  {'query_operation': {'path': '/query', 'method...  \n",
       "367202  {'query_operation': {'path': '/query', 'method...  \n",
       "367203  {'query_operation': {'path': '/query', 'method...  \n",
       "367204  {'query_operation': {'path': '/query', 'method...  \n",
       "367205  {'query_operation': {'path': '/query', 'method...  \n",
       "\n",
       "[367206 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# older input file from Chunlei\n",
    "#df = pd.read_json('data/smartapi_metakg_20230413.ndjson.gz', lines=True)\n",
    "\n",
    "# newer input file from Chunlei -- preprocessed using following command:\n",
    "#      gzip -cd smartapi_metakg_20230413.ndjson.gz | jq -c ' ._source' | gzip > smartapi_metakg_20230413b.ndjson.gz\n",
    "df = pd.read_json('data/smartapi_metakg_20230413b.ndjson.gz', lines=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fcb411e-fede-45f3-a2c5-4990eb924e32",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Option 2 (preferred) -- Query the SmartAPI API\n",
    "\n",
    "This method takes longer, but it retrieves the most up-to-date data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5644e1a0-fc78-473f-adf2-0184168862fb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "c = biothings_client.get_client('metakg', url='https://smart-api.info/api/metakg')\n",
    "c._query_endpoint=''\n",
    "a = c.query('*', fetch_all=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b8279669-8d85-4c00-8c26-71cda255bb73",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching 338605 metakg(s) . . .\n",
      "No more results to return.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>api</th>\n",
       "      <th>bte</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1m-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2G-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2W-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>acts_upstream_of_negative_effect</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2m-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>acts_upstream_of_negative_effect</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338600</th>\n",
       "      <td>anO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>GeographicLocation</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338601</th>\n",
       "      <td>a3O0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Activity</td>\n",
       "      <td>related_to</td>\n",
       "      <td>ClinicalIntervention</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338602</th>\n",
       "      <td>bHO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338603</th>\n",
       "      <td>bXO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338604</th>\n",
       "      <td>bnO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Agent</td>\n",
       "      <td>treats</td>\n",
       "      <td>Agent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>338605 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         _id  _score  \\\n",
       "0       1m-wcYcBXoKXcforQySO     1.0   \n",
       "1       12-wcYcBXoKXcforQySO     1.0   \n",
       "2       2G-wcYcBXoKXcforQySO     1.0   \n",
       "3       2W-wcYcBXoKXcforQySO     1.0   \n",
       "4       2m-wcYcBXoKXcforQySO     1.0   \n",
       "...                      ...     ...   \n",
       "338600  anO0cYcBXoKXcforsf-o     1.0   \n",
       "338601  a3O0cYcBXoKXcforsf-o     1.0   \n",
       "338602  bHO0cYcBXoKXcforsf-o     1.0   \n",
       "338603  bXO0cYcBXoKXcforsf-o     1.0   \n",
       "338604  bnO0cYcBXoKXcforsf-o     1.0   \n",
       "\n",
       "                                                      api  \\\n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "...                                                   ...   \n",
       "338600  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338601  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338602  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338603  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338604  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                                                      bte  \\\n",
       "0       {'query_operation': {'input_separator': ',', '...   \n",
       "1       {'query_operation': {'input_separator': ',', '...   \n",
       "2       {'query_operation': {'input_separator': ',', '...   \n",
       "3       {'query_operation': {'input_separator': ',', '...   \n",
       "4       {'query_operation': {'input_separator': ',', '...   \n",
       "...                                                   ...   \n",
       "338600  {'query_operation': {'input_separator': ',', '...   \n",
       "338601  {'query_operation': {'input_separator': ',', '...   \n",
       "338602  {'query_operation': {'input_separator': ',', '...   \n",
       "338603  {'query_operation': {'input_separator': ',', '...   \n",
       "338604  {'query_operation': {'input_separator': ',', '...   \n",
       "\n",
       "                            object                         predicate  \\\n",
       "0       MacromolecularMachineMixin                  acts_upstream_of   \n",
       "1                MolecularActivity                  acts_upstream_of   \n",
       "2                       NamedThing                  acts_upstream_of   \n",
       "3                 BiologicalEntity  acts_upstream_of_negative_effect   \n",
       "4                BiologicalProcess  acts_upstream_of_negative_effect   \n",
       "...                            ...                               ...   \n",
       "338600                   Procedure                      derives_from   \n",
       "338601                    Activity                        related_to   \n",
       "338602            AnatomicalEntity                     coexists_with   \n",
       "338603                  Phenomenon                          has_part   \n",
       "338604                       Agent                            treats   \n",
       "\n",
       "                     subject provided_by  \n",
       "0              GenomicEntity         NaN  \n",
       "1              GenomicEntity         NaN  \n",
       "2              GenomicEntity         NaN  \n",
       "3              GenomicEntity         NaN  \n",
       "4              GenomicEntity         NaN  \n",
       "...                      ...         ...  \n",
       "338600    GeographicLocation         NaN  \n",
       "338601  ClinicalIntervention         NaN  \n",
       "338602           Polypeptide         NaN  \n",
       "338603            NamedThing         NaN  \n",
       "338604                 Agent         NaN  \n",
       "\n",
       "[338605 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(a)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2158f491-600f-48d6-bac4-9080ae188d8a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Post-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e981e0d-f064-4f5b-b7a3-50929a1e58b7",
   "metadata": {},
   "source": [
    "parse out a couple lines for the API name and ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "479dce85-28b0-45e2-b494-7756a1b42bed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>api</th>\n",
       "      <th>bte</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1m-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>MacromolecularMachineMixin</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>642b1dbcd0943f3a119d714cf7535fba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>642b1dbcd0943f3a119d714cf7535fba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2G-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>acts_upstream_of</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>642b1dbcd0943f3a119d714cf7535fba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2W-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>acts_upstream_of_negative_effect</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>642b1dbcd0943f3a119d714cf7535fba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2m-wcYcBXoKXcforQySO</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'CAM-KP API', 'smartapi': {'id': '642...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>acts_upstream_of_negative_effect</td>\n",
       "      <td>GenomicEntity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAM-KP API</td>\n",
       "      <td>642b1dbcd0943f3a119d714cf7535fba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338600</th>\n",
       "      <td>anO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>GeographicLocation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338601</th>\n",
       "      <td>a3O0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Activity</td>\n",
       "      <td>related_to</td>\n",
       "      <td>ClinicalIntervention</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338602</th>\n",
       "      <td>bHO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338603</th>\n",
       "      <td>bXO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Phenomenon</td>\n",
       "      <td>has_part</td>\n",
       "      <td>NamedThing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338604</th>\n",
       "      <td>bnO0cYcBXoKXcforsf-o</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'name': 'ARAX Translator Reasoner - TRAPI 1.3...</td>\n",
       "      <td>{'query_operation': {'input_separator': ',', '...</td>\n",
       "      <td>Agent</td>\n",
       "      <td>treats</td>\n",
       "      <td>Agent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAX Translator Reasoner - TRAPI 1.3.0</td>\n",
       "      <td>e248aefca0f469229e82cca80fbabc89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>338605 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         _id  _score  \\\n",
       "0       1m-wcYcBXoKXcforQySO     1.0   \n",
       "1       12-wcYcBXoKXcforQySO     1.0   \n",
       "2       2G-wcYcBXoKXcforQySO     1.0   \n",
       "3       2W-wcYcBXoKXcforQySO     1.0   \n",
       "4       2m-wcYcBXoKXcforQySO     1.0   \n",
       "...                      ...     ...   \n",
       "338600  anO0cYcBXoKXcforsf-o     1.0   \n",
       "338601  a3O0cYcBXoKXcforsf-o     1.0   \n",
       "338602  bHO0cYcBXoKXcforsf-o     1.0   \n",
       "338603  bXO0cYcBXoKXcforsf-o     1.0   \n",
       "338604  bnO0cYcBXoKXcforsf-o     1.0   \n",
       "\n",
       "                                                      api  \\\n",
       "0       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "1       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "2       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "3       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "4       {'name': 'CAM-KP API', 'smartapi': {'id': '642...   \n",
       "...                                                   ...   \n",
       "338600  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338601  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338602  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338603  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "338604  {'name': 'ARAX Translator Reasoner - TRAPI 1.3...   \n",
       "\n",
       "                                                      bte  \\\n",
       "0       {'query_operation': {'input_separator': ',', '...   \n",
       "1       {'query_operation': {'input_separator': ',', '...   \n",
       "2       {'query_operation': {'input_separator': ',', '...   \n",
       "3       {'query_operation': {'input_separator': ',', '...   \n",
       "4       {'query_operation': {'input_separator': ',', '...   \n",
       "...                                                   ...   \n",
       "338600  {'query_operation': {'input_separator': ',', '...   \n",
       "338601  {'query_operation': {'input_separator': ',', '...   \n",
       "338602  {'query_operation': {'input_separator': ',', '...   \n",
       "338603  {'query_operation': {'input_separator': ',', '...   \n",
       "338604  {'query_operation': {'input_separator': ',', '...   \n",
       "\n",
       "                            object                         predicate  \\\n",
       "0       MacromolecularMachineMixin                  acts_upstream_of   \n",
       "1                MolecularActivity                  acts_upstream_of   \n",
       "2                       NamedThing                  acts_upstream_of   \n",
       "3                 BiologicalEntity  acts_upstream_of_negative_effect   \n",
       "4                BiologicalProcess  acts_upstream_of_negative_effect   \n",
       "...                            ...                               ...   \n",
       "338600                   Procedure                      derives_from   \n",
       "338601                    Activity                        related_to   \n",
       "338602            AnatomicalEntity                     coexists_with   \n",
       "338603                  Phenomenon                          has_part   \n",
       "338604                       Agent                            treats   \n",
       "\n",
       "                     subject provided_by  \\\n",
       "0              GenomicEntity         NaN   \n",
       "1              GenomicEntity         NaN   \n",
       "2              GenomicEntity         NaN   \n",
       "3              GenomicEntity         NaN   \n",
       "4              GenomicEntity         NaN   \n",
       "...                      ...         ...   \n",
       "338600    GeographicLocation         NaN   \n",
       "338601  ClinicalIntervention         NaN   \n",
       "338602           Polypeptide         NaN   \n",
       "338603            NamedThing         NaN   \n",
       "338604                 Agent         NaN   \n",
       "\n",
       "                                      api_name  \\\n",
       "0                                   CAM-KP API   \n",
       "1                                   CAM-KP API   \n",
       "2                                   CAM-KP API   \n",
       "3                                   CAM-KP API   \n",
       "4                                   CAM-KP API   \n",
       "...                                        ...   \n",
       "338600  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "338601  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "338602  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "338603  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "338604  ARAX Translator Reasoner - TRAPI 1.3.0   \n",
       "\n",
       "                                  api_id  \n",
       "0       642b1dbcd0943f3a119d714cf7535fba  \n",
       "1       642b1dbcd0943f3a119d714cf7535fba  \n",
       "2       642b1dbcd0943f3a119d714cf7535fba  \n",
       "3       642b1dbcd0943f3a119d714cf7535fba  \n",
       "4       642b1dbcd0943f3a119d714cf7535fba  \n",
       "...                                  ...  \n",
       "338600  e248aefca0f469229e82cca80fbabc89  \n",
       "338601  e248aefca0f469229e82cca80fbabc89  \n",
       "338602  e248aefca0f469229e82cca80fbabc89  \n",
       "338603  e248aefca0f469229e82cca80fbabc89  \n",
       "338604  e248aefca0f469229e82cca80fbabc89  \n",
       "\n",
       "[338605 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(api_name = lambda x: pd.json_normalize(x['api'])['name'])\n",
    "df = df.assign(api_id = lambda x: pd.json_normalize(x['api'])['smartapi.id'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aafe106-7f9a-4713-84fe-59ba83195f1b",
   "metadata": {},
   "source": [
    "## Read in the BTE config file that specifies currently-allowed APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "837e2003-2ac6-429b-a123-2715f1b93da0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       " {'id': '0212611d1c670f9107baf00b77f0889a',\n",
       "  'name': 'CTD API',\n",
       "  'primarySource': True},\n",
       " {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       " {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       " {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       " {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "  'name': 'Ontology Lookup Service API'},\n",
       " {'id': '38e9e5169a72aee3659c9ddba956790d', 'name': 'BioThings BindingDB API'},\n",
       " {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "  'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       " {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "  'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       " {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       " {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       " {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       " {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "  'name': 'BioThings EBIgene2phenotype API'},\n",
       " {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "  'name': 'BioThings GO Biological Process API'},\n",
       " {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "  'name': 'BioThings GO Cellular Component API'},\n",
       " {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "  'name': 'BioThings GO Molecular Function API'},\n",
       " {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       " {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       " {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       " {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "  'name': 'BioThings MGIgene2phenotype API'},\n",
       " {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       " {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       " {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       " {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       " {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       " {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       " {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       " {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       " {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "  'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       " {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "  'name': 'Multiomics ClinicalTrials KP'},\n",
       " {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "  'name': 'Multiomics EHR Risk KP API',\n",
       "  'primarySource': True},\n",
       " {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "  'name': 'Multiomics Wellness KP API',\n",
       "  'primarySource': True},\n",
       " {'id': '978fe380a147a8641caf72320862697b',\n",
       "  'name': 'Text Mining Targeted Association API'},\n",
       " {'id': '25085b05fd1afcebb497724d147cfb44',\n",
       "  'name': 'Automat-biolink(Trapi v1.3.0)'},\n",
       " {'id': '89aa98098ba4329aa49b43ff8d21ffbb',\n",
       "  'name': 'Automat-ctd(Trapi v1.3.0)'},\n",
       " {'id': '463525b319d7e1ef8f705bce953d01bd',\n",
       "  'name': 'Automat-cord19(Trapi v1.3.0)'},\n",
       " {'id': '539873b1f2f2eb913efaee411e09eaa7',\n",
       "  'name': 'Automat-drug-central(Trapi v1.3.0)'},\n",
       " {'id': '272c2357763e4faf737f1cba94beaa1d',\n",
       "  'name': 'Automat-gtex(Trapi v1.3.0)'},\n",
       " {'id': 'fb1f3780ad67b030cb0617363afa8f61',\n",
       "  'name': 'Automat-gwas-catalog(Trapi v1.3.0)'},\n",
       " {'id': 'aacdd9e863bb659574d302f084084976',\n",
       "  'name': 'Automat-gtopdb(Trapi v1.3.0)'},\n",
       " {'id': '7382f0fabffce3cc7f7b8b6358c69259',\n",
       "  'name': 'Automat-hgnc(Trapi v1.3.0)'},\n",
       " {'id': 'ee2d2eae42ca30fe82946d3f42febaa0',\n",
       "  'name': 'Automat-hmdb(Trapi v1.3.0)'},\n",
       " {'id': '830600da121a5accc955cbf62e60f802',\n",
       "  'name': 'Automat-hetio(Trapi v1.3.0)'},\n",
       " {'id': 'f30339a6894a146a19a34974712ca2e3',\n",
       "  'name': 'Automat-human-goa(Trapi v1.3.0)'},\n",
       " {'id': '44e7a1147ca8657f50af6bb25982762d',\n",
       "  'name': 'Automat-icees-kg(Trapi v1.3.0)'},\n",
       " {'id': 'e0687431a9ff88344d20e83e0c99ee7d',\n",
       "  'name': 'Automat-intact(Trapi v1.3.0)'},\n",
       " {'id': '4c6f9117581531161849e60ea906f0be',\n",
       "  'name': 'Automat-ontology-hierarchy(Trapi v1.3.0)'},\n",
       " {'id': '2a879882329b000c7e7f08c2d71ccffd',\n",
       "  'name': 'Automat-panther(Trapi v1.3.0)'},\n",
       " {'id': '0295640b4d060133e4296dda4c31da47',\n",
       "  'name': 'Automat-pharos(Trapi v1.3.0)'},\n",
       " {'id': 'ef9027a7d2246c6540cc7b3ce202d89f',\n",
       "  'name': 'Automat-uberongraph(Trapi v1.3.0)'},\n",
       " {'id': '79310edd01cae96f3d8495250e625886',\n",
       "  'name': 'Automat-viral-proteome(Trapi v1.3.0)'},\n",
       " {'id': '51c178099fa2dc99b5d8fff8bf9f1a0d', 'name': 'COHD TRAPI 1.3'},\n",
       " {'id': '855adaa128ce5aa58a091d99e520d396',\n",
       "  'name': 'Connections Hypothesis Provider API'}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_url = \"https://raw.githubusercontent.com/biothings/BioThings_Explorer_TRAPI/main/src/config/apis.js\"\n",
    "r = requests.get(bte_config_url)\n",
    "str_bte_config = r.text\n",
    "#print(str_bte_config)\n",
    "str_bte_config = re.sub(\"exports.API_LIST = \",  \"\", str_bte_config)                       # remove variable assignment step\n",
    "str_bte_config = re.sub(\"\\s*//.*\",              \"\", str_bte_config)                       # remove commented lines\n",
    "str_bte_config = re.sub(r'^$\\n',                '', str_bte_config, flags=re.MULTILINE)   # remove blank lines\n",
    "str_bte_config = re.sub(r',\\s*exclude:[^\\]]*]', '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "str_bte_config = re.sub(r';$',                  '', str_bte_config, flags=re.MULTILINE)   # remove 'exclude' section\n",
    "#print(str_bte_config)\n",
    "\n",
    "bte_config = json5.loads(str_bte_config)['include']\n",
    "bte_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "614a47e4-d78a-4275-bc61-051cc0e2ac27",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'd22b657426375a5295e7da8a303b9893', 'name': 'BioLink API'},\n",
       " {'id': '0212611d1c670f9107baf00b77f0889a',\n",
       "  'name': 'CTD API',\n",
       "  'primarySource': True},\n",
       " {'id': '43af91b3d7cae43591083bff9d75c6dd', 'name': 'EBI Proteins API'},\n",
       " {'id': 'dca415f2d792976af9d642b7e73f7a41', 'name': 'LitVar API'},\n",
       " {'id': '1f277e1563fcfd124bfae2cc3c4bcdec', 'name': 'QuickGO API'},\n",
       " {'id': '1c056ffc7ed0dd1229e71c4752239465',\n",
       "  'name': 'Ontology Lookup Service API'},\n",
       " {'id': '38e9e5169a72aee3659c9ddba956790d', 'name': 'BioThings BindingDB API'},\n",
       " {'id': '55a223c6c6e0291dbd05f2faf27d16f4',\n",
       "  'name': 'BioThings BioPlanet Pathway-Disease API'},\n",
       " {'id': 'b99c6dd64abcefe87dcd0a51c249ee6d',\n",
       "  'name': 'BioThings BioPlanet Pathway-Gene API'},\n",
       " {'id': '00fb85fc776279163199e6c50f6ddfc6', 'name': 'BioThings DDInter API'},\n",
       " {'id': 'e3edd325c76f2992a111b43a907a4870', 'name': 'BioThings DGIdb API'},\n",
       " {'id': 'a7f784626a426d054885a5f33f17d3f8', 'name': 'BioThings DISEASES API'},\n",
       " {'id': '1f47552dabd67351d4c625adb0a10d00',\n",
       "  'name': 'BioThings EBIgene2phenotype API'},\n",
       " {'id': 'cc857d5b7c8b7609b5bbb38ff990bfff',\n",
       "  'name': 'BioThings GO Biological Process API'},\n",
       " {'id': 'f339b28426e7bf72028f60feefcd7465',\n",
       "  'name': 'BioThings GO Cellular Component API'},\n",
       " {'id': '34bad236d77bea0a0ee6c6cba5be54a6',\n",
       "  'name': 'BioThings GO Molecular Function API'},\n",
       " {'id': '316eab811fd9ef1097df98bcaa9f7361', 'name': 'BioThings GTRx API'},\n",
       " {'id': 'a5b0ec6bfde5008984d4b6cde402d61f', 'name': 'BioThings HPO API'},\n",
       " {'id': '32f36164fabed5d3abe6c2fd899c9418', 'name': 'BioThings IDISK API'},\n",
       " {'id': '77ed27f111262d0289ed4f4071faa619',\n",
       "  'name': 'BioThings MGIgene2phenotype API'},\n",
       " {'id': 'edeb26858bd27d0322af93e7a9e08761', 'name': 'BioThings PFOCR API'},\n",
       " {'id': '03283cc2b21c077be6794e1704b1d230', 'name': 'BioThings Rhea API'},\n",
       " {'id': '1d288b3a3caf75d541ffaae3aab386c8', 'name': 'BioThings SEMMEDDB API'},\n",
       " {'id': 'ec6d76016ef40f284359d17fbf78df20', 'name': 'BioThings UBERON API'},\n",
       " {'id': '8f08d1446e0bb9c2b323713ce83e2bd3', 'name': 'MyChem.info API'},\n",
       " {'id': '671b45c0301c8624abbd26ae78449ca2', 'name': 'MyDisease.info API'},\n",
       " {'id': '59dce17363dce279d389100834e43648', 'name': 'MyGene.info API'},\n",
       " {'id': '09c8782d9f4027712e65b95424adba79', 'name': 'MyVariant.info API'},\n",
       " {'id': 'adf20dd6ff23dfe18e8e012bde686e31',\n",
       "  'name': 'Multiomics BigGIM-DrugResponse KP API'},\n",
       " {'id': '08a5ddcde71b4bf838327ef469076acd',\n",
       "  'name': 'Multiomics ClinicalTrials KP'},\n",
       " {'id': 'd86a24f6027ffe778f84ba10a7a1861a',\n",
       "  'name': 'Multiomics EHR Risk KP API',\n",
       "  'primarySource': True},\n",
       " {'id': '02af7d098ab304e80d6f4806c3527027',\n",
       "  'name': 'Multiomics Wellness KP API',\n",
       "  'primarySource': True},\n",
       " {'id': '978fe380a147a8641caf72320862697b',\n",
       "  'name': 'Text Mining Targeted Association API'}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bte_config_notrapi = [ item for item in bte_config if not re.search(\"trapi\", item['name'], re.IGNORECASE)]\n",
    "\n",
    "# temporary hack on next line pending merge of https://github.com/biothings/biothings_explorer/pull/624\n",
    "bte_config_notrapi = [ item for item in bte_config_notrapi if item['name'] != \"Connections Hypothesis Provider API\" ]\n",
    "\n",
    "bte_config_notrapi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "689ae4af-34a1-4a5c-97e3-9f97fce9ef7e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33\n",
      "['d22b657426375a5295e7da8a303b9893', '0212611d1c670f9107baf00b77f0889a', '43af91b3d7cae43591083bff9d75c6dd', 'dca415f2d792976af9d642b7e73f7a41', '1f277e1563fcfd124bfae2cc3c4bcdec', '1c056ffc7ed0dd1229e71c4752239465', '38e9e5169a72aee3659c9ddba956790d', '55a223c6c6e0291dbd05f2faf27d16f4', 'b99c6dd64abcefe87dcd0a51c249ee6d', '00fb85fc776279163199e6c50f6ddfc6', 'e3edd325c76f2992a111b43a907a4870', 'a7f784626a426d054885a5f33f17d3f8', '1f47552dabd67351d4c625adb0a10d00', 'cc857d5b7c8b7609b5bbb38ff990bfff', 'f339b28426e7bf72028f60feefcd7465', '34bad236d77bea0a0ee6c6cba5be54a6', '316eab811fd9ef1097df98bcaa9f7361', 'a5b0ec6bfde5008984d4b6cde402d61f', '32f36164fabed5d3abe6c2fd899c9418', '77ed27f111262d0289ed4f4071faa619', 'edeb26858bd27d0322af93e7a9e08761', '03283cc2b21c077be6794e1704b1d230', '1d288b3a3caf75d541ffaae3aab386c8', 'ec6d76016ef40f284359d17fbf78df20', '8f08d1446e0bb9c2b323713ce83e2bd3', '671b45c0301c8624abbd26ae78449ca2', '59dce17363dce279d389100834e43648', '09c8782d9f4027712e65b95424adba79', 'adf20dd6ff23dfe18e8e012bde686e31', '08a5ddcde71b4bf838327ef469076acd', 'd86a24f6027ffe778f84ba10a7a1861a', '02af7d098ab304e80d6f4806c3527027', '978fe380a147a8641caf72320862697b']\n"
     ]
    }
   ],
   "source": [
    "# uncomment one of the following two lines to look at all APIs, or just x-bte APIs\n",
    "#bte_config_ids = [ x['id'] for x in bte_config ]\n",
    "bte_config_ids = [ x['id'] for x in bte_config_notrapi ]\n",
    "\n",
    "print(len(bte_config_ids))\n",
    "print(bte_config_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96ca64b-7295-4039-a8c7-9fe1c9912473",
   "metadata": {},
   "source": [
    "## Join SmartAPI data with BTE config IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "462ed5f8-c015-438a-886c-650c4fdaa2fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>predicate</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17393</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17394</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17395</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17397</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17398</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CellularComponent</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293770</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>contribution_from</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293773</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>affects</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293774</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293775</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>affects</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293776</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1041 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject             object          predicate  \\\n",
       "17393       SmallMolecule               Gene         related_to   \n",
       "17394                Gene      SmallMolecule         related_to   \n",
       "17395             Disease      SmallMolecule         related_to   \n",
       "17397       SmallMolecule  BiologicalProcess         related_to   \n",
       "17398       SmallMolecule  CellularComponent         related_to   \n",
       "...                   ...                ...                ...   \n",
       "293770            Disease      SmallMolecule  contribution_from   \n",
       "293773               Gene            Disease            affects   \n",
       "293774            Disease               Gene        affected_by   \n",
       "293775               Gene  PhenotypicFeature            affects   \n",
       "293776  PhenotypicFeature               Gene        affected_by   \n",
       "\n",
       "                                    api_name                            api_id  \n",
       "17393                                CTD API  0212611d1c670f9107baf00b77f0889a  \n",
       "17394                                CTD API  0212611d1c670f9107baf00b77f0889a  \n",
       "17395                                CTD API  0212611d1c670f9107baf00b77f0889a  \n",
       "17397                                CTD API  0212611d1c670f9107baf00b77f0889a  \n",
       "17398                                CTD API  0212611d1c670f9107baf00b77f0889a  \n",
       "...                                      ...                               ...  \n",
       "293770  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "293773  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "293774  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "293775  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "293776  Text Mining Targeted Association API  978fe380a147a8641caf72320862697b  \n",
       "\n",
       "[1041 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte = df.query('api_id in @bte_config_ids')[['subject','object','predicate','api_name','api_id']].drop_duplicates()\n",
    "df_bte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6764f282-a1e6-4b3c-936f-11b4d3971792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_bte.to_csv(\"results/bte_operations.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ac384894-b9af-47c0-b0b5-e4920b30cf17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BioThings SEMMEDDB API                     826\n",
       "BioLink API                                 21\n",
       "Text Mining Targeted Association API        20\n",
       "MyDisease.info API                          17\n",
       "Multiomics Wellness KP API                  17\n",
       "Multiomics BigGIM-DrugResponse KP API       14\n",
       "Multiomics EHR Risk KP API                  14\n",
       "MyChem.info API                             12\n",
       "MyGene.info API                             11\n",
       "MyVariant.info API                          10\n",
       "BioThings UBERON API                         9\n",
       "CTD API                                      8\n",
       "BioThings IDISK API                          7\n",
       "BioThings GO Biological Process API          6\n",
       "BioThings PFOCR API                          6\n",
       "BioThings DGIdb API                          6\n",
       "BioThings GO Cellular Component API          4\n",
       "BioThings MGIgene2phenotype API              4\n",
       "BioThings GO Molecular Function API          4\n",
       "BioThings EBIgene2phenotype API              4\n",
       "BioThings HPO API                            2\n",
       "BioThings Rhea API                           2\n",
       "BioThings GTRx API                           2\n",
       "BioThings BioPlanet Pathway-Disease API      2\n",
       "BioThings BindingDB API                      2\n",
       "BioThings BioPlanet Pathway-Gene API         2\n",
       "Multiomics ClinicalTrials KP                 2\n",
       "BioThings DISEASES API                       2\n",
       "LitVar API                                   1\n",
       "EBI Proteins API                             1\n",
       "Ontology Lookup Service API                  1\n",
       "QuickGO API                                  1\n",
       "BioThings DDInter API                        1\n",
       "Name: api_name, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bte['api_name'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398517c7-4afd-4e3c-b9e4-993dce33e07d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Summarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515b78f8-2606-4c88-9403-60e0fc2e34c6",
   "metadata": {},
   "source": [
    "### by subject, object; count # of APIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bde741c7-5a9d-4644-8f3c-13f86a0474f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>api_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17393</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CTD API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17394</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CTD API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17395</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CTD API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17397</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BiologicalProcess</td>\n",
       "      <td>CTD API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17398</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CellularComponent</td>\n",
       "      <td>CTD API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293757</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293761</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293762</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293763</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293764</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject             object  \\\n",
       "17393       SmallMolecule               Gene   \n",
       "17394                Gene      SmallMolecule   \n",
       "17395             Disease      SmallMolecule   \n",
       "17397       SmallMolecule  BiologicalProcess   \n",
       "17398       SmallMolecule  CellularComponent   \n",
       "...                   ...                ...   \n",
       "293757               Gene               Gene   \n",
       "293761               Gene            Disease   \n",
       "293762            Disease               Gene   \n",
       "293763               Gene  PhenotypicFeature   \n",
       "293764  PhenotypicFeature               Gene   \n",
       "\n",
       "                                    api_name  \n",
       "17393                                CTD API  \n",
       "17394                                CTD API  \n",
       "17395                                CTD API  \n",
       "17397                                CTD API  \n",
       "17398                                CTD API  \n",
       "...                                      ...  \n",
       "293757  Text Mining Targeted Association API  \n",
       "293761  Text Mining Targeted Association API  \n",
       "293762  Text Mining Targeted Association API  \n",
       "293763  Text Mining Targeted Association API  \n",
       "293764  Text Mining Targeted Association API  \n",
       "\n",
       "[326 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"api_name\"]].drop_duplicates()\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a53da5f5-34dd-4f43-bdad-855bc4139fb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>12</td>\n",
       "      <td>[CTD API, MyDisease.info API, MyVariant.info A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>11</td>\n",
       "      <td>[MyDisease.info API, MyVariant.info API, MyGen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>9</td>\n",
       "      <td>[CTD API, MyChem.info API, BioThings BindingDB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[CTD API, MyChem.info API, BioThings BindingDB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[CTD API, MyDisease.info API, MyChem.info API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>8</td>\n",
       "      <td>[MyDisease.info API, MyChem.info API, BioThing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>6</td>\n",
       "      <td>[MyGene.info API, Multiomics Wellness KP API, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>5</td>\n",
       "      <td>[BioThings MGIgene2phenotype API, BioThings EB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>[BioThings MGIgene2phenotype API, BioThings EB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, Multiomics EHR Risk KP A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[MyDisease.info API, Multiomics EHR Risk KP AP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>4</td>\n",
       "      <td>[CTD API, BioThings Rhea API, MyChem.info API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[Ontology Lookup Service API, MyDisease.info A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, Multiomics EHR Risk KP A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[MyDisease.info API, Multiomics EHR Risk KP AP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject             object  count  \\\n",
       "50             Disease               Gene     12   \n",
       "73                Gene            Disease     11   \n",
       "196      SmallMolecule               Gene      9   \n",
       "84                Gene      SmallMolecule      9   \n",
       "61             Disease      SmallMolecule      9   \n",
       "195      SmallMolecule            Disease      8   \n",
       "74                Gene               Gene      6   \n",
       "79                Gene  PhenotypicFeature      5   \n",
       "128  PhenotypicFeature               Gene      5   \n",
       "136  PhenotypicFeature      SmallMolecule      4   \n",
       "127  PhenotypicFeature            Disease      4   \n",
       "198      SmallMolecule  MolecularActivity      4   \n",
       "47             Disease            Disease      4   \n",
       "200      SmallMolecule  PhenotypicFeature      4   \n",
       "55             Disease  PhenotypicFeature      4   \n",
       "\n",
       "                                                  list  \n",
       "50   [CTD API, MyDisease.info API, MyVariant.info A...  \n",
       "73   [MyDisease.info API, MyVariant.info API, MyGen...  \n",
       "196  [CTD API, MyChem.info API, BioThings BindingDB...  \n",
       "84   [CTD API, MyChem.info API, BioThings BindingDB...  \n",
       "61   [CTD API, MyDisease.info API, MyChem.info API,...  \n",
       "195  [MyDisease.info API, MyChem.info API, BioThing...  \n",
       "74   [MyGene.info API, Multiomics Wellness KP API, ...  \n",
       "79   [BioThings MGIgene2phenotype API, BioThings EB...  \n",
       "128  [BioThings MGIgene2phenotype API, BioThings EB...  \n",
       "136  [BioThings IDISK API, Multiomics EHR Risk KP A...  \n",
       "127  [MyDisease.info API, Multiomics EHR Risk KP AP...  \n",
       "198  [CTD API, BioThings Rhea API, MyChem.info API,...  \n",
       "47   [Ontology Lookup Service API, MyDisease.info A...  \n",
       "200  [BioThings IDISK API, Multiomics EHR Risk KP A...  \n",
       "55   [MyDisease.info API, Multiomics EHR Risk KP AP...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats = df1.groupby(['subject','object'], group_keys=False)['api_name'].nunique().rename(\"count\").to_frame()\n",
    "api_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['api_name'].unique().apply(list)\n",
    "api_stats = api_stats.reset_index().sort_values(by=['count'],ascending=False)\n",
    "\n",
    "api_stats.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d73842dd-d6a2-4309-b820-1594a85c44d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "api_stats.to_csv(\"results/api_stats.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abbb70cd-070a-4380-b9e2-63cb798d09f8",
   "metadata": {},
   "source": [
    "### by subject, object; count # of predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cc40d6f8-87ed-48b1-b8a1-d3c04a9d8893",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>26</td>\n",
       "      <td>[superclass_of, subclass_of, contributes_to, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>16</td>\n",
       "      <td>[homologous_to, correlated_with, physically_in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, treated_by, has_contraindication,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, interacts_with, physically_intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Disease</td>\n",
       "      <td>15</td>\n",
       "      <td>[coexists_with, caused_by, causes, affected_by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>15</td>\n",
       "      <td>[coexists_with, causes, caused_by, affects, pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Gene</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, interacts_with, affects, physical...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, treats, contraindicated_for, has_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[related_to, condition_associated_with_gene, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "      <td>13</td>\n",
       "      <td>[gene_associated_with_condition, affects, rela...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>12</td>\n",
       "      <td>[interacts_with, correlated_with, related_to, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>11</td>\n",
       "      <td>[coexists_with, causes, caused_by, affects, af...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>10</td>\n",
       "      <td>[has_adverse_event, associated_with, treats, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Disease</td>\n",
       "      <td>10</td>\n",
       "      <td>[related_to, causes, affects, predisposes, pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[interacts_with, affects, affected_by, coexist...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 subject               object  count  \\\n",
       "47               Disease              Disease     26   \n",
       "74                  Gene                 Gene     16   \n",
       "61               Disease        SmallMolecule     15   \n",
       "84                  Gene        SmallMolecule     15   \n",
       "112  PathologicalProcess              Disease     15   \n",
       "53               Disease  PathologicalProcess     15   \n",
       "196        SmallMolecule                 Gene     15   \n",
       "195        SmallMolecule              Disease     15   \n",
       "50               Disease                 Gene     13   \n",
       "73                  Gene              Disease     13   \n",
       "205        SmallMolecule        SmallMolecule     12   \n",
       "117  PathologicalProcess  PathologicalProcess     11   \n",
       "200        SmallMolecule    PhenotypicFeature     10   \n",
       "155          Polypeptide              Disease     10   \n",
       "162          Polypeptide          Polypeptide     10   \n",
       "\n",
       "                                                  list  \n",
       "47   [superclass_of, subclass_of, contributes_to, c...  \n",
       "74   [homologous_to, correlated_with, physically_in...  \n",
       "61   [related_to, treated_by, has_contraindication,...  \n",
       "84   [related_to, interacts_with, physically_intera...  \n",
       "112  [coexists_with, caused_by, causes, affected_by...  \n",
       "53   [coexists_with, causes, caused_by, affects, pr...  \n",
       "196  [related_to, interacts_with, affects, physical...  \n",
       "195  [related_to, treats, contraindicated_for, has_...  \n",
       "50   [related_to, condition_associated_with_gene, a...  \n",
       "73   [gene_associated_with_condition, affects, rela...  \n",
       "205  [interacts_with, correlated_with, related_to, ...  \n",
       "117  [coexists_with, causes, caused_by, affects, af...  \n",
       "200  [has_adverse_event, associated_with, treats, c...  \n",
       "155  [related_to, causes, affects, predisposes, pro...  \n",
       "162  [interacts_with, affects, affected_by, coexist...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df_bte[[\"subject\",\"object\",\"predicate\"]].drop_duplicates()\n",
    "predicate_stats = df1.groupby(['subject','object'], group_keys=False)['predicate'].nunique().rename(\"count\").to_frame()\n",
    "predicate_stats['list'] = df1.groupby(['subject','object'], group_keys=False)['predicate'].unique().apply(list)\n",
    "predicate_stats = predicate_stats.reset_index().sort_values(by=['count'],ascending=False)\n",
    "\n",
    "predicate_stats.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0a2da419-97e7-438d-a294-69fa0946081f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "predicate_stats.to_csv(\"results/predicate_stats.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e955923-8436-4453-b380-2f8b8bb9ce3f",
   "metadata": {},
   "source": [
    "## Filter by most common types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9104cf-516b-4457-9632-0ff9a42ff036",
   "metadata": {},
   "source": [
    "Filter to only include the most common types of entities.  Also, since we _mostly_ have the same info in both directions, only keep one direction to simplify visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b2b9bffb-2178-4266-9b12-845b75412e6e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Disease                     323\n",
       "Gene                        266\n",
       "SmallMolecule               256\n",
       "Polypeptide                 182\n",
       "PathologicalProcess         174\n",
       "ChemicalEntity              132\n",
       "PhenotypicFeature           129\n",
       "PhysiologicalProcess        124\n",
       "Procedure                    80\n",
       "Cell                         72\n",
       "CellularComponent            71\n",
       "GrossAnatomicalStructure     70\n",
       "MolecularActivity            62\n",
       "Protein                      40\n",
       "AnatomicalEntity             18\n",
       "Drug                         16\n",
       "BiologicalProcess            15\n",
       "SequenceVariant              15\n",
       "Food                         14\n",
       "ClinicalFinding              12\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ad7563bf-3c03-43ac-811c-186145df2acb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ChemicalEntity',\n",
       " 'Disease',\n",
       " 'Gene',\n",
       " 'PathologicalProcess',\n",
       " 'PhenotypicFeature',\n",
       " 'PhysiologicalProcess',\n",
       " 'Polypeptide',\n",
       " 'SmallMolecule'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NUM_TYPES_TO_KEEP = 8\n",
    "\n",
    "keep = set(pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4b29e182-3e47-4e8f-9925-3eb0803566a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>26</td>\n",
       "      <td>[superclass_of, subclass_of, contributes_to, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>16</td>\n",
       "      <td>[homologous_to, correlated_with, physically_in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, treated_by, has_contraindication,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>15</td>\n",
       "      <td>[related_to, interacts_with, physically_intera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>15</td>\n",
       "      <td>[coexists_with, causes, caused_by, affects, pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13</td>\n",
       "      <td>[related_to, condition_associated_with_gene, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>12</td>\n",
       "      <td>[interacts_with, correlated_with, related_to, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>11</td>\n",
       "      <td>[coexists_with, causes, caused_by, affects, af...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>10</td>\n",
       "      <td>[adverse_event_of, associated_with, treated_by...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[related_to, caused_by, affected_by, has_predi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>10</td>\n",
       "      <td>[interacts_with, affects, affected_by, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[treated_by, caused_by, related_to, prevented_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>9</td>\n",
       "      <td>[affected_by, coexists_with, interacts_with, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>9</td>\n",
       "      <td>[has_phenotype, associated_with, related_to, c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>9</td>\n",
       "      <td>[contains_process, related_to, causes, affects...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[has_part, interacts_with, affects, coexists_w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>9</td>\n",
       "      <td>[related_to, caused_by, affected_by, has_predi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>8</td>\n",
       "      <td>[related_to, prevents, causes, produced_by, af...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>7</td>\n",
       "      <td>[affects, interacts_with, produced_by, coexist...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>7</td>\n",
       "      <td>[coexists_with, affects, affected_by, causes, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>7</td>\n",
       "      <td>[related_to, affected_by, disrupted_by, has_pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[related_to, causes, affects, disrupts, predis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>7</td>\n",
       "      <td>[causes, related_to, produced_by, affects, pre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>7</td>\n",
       "      <td>[interacts_with, coexists_with, affected_by, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>7</td>\n",
       "      <td>[superclass_of, subclass_of, associated_with, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>6</td>\n",
       "      <td>[related_to, affects, causes, predisposes, dis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>6</td>\n",
       "      <td>[affected_by, disrupted_by, caused_by, affects...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>6</td>\n",
       "      <td>[caused_by, disrupted_by, affected_by, affects...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>6</td>\n",
       "      <td>[causes, affected_by, affects, precedes, disru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>6</td>\n",
       "      <td>[coexists_with, interacts_with, affects, affec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>5</td>\n",
       "      <td>[coexists_with, interacts_with, affects, affec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>4</td>\n",
       "      <td>[produced_by, affects, disrupts, causes]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[causes, related_to, affects, predisposes]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>4</td>\n",
       "      <td>[causes, affects, disrupts, affected_by]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[causes, predisposes, has_manifestation, coexi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>3</td>\n",
       "      <td>[related_to, causes, manifestation_of]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "47                Disease               Disease     26   \n",
       "74                   Gene                  Gene     16   \n",
       "61                Disease         SmallMolecule     15   \n",
       "84                   Gene         SmallMolecule     15   \n",
       "53                Disease   PathologicalProcess     15   \n",
       "50                Disease                  Gene     13   \n",
       "205         SmallMolecule         SmallMolecule     12   \n",
       "117   PathologicalProcess   PathologicalProcess     11   \n",
       "136     PhenotypicFeature         SmallMolecule     10   \n",
       "57                Disease           Polypeptide     10   \n",
       "162           Polypeptide           Polypeptide     10   \n",
       "123   PathologicalProcess         SmallMolecule      9   \n",
       "81                   Gene           Polypeptide      9   \n",
       "55                Disease     PhenotypicFeature      9   \n",
       "56                Disease  PhysiologicalProcess      9   \n",
       "165           Polypeptide         SmallMolecule      9   \n",
       "120   PathologicalProcess           Polypeptide      9   \n",
       "31         ChemicalEntity               Disease      8   \n",
       "32         ChemicalEntity                  Gene      7   \n",
       "147  PhysiologicalProcess  PhysiologicalProcess      7   \n",
       "133     PhenotypicFeature           Polypeptide      7   \n",
       "79                   Gene     PhenotypicFeature      7   \n",
       "35         ChemicalEntity   PathologicalProcess      7   \n",
       "38         ChemicalEntity           Polypeptide      7   \n",
       "131     PhenotypicFeature     PhenotypicFeature      7   \n",
       "77                   Gene   PathologicalProcess      6   \n",
       "148  PhysiologicalProcess           Polypeptide      6   \n",
       "151  PhysiologicalProcess         SmallMolecule      6   \n",
       "119   PathologicalProcess  PhysiologicalProcess      6   \n",
       "30         ChemicalEntity        ChemicalEntity      6   \n",
       "40         ChemicalEntity         SmallMolecule      5   \n",
       "37         ChemicalEntity  PhysiologicalProcess      4   \n",
       "36         ChemicalEntity     PhenotypicFeature      4   \n",
       "80                   Gene  PhysiologicalProcess      4   \n",
       "118   PathologicalProcess     PhenotypicFeature      4   \n",
       "132     PhenotypicFeature  PhysiologicalProcess      3   \n",
       "\n",
       "                                                  list  \n",
       "47   [superclass_of, subclass_of, contributes_to, c...  \n",
       "74   [homologous_to, correlated_with, physically_in...  \n",
       "61   [related_to, treated_by, has_contraindication,...  \n",
       "84   [related_to, interacts_with, physically_intera...  \n",
       "53   [coexists_with, causes, caused_by, affects, pr...  \n",
       "50   [related_to, condition_associated_with_gene, a...  \n",
       "205  [interacts_with, correlated_with, related_to, ...  \n",
       "117  [coexists_with, causes, caused_by, affects, af...  \n",
       "136  [adverse_event_of, associated_with, treated_by...  \n",
       "57   [related_to, caused_by, affected_by, has_predi...  \n",
       "162  [interacts_with, affects, affected_by, coexist...  \n",
       "123  [treated_by, caused_by, related_to, prevented_...  \n",
       "81   [affected_by, coexists_with, interacts_with, p...  \n",
       "55   [has_phenotype, associated_with, related_to, c...  \n",
       "56   [contains_process, related_to, causes, affects...  \n",
       "165  [has_part, interacts_with, affects, coexists_w...  \n",
       "120  [related_to, caused_by, affected_by, has_predi...  \n",
       "31   [related_to, prevents, causes, produced_by, af...  \n",
       "32   [affects, interacts_with, produced_by, coexist...  \n",
       "147  [coexists_with, affects, affected_by, causes, ...  \n",
       "133  [related_to, affected_by, disrupted_by, has_pr...  \n",
       "79   [related_to, causes, affects, disrupts, predis...  \n",
       "35   [causes, related_to, produced_by, affects, pre...  \n",
       "38   [interacts_with, coexists_with, affected_by, a...  \n",
       "131  [superclass_of, subclass_of, associated_with, ...  \n",
       "77   [related_to, affects, causes, predisposes, dis...  \n",
       "148  [affected_by, disrupted_by, caused_by, affects...  \n",
       "151  [caused_by, disrupted_by, affected_by, affects...  \n",
       "119  [causes, affected_by, affects, precedes, disru...  \n",
       "30   [coexists_with, interacts_with, affects, affec...  \n",
       "40   [coexists_with, interacts_with, affects, affec...  \n",
       "37            [produced_by, affects, disrupts, causes]  \n",
       "36          [causes, related_to, affects, predisposes]  \n",
       "80            [causes, affects, disrupts, affected_by]  \n",
       "118  [causes, predisposes, has_manifestation, coexi...  \n",
       "132             [related_to, causes, manifestation_of]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicate_stats_filt = predicate_stats.query(\"subject in @keep & object in @keep & subject <= object\").sort_values(by=['count'], ascending=False)\n",
    "predicate_stats_filt.to_csv(\"results/predicate_stats_filt.tsv\", sep=\"\\t\")\n",
    "predicate_stats_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dabb9bd0-808b-4e46-a51c-ff15d25ac4dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>count</th>\n",
       "      <th>list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>12</td>\n",
       "      <td>[CTD API, MyDisease.info API, MyVariant.info A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[CTD API, MyDisease.info API, MyChem.info API,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>9</td>\n",
       "      <td>[CTD API, MyChem.info API, BioThings BindingDB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>6</td>\n",
       "      <td>[MyGene.info API, Multiomics Wellness KP API, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>5</td>\n",
       "      <td>[BioThings MGIgene2phenotype API, BioThings EB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings IDISK API, Multiomics EHR Risk KP A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>4</td>\n",
       "      <td>[Ontology Lookup Service API, MyDisease.info A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>4</td>\n",
       "      <td>[MyDisease.info API, Multiomics EHR Risk KP AP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>4</td>\n",
       "      <td>[BioThings DDInter API, BioThings IDISK API, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>3</td>\n",
       "      <td>[BioThings HPO API, Multiomics EHR Risk KP API...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PhenotypicFeature</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>1</td>\n",
       "      <td>[BioThings SEMMEDDB API]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  subject                object  count  \\\n",
       "50                Disease                  Gene     12   \n",
       "61                Disease         SmallMolecule      9   \n",
       "84                   Gene         SmallMolecule      9   \n",
       "74                   Gene                  Gene      6   \n",
       "79                   Gene     PhenotypicFeature      5   \n",
       "136     PhenotypicFeature         SmallMolecule      4   \n",
       "47                Disease               Disease      4   \n",
       "55                Disease     PhenotypicFeature      4   \n",
       "205         SmallMolecule         SmallMolecule      4   \n",
       "131     PhenotypicFeature     PhenotypicFeature      3   \n",
       "40         ChemicalEntity         SmallMolecule      1   \n",
       "32         ChemicalEntity                  Gene      1   \n",
       "35         ChemicalEntity   PathologicalProcess      1   \n",
       "37         ChemicalEntity  PhysiologicalProcess      1   \n",
       "38         ChemicalEntity           Polypeptide      1   \n",
       "77                   Gene   PathologicalProcess      1   \n",
       "53                Disease   PathologicalProcess      1   \n",
       "56                Disease  PhysiologicalProcess      1   \n",
       "57                Disease           Polypeptide      1   \n",
       "30         ChemicalEntity        ChemicalEntity      1   \n",
       "80                   Gene  PhysiologicalProcess      1   \n",
       "31         ChemicalEntity               Disease      1   \n",
       "147  PhysiologicalProcess  PhysiologicalProcess      1   \n",
       "36         ChemicalEntity     PhenotypicFeature      1   \n",
       "165           Polypeptide         SmallMolecule      1   \n",
       "162           Polypeptide           Polypeptide      1   \n",
       "151  PhysiologicalProcess         SmallMolecule      1   \n",
       "148  PhysiologicalProcess           Polypeptide      1   \n",
       "132     PhenotypicFeature  PhysiologicalProcess      1   \n",
       "123   PathologicalProcess         SmallMolecule      1   \n",
       "120   PathologicalProcess           Polypeptide      1   \n",
       "119   PathologicalProcess  PhysiologicalProcess      1   \n",
       "118   PathologicalProcess     PhenotypicFeature      1   \n",
       "117   PathologicalProcess   PathologicalProcess      1   \n",
       "133     PhenotypicFeature           Polypeptide      1   \n",
       "81                   Gene           Polypeptide      1   \n",
       "\n",
       "                                                  list  \n",
       "50   [CTD API, MyDisease.info API, MyVariant.info A...  \n",
       "61   [CTD API, MyDisease.info API, MyChem.info API,...  \n",
       "84   [CTD API, MyChem.info API, BioThings BindingDB...  \n",
       "74   [MyGene.info API, Multiomics Wellness KP API, ...  \n",
       "79   [BioThings MGIgene2phenotype API, BioThings EB...  \n",
       "136  [BioThings IDISK API, Multiomics EHR Risk KP A...  \n",
       "47   [Ontology Lookup Service API, MyDisease.info A...  \n",
       "55   [MyDisease.info API, Multiomics EHR Risk KP AP...  \n",
       "205  [BioThings DDInter API, BioThings IDISK API, M...  \n",
       "131  [BioThings HPO API, Multiomics EHR Risk KP API...  \n",
       "40                            [BioThings SEMMEDDB API]  \n",
       "32                            [BioThings SEMMEDDB API]  \n",
       "35                            [BioThings SEMMEDDB API]  \n",
       "37                            [BioThings SEMMEDDB API]  \n",
       "38                            [BioThings SEMMEDDB API]  \n",
       "77                            [BioThings SEMMEDDB API]  \n",
       "53                            [BioThings SEMMEDDB API]  \n",
       "56                            [BioThings SEMMEDDB API]  \n",
       "57                            [BioThings SEMMEDDB API]  \n",
       "30                            [BioThings SEMMEDDB API]  \n",
       "80                            [BioThings SEMMEDDB API]  \n",
       "31                            [BioThings SEMMEDDB API]  \n",
       "147                           [BioThings SEMMEDDB API]  \n",
       "36                            [BioThings SEMMEDDB API]  \n",
       "165                           [BioThings SEMMEDDB API]  \n",
       "162                           [BioThings SEMMEDDB API]  \n",
       "151                           [BioThings SEMMEDDB API]  \n",
       "148                           [BioThings SEMMEDDB API]  \n",
       "132                           [BioThings SEMMEDDB API]  \n",
       "123                           [BioThings SEMMEDDB API]  \n",
       "120                           [BioThings SEMMEDDB API]  \n",
       "119                           [BioThings SEMMEDDB API]  \n",
       "118                           [BioThings SEMMEDDB API]  \n",
       "117                           [BioThings SEMMEDDB API]  \n",
       "133                           [BioThings SEMMEDDB API]  \n",
       "81                            [BioThings SEMMEDDB API]  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_stats_filt = api_stats.query(\"subject in @keep & object in @keep & subject <= object\").sort_values(by=['count'], ascending=False)\n",
    "api_stats_filt.to_csv(\"results/api_stats_filt.tsv\", sep=\"\\t\")\n",
    "api_stats_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0eda096-761d-486b-8bee-a00585a0774f",
   "metadata": {},
   "source": [
    "## Export to graphml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "84d2d692-7498-4e95-8c2f-7ad227941905",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_graph(df2, filename):\n",
    "    G = nx.Graph()\n",
    "\n",
    "    node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "    for node_type in node_types:\n",
    "        G.add_node(node_type, label = add_spacing(node_type))\n",
    "\n",
    "    for index,row in df2.iterrows():\n",
    "        if 'count' in row.keys():\n",
    "            G.add_edge(row['subject'], row['object'], weight=row['count'])\n",
    "        else:\n",
    "            G.add_edge(row['subject'], row['object'])\n",
    "    \n",
    "    nx.write_graphml(G, filename, infer_numeric_types=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ed33943d-4a08-4927-92e9-70df8f2dc190",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_spacing(str):\n",
    "    key = {\n",
    "        \"BiologicalProcess\":               \"Biological\\nProcess\",\n",
    "        \"ChemicalEntity\":                  \"Chemical\\nEntity\",\n",
    "        \"MolecularMixture\":                \"Molecular\\nMixture\", \n",
    "        \"PhysiologicalProcess\":            \"Physiological\\nProcess\",\n",
    "        \"SmallMolecule\":                   \"Small\\nMolecule\",\n",
    "        \"PhenotypicFeature\":               \"Phenotypic\\nFeature\",\n",
    "        'ChemicalExposure':                'Chemical\\nExposure',\n",
    "        'ClinicalAttribute':               'Clinical\\nAttribute',\n",
    "        'ClinicalIntervention':            'Clinical\\nIntervention',\n",
    "        'ComplexMolecularMixture':         'Complex\\nMolecular\\nMixture',\n",
    "        'EnvironmentalExposure':           'Environmental\\nExposure',\n",
    "        'InformationContentEntity':        'Information\\nContentEntity',\n",
    "        'MolecularMixture':                'Molecular\\nMixture',\n",
    "        'PhysiologicalProcess':            'Physiological\\nProcess',\n",
    "        'PopulationOfIndividualOrganisms': 'PopulationOf\\nIndividualOrganisms',\n",
    "        'GrossAnatomicalStructure':        'Gross\\nAnatomical\\nStructure',\n",
    "        'PathologicalProcess':             'Pathological\\nProcess'\n",
    "    }\n",
    "    if str in key.keys():\n",
    "        return(key[str])\n",
    "    else:\n",
    "        return(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a53dc509-2c6c-4c70-ba6d-be046363e844",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "create_graph(api_stats_filt, \"results/api_stats_filt.graphml\")\n",
    "create_graph(predicate_stats_filt, \"results/predicate_stats_filt.graphml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790e6e61-eaac-4719-8e0d-a0315eaf9f62",
   "metadata": {},
   "source": [
    "## Restricted subway diagram\n",
    "\n",
    "Ultimately wasn't able to get a usable figure out of this, but keeping this section in case I want to experiment with this again later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b12e2f4d-184c-4380-b87e-d4c381bb5675",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Disease', 'Gene', 'PathologicalProcess', 'Polypeptide', 'SmallMolecule'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NUM_TYPES_TO_KEEP = 5\n",
    "\n",
    "keep = set(pd.concat([df_bte['subject'], df_bte['object']]).value_counts().head(NUM_TYPES_TO_KEEP).keys())\n",
    "keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9d2a8dcf-4d78-4a6c-bf51-783796dd6299",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>api_name</th>\n",
       "      <th>api_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17394</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17395</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17400</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>0212611d1c670f9107baf00b77f0889a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65526</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MyDisease.info API</td>\n",
       "      <td>671b45c0301c8624abbd26ae78449ca2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65530</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>MyDisease.info API</td>\n",
       "      <td>671b45c0301c8624abbd26ae78449ca2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65553</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MyVariant.info API</td>\n",
       "      <td>09c8782d9f4027712e65b95424adba79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66886</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MyGene.info API</td>\n",
       "      <td>59dce17363dce279d389100834e43648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67276</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>MyChem.info API</td>\n",
       "      <td>8f08d1446e0bb9c2b323713ce83e2bd3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67282</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>MyChem.info API</td>\n",
       "      <td>8f08d1446e0bb9c2b323713ce83e2bd3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67312</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings BindingDB API</td>\n",
       "      <td>38e9e5169a72aee3659c9ddba956790d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67314</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings PFOCR API</td>\n",
       "      <td>edeb26858bd27d0322af93e7a9e08761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67316</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings PFOCR API</td>\n",
       "      <td>edeb26858bd27d0322af93e7a9e08761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67318</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioThings PFOCR API</td>\n",
       "      <td>edeb26858bd27d0322af93e7a9e08761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67320</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioThings DISEASES API</td>\n",
       "      <td>a7f784626a426d054885a5f33f17d3f8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67368</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioThings MGIgene2phenotype API</td>\n",
       "      <td>77ed27f111262d0289ed4f4071faa619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67380</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioThings EBIgene2phenotype API</td>\n",
       "      <td>1f47552dabd67351d4c625adb0a10d00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195392</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings GTRx API</td>\n",
       "      <td>316eab811fd9ef1097df98bcaa9f7361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195394</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings IDISK API</td>\n",
       "      <td>32f36164fabed5d3abe6c2fd899c9418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201181</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings DGIdb API</td>\n",
       "      <td>e3edd325c76f2992a111b43a907a4870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246095</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Multiomics Wellness KP API</td>\n",
       "      <td>02af7d098ab304e80d6f4806c3527027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291136</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Multiomics EHR Risk KP API</td>\n",
       "      <td>d86a24f6027ffe778f84ba10a7a1861a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291205</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Multiomics BigGIM-DrugResponse KP API</td>\n",
       "      <td>adf20dd6ff23dfe18e8e012bde686e31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291210</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Multiomics BigGIM-DrugResponse KP API</td>\n",
       "      <td>adf20dd6ff23dfe18e8e012bde686e31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291236</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291244</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291246</th>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291251</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291260</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291278</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291289</th>\n",
       "      <td>Disease</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291370</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291422</th>\n",
       "      <td>Gene</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291432</th>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>BioThings SEMMEDDB API</td>\n",
       "      <td>1d288b3a3caf75d541ffaae3aab386c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293721</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>d22b657426375a5295e7da8a303b9893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293742</th>\n",
       "      <td>Disease</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293750</th>\n",
       "      <td>Gene</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293762</th>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Text Mining Targeted Association API</td>\n",
       "      <td>978fe380a147a8641caf72320862697b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    subject               object  \\\n",
       "17394                  Gene        SmallMolecule   \n",
       "17395               Disease        SmallMolecule   \n",
       "17400               Disease                 Gene   \n",
       "65526               Disease                 Gene   \n",
       "65530               Disease        SmallMolecule   \n",
       "65553               Disease                 Gene   \n",
       "66886               Disease                 Gene   \n",
       "67276               Disease        SmallMolecule   \n",
       "67282                  Gene        SmallMolecule   \n",
       "67312                  Gene        SmallMolecule   \n",
       "67314                  Gene        SmallMolecule   \n",
       "67316               Disease        SmallMolecule   \n",
       "67318               Disease                 Gene   \n",
       "67320               Disease                 Gene   \n",
       "67368               Disease                 Gene   \n",
       "67380               Disease                 Gene   \n",
       "195392              Disease        SmallMolecule   \n",
       "195394              Disease        SmallMolecule   \n",
       "201181                 Gene        SmallMolecule   \n",
       "246095                 Gene        SmallMolecule   \n",
       "291136              Disease        SmallMolecule   \n",
       "291205                 Gene        SmallMolecule   \n",
       "291210              Disease                 Gene   \n",
       "291236              Disease        SmallMolecule   \n",
       "291244                 Gene          Polypeptide   \n",
       "291246          Polypeptide        SmallMolecule   \n",
       "291251              Disease          Polypeptide   \n",
       "291260              Disease                 Gene   \n",
       "291278                 Gene        SmallMolecule   \n",
       "291289              Disease  PathologicalProcess   \n",
       "291370  PathologicalProcess          Polypeptide   \n",
       "291422                 Gene  PathologicalProcess   \n",
       "291432  PathologicalProcess        SmallMolecule   \n",
       "293721              Disease                 Gene   \n",
       "293742              Disease        SmallMolecule   \n",
       "293750                 Gene        SmallMolecule   \n",
       "293762              Disease                 Gene   \n",
       "\n",
       "                                     api_name  \\\n",
       "17394                                 CTD API   \n",
       "17395                                 CTD API   \n",
       "17400                                 CTD API   \n",
       "65526                      MyDisease.info API   \n",
       "65530                      MyDisease.info API   \n",
       "65553                      MyVariant.info API   \n",
       "66886                         MyGene.info API   \n",
       "67276                         MyChem.info API   \n",
       "67282                         MyChem.info API   \n",
       "67312                 BioThings BindingDB API   \n",
       "67314                     BioThings PFOCR API   \n",
       "67316                     BioThings PFOCR API   \n",
       "67318                     BioThings PFOCR API   \n",
       "67320                  BioThings DISEASES API   \n",
       "67368         BioThings MGIgene2phenotype API   \n",
       "67380         BioThings EBIgene2phenotype API   \n",
       "195392                     BioThings GTRx API   \n",
       "195394                    BioThings IDISK API   \n",
       "201181                    BioThings DGIdb API   \n",
       "246095             Multiomics Wellness KP API   \n",
       "291136             Multiomics EHR Risk KP API   \n",
       "291205  Multiomics BigGIM-DrugResponse KP API   \n",
       "291210  Multiomics BigGIM-DrugResponse KP API   \n",
       "291236                 BioThings SEMMEDDB API   \n",
       "291244                 BioThings SEMMEDDB API   \n",
       "291246                 BioThings SEMMEDDB API   \n",
       "291251                 BioThings SEMMEDDB API   \n",
       "291260                 BioThings SEMMEDDB API   \n",
       "291278                 BioThings SEMMEDDB API   \n",
       "291289                 BioThings SEMMEDDB API   \n",
       "291370                 BioThings SEMMEDDB API   \n",
       "291422                 BioThings SEMMEDDB API   \n",
       "291432                 BioThings SEMMEDDB API   \n",
       "293721                            BioLink API   \n",
       "293742   Text Mining Targeted Association API   \n",
       "293750   Text Mining Targeted Association API   \n",
       "293762   Text Mining Targeted Association API   \n",
       "\n",
       "                                  api_id  \n",
       "17394   0212611d1c670f9107baf00b77f0889a  \n",
       "17395   0212611d1c670f9107baf00b77f0889a  \n",
       "17400   0212611d1c670f9107baf00b77f0889a  \n",
       "65526   671b45c0301c8624abbd26ae78449ca2  \n",
       "65530   671b45c0301c8624abbd26ae78449ca2  \n",
       "65553   09c8782d9f4027712e65b95424adba79  \n",
       "66886   59dce17363dce279d389100834e43648  \n",
       "67276   8f08d1446e0bb9c2b323713ce83e2bd3  \n",
       "67282   8f08d1446e0bb9c2b323713ce83e2bd3  \n",
       "67312   38e9e5169a72aee3659c9ddba956790d  \n",
       "67314   edeb26858bd27d0322af93e7a9e08761  \n",
       "67316   edeb26858bd27d0322af93e7a9e08761  \n",
       "67318   edeb26858bd27d0322af93e7a9e08761  \n",
       "67320   a7f784626a426d054885a5f33f17d3f8  \n",
       "67368   77ed27f111262d0289ed4f4071faa619  \n",
       "67380   1f47552dabd67351d4c625adb0a10d00  \n",
       "195392  316eab811fd9ef1097df98bcaa9f7361  \n",
       "195394  32f36164fabed5d3abe6c2fd899c9418  \n",
       "201181  e3edd325c76f2992a111b43a907a4870  \n",
       "246095  02af7d098ab304e80d6f4806c3527027  \n",
       "291136  d86a24f6027ffe778f84ba10a7a1861a  \n",
       "291205  adf20dd6ff23dfe18e8e012bde686e31  \n",
       "291210  adf20dd6ff23dfe18e8e012bde686e31  \n",
       "291236  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291244  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291246  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291251  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291260  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291278  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291289  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291370  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291422  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "291432  1d288b3a3caf75d541ffaae3aab386c8  \n",
       "293721  d22b657426375a5295e7da8a303b9893  \n",
       "293742  978fe380a147a8641caf72320862697b  \n",
       "293750  978fe380a147a8641caf72320862697b  \n",
       "293762  978fe380a147a8641caf72320862697b  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_subway = df_bte.query('object in @keep & subject in @keep & subject < object').drop(columns=['predicate']).drop_duplicates()\n",
    "df_subway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3ab3bc7a-1673-411f-bd86-c5d8899eb61a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "G = nx.MultiGraph()\n",
    "\n",
    "df2 = df_subway\n",
    "\n",
    "node_types = set(pd.concat([df2['subject'], df2['object']]))\n",
    "        \n",
    "for node_type in node_types:\n",
    "    G.add_node(node_type, label = add_spacing(node_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e344e620-d5c4-44ac-bf8a-e8efaa2eec7f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "idcount = 0\n",
    "for index,row in df2.iterrows():\n",
    "    G.add_edge(row['subject'], row['object'], api_name=row['api_name'], key=idcount)\n",
    "    idcount = idcount + 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cb3f8b64-c55b-476a-8386-e26be9c527b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "nx.write_graphml(G, \"results/subway.graphml\", infer_numeric_types=True,named_key_ids=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
